{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from eda_import import *\n",
    "\n",
    "from utils.parse_gtf_utils import extract_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def fisher_exact_wrapper(row):\n",
    "    contingency_table = np.array([\n",
    "            [row.N_on, row.N_off],\n",
    "            [row.T_on, row.T_off]])\n",
    "    _, pval = stats.fisher_exact(contingency_table)\n",
    "    return pval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def calc_fisher_exact_pval(grp):\n",
    "    count_by_sstype = lambda grp: grp[['sstype', 'analysis_id']].drop_duplicates().sstype.value_counts().astype(int)\n",
    "    \n",
    "    if grp.sstype.unique().shape[0] == 1:\n",
    "        return\n",
    "    sample_counts = count_by_sstype(grp).to_dict()\n",
    "    # These are the numbers of samples that have at least on mclv sites detected for this gene\n",
    "    num_normal = sample_counts['normal']\n",
    "    num_tumour = sample_counts['tumour']\n",
    "\n",
    "    grped_by_mclv_site = grp.groupby(['mkid', 'mclv', 'strand', 'aclv'])\n",
    "    # calculate the number of normal or tumour samples that have a mclv site\n",
    "    count_df = grped_by_mclv_site.apply(lambda g: count_by_sstype(g).to_frame().sort_index().T)\n",
    "    count_df.fillna(0, inplace=True)\n",
    "    \n",
    "    # rename cols and add off cols, preparing for input for Fishers exact test\n",
    "    count_df.rename(columns={'normal': 'N_on', 'tumour': 'T_on'}, inplace=True)\n",
    "    count_df['N_off'] = num_normal - count_df['N_on']\n",
    "    count_df['T_off'] = num_tumour - count_df['T_on']\n",
    "    count_df['fisher_exact_p'] = count_df.apply(fisher_exact_wrapper, axis=1)\n",
    "    \n",
    "    # add other helpful cols\n",
    "    count_df['N_base'] = num_normal\n",
    "    count_df['T_base'] = num_tumour\n",
    "    count_df['N_on_ratio'] = count_df['N_on'] / num_normal\n",
    "    count_df['T_on_ratio'] = count_df['T_on'] / num_tumour\n",
    "    \n",
    "    # from normal to tumour, does the on ratio go up?\n",
    "    count_df['N2T_ratio_diff'] = count_df['T_on_ratio'] - count_df['N_on_ratio']\n",
    "    count_df['diff_is_significant'] = count_df.apply(lambda row: True if row.fisher_exact_p < PVAL_CUTOFF else False, axis=1)\n",
    "    count_df['N2T_ratio_change'] = count_df.apply(lambda row: 'up' if row['N2T_ratio_diff'] > 0 else 'down', axis=1)\n",
    "    count_df['N2T_ratio_change_sig'] = count_df.apply(lambda row: row.N2T_ratio_change if row.fisher_exact_p < PVAL_CUTOFF else None, axis=1)\n",
    "    return count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def has_diff_aca(grp):\n",
    "    # it means for all mclv sites with pval < cutoff, and there are both an increase and a decrease\n",
    "    return grp.query('fisher_exact_p < {0}'.format(PVAL_CUTOFF)).N2T_ratio_change.unique().shape[0] > 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# diseases with sufficient normal samples (>= 15)\n",
    "NT_diseases = ['BRCA', 'KIRC', 'COAD', 'KICH', \n",
    "               'KIRP', 'LIHC', 'LUAD', 'LUSC',\n",
    "               'THCA', 'HNSC', 'STAD', 'UCEC',\n",
    "               'PRAD', 'BLCA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Load RPKMS\n",
    "df_expr = pd.read_csv('./results_data/RPKMS.csv.gz', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "PVAL_CUTOFF = 0.01"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 24.6 s, sys: 570 ms, total: 25.2 s\n",
      "Wall time: 25.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "_cols = ['analysis_id', 'barcode', 'disease', 'sstype', 'mkid',\n",
    "         'gene_name', 'strand', 'participant_id', 'mclv', 'aclv']\n",
    "adf = pd.read_csv('./results_data/all_cba.KLEAT.on-target-cleaned.filtered.clustered.csv.gz',\n",
    "                  usecols=_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# dedup is required after clustering to work on mclv\n",
    "assert adf.shape[0] == 3420253\n",
    "\n",
    "adf_dedup = adf.drop_duplicates()\n",
    "\n",
    "assert adf_dedup.shape[0] == 2213818\n",
    "\n",
    "assert adf_dedup.gene_name.unique().shape[0] == 114"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "bdf = adf_dedup.query('disease in {0}'.format(NT_diseases))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>analysis_id</th>\n",
       "      <th>barcode</th>\n",
       "      <th>disease</th>\n",
       "      <th>participant_id</th>\n",
       "      <th>mclv</th>\n",
       "      <th>aclv</th>\n",
       "      <th>sstype</th>\n",
       "      <th>mkid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>+</td>\n",
       "      <td>1d811d54-caa1-41f3-bfc2-400eebdd97ef</td>\n",
       "      <td>TCGA-KN-8425-01A-11R-2315-07</td>\n",
       "      <td>KICH</td>\n",
       "      <td>cc316faf-dcb1-4dea-b8b0-a17d000ff40f</td>\n",
       "      <td>133763062</td>\n",
       "      <td>133763062</td>\n",
       "      <td>tumour</td>\n",
       "      <td>chr9|ABL1|+|133763062</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name strand                           analysis_id                       barcode disease                        participant_id       mclv       aclv  sstype                   mkid\n",
       "0  ABL1      +      1d811d54-caa1-41f3-bfc2-400eebdd97ef  TCGA-KN-8425-01A-11R-2315-07  KICH    cc316faf-dcb1-4dea-b8b0-a17d000ff40f  133763062  133763062  tumour  chr9|ABL1|+|133763062"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bdf.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 35s, sys: 2.77 s, total: 1min 38s\n",
      "Wall time: 1min 38s\n"
     ]
    }
   ],
   "source": [
    "%time raw_pval_df = bdf.groupby(['disease', 'gene_name']).apply(calc_fisher_exact_pval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mkid</th>\n",
       "      <th>mclv</th>\n",
       "      <th>strand</th>\n",
       "      <th>aclv</th>\n",
       "      <th>level_6</th>\n",
       "      <th>N_on</th>\n",
       "      <th>T_on</th>\n",
       "      <th>N_off</th>\n",
       "      <th>T_off</th>\n",
       "      <th>fisher_exact_p</th>\n",
       "      <th>N_base</th>\n",
       "      <th>T_base</th>\n",
       "      <th>N_on_ratio</th>\n",
       "      <th>T_on_ratio</th>\n",
       "      <th>N2T_ratio_diff</th>\n",
       "      <th>diff_is_significant</th>\n",
       "      <th>N2T_ratio_change</th>\n",
       "      <th>N2T_ratio_change_sig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>AKT1</td>\n",
       "      <td>chr14|AKT1|-|105230220</td>\n",
       "      <td>105230220</td>\n",
       "      <td>-</td>\n",
       "      <td>105235686</td>\n",
       "      <td>sstype</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>413.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>19</td>\n",
       "      <td>414</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002415</td>\n",
       "      <td>0.002415</td>\n",
       "      <td>False</td>\n",
       "      <td>up</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>ALK</td>\n",
       "      <td>chr2|ALK|-|29737056</td>\n",
       "      <td>29737056</td>\n",
       "      <td>-</td>\n",
       "      <td>29443735</td>\n",
       "      <td>sstype</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8</td>\n",
       "      <td>115</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008696</td>\n",
       "      <td>0.008696</td>\n",
       "      <td>False</td>\n",
       "      <td>up</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>ALK</td>\n",
       "      <td>chr2|ALK|-|29743239</td>\n",
       "      <td>29743239</td>\n",
       "      <td>-</td>\n",
       "      <td>29443735</td>\n",
       "      <td>sstype</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8</td>\n",
       "      <td>115</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.008696</td>\n",
       "      <td>0.008696</td>\n",
       "      <td>False</td>\n",
       "      <td>up</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>APC</td>\n",
       "      <td>chr5|APC|+|112116697</td>\n",
       "      <td>112116697</td>\n",
       "      <td>+</td>\n",
       "      <td>112178452</td>\n",
       "      <td>sstype</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>378.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18</td>\n",
       "      <td>379</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002639</td>\n",
       "      <td>0.002639</td>\n",
       "      <td>False</td>\n",
       "      <td>up</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>APC</td>\n",
       "      <td>chr5|APC|+|112170731</td>\n",
       "      <td>112170731</td>\n",
       "      <td>+</td>\n",
       "      <td>112178452</td>\n",
       "      <td>sstype</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>376.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18</td>\n",
       "      <td>379</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007916</td>\n",
       "      <td>0.007916</td>\n",
       "      <td>False</td>\n",
       "      <td>up</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  disease gene_name                    mkid       mclv strand       aclv level_6  N_on  T_on  N_off  T_off  fisher_exact_p  N_base  T_base  N_on_ratio  T_on_ratio  N2T_ratio_diff  diff_is_significant N2T_ratio_change N2T_ratio_change_sig\n",
       "0  BLCA    AKT1      chr14|AKT1|-|105230220  105230220  -      105235686  sstype  0.0   1.0   19.0   413.0  1.0             19      414     0.0         0.002415    0.002415        False                up               None               \n",
       "1  BLCA    ALK       chr2|ALK|-|29737056     29737056   -      29443735   sstype  0.0   1.0   8.0    114.0  1.0             8       115     0.0         0.008696    0.008696        False                up               None               \n",
       "2  BLCA    ALK       chr2|ALK|-|29743239     29743239   -      29443735   sstype  0.0   1.0   8.0    114.0  1.0             8       115     0.0         0.008696    0.008696        False                up               None               \n",
       "3  BLCA    APC       chr5|APC|+|112116697    112116697  +      112178452  sstype  0.0   1.0   18.0   378.0  1.0             18      379     0.0         0.002639    0.002639        False                up               None               \n",
       "4  BLCA    APC       chr5|APC|+|112170731    112170731  +      112178452  sstype  0.0   3.0   18.0   376.0  1.0             18      379     0.0         0.007916    0.007916        False                up               None               "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_pval_df.query('N_on_ratio < 0.01').query('T_on_ratio < 0.01').reset_index().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAArQAAAElCAYAAAAV7HGhAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAAPYQAAD2EBqD+naQAAIABJREFUeJzt3X2cXVV97/HPDyLEAIamQKKlVLylMdreV81AJZdW6a2J\nqdVTtaUhhGomWAUTEHyZVO7lmgC2mpQnb5I+wJ0r9kIntA0dH4pkjBQlFqjMaC0yQURw1DzAEEMK\nQ8DA7/6x94Q9JzsPe8/ss89a+b5fr/OazN7rnLO+mZWV3+yz99rm7oiIiIiIhOqIujsgIiIiIjIW\nKmhFREREJGgqaEVEREQkaCpoRURERCRoKmhFREREJGgqaEVEREQkaCpoRURERCRoKmhFREREJGgq\naEVEREQkaCpoRSpkZm5mK+ruh4jI4cDMzk7n3bPr7ou0lgpaKc3MFqYTx24z+4Wc/Xeb2YN19K2V\nzOwdKlpFpBXSOfdQHmfX3dcqmdmHzWxh3f2Q9jGh7g5IFI4GPg5cXHdHavIOYDGwImffK4E9Le2N\niMTsj5u+fx8wO2f7QGu6U5sPA0PAzU3bv04y777Q6g5JvVTQynj4NvAnZvYpd99Sd2fGysyOcfdn\nx+O13H33eLyOiAiAu9+S/d7MzgRmN28PiZkZMNHdnxvra7n7S4Dm3cOQTjmQ8fDnwJEkR2nHxMz+\nu5ndY2bPmtlOM/u8mc1oarMi/Ujtl83s5rTd02b2WTObVPD9bjazZ8zsv5jZHWb2n8Ct6b7fMrN/\nMLNBM3vezH5kZteb2Suzzyc5Ojvqo8DM/n3OoTWzN5nZl81sV/reX03/UxIRabn04/vvpvPcFjNb\na2bHN7W528weNLM3mNm/mNmwmf3EzJaVeL/HzexLZvZ2M3sAeA74ULqv08zuMrMn0v48ZGYXNT8f\neCPw1sy8e3e6L/ccWjM7x8z6zOw5Mxsys1vyTpWTcOkIrYyHx4C/JTlK++myR2nN7G3Al4EfkHx8\n/0qS0xi+YWYz3f3xpqf8ffrelwMzgQ8ATwB/WvCtJwAbgE3Ax4DhdPs5wCTgr4CngN9I+3Nyug/g\nb4DXkP+RX17GNwL3ALuAVcDPSCbyu83sre5+f8G+i4iUlv7CvRzYSDLXTQcuAs4ws7Pc/WeZ5j8H\n3AncTjL//iGw0sz+w92/XPCtpwPdJHPoTcDD6faLgO8CXyA5XetdwF+a2RHuvjZtcymwGngG+LN0\n2/YDZFwIfBb4Jsn/F1OBjwBnmdmb3H1nwb5LO3J3PfQo9QAWAg6cDryOpDj7TGb/3cCDBV7vWyST\n0pTMtv8KvAh8LrNtRfq+XU3Pvx0YKpjh5vS1PpWz75U52z4OvAScktm2JvmnlPv6DqzIfP9PwPPA\n6zLbXk1S4H6t7p+pHnroEdbjQPPPITz3xHQ+2gAckdm+OJ27OjPb7k63/XFm21HAVuAfC77v4+lr\nvT1nX968eyfwaNO2B4G7c9qenb722en3r0j/X/kPktMaRtr9Xtruyrp/hnqMz0OnHMi4cPcfAP8P\n+KCZvbro89Pn/Dpws7vvyLzud4CvkFx41eyvm76/B/h5M3tV0fcnOTIximfO5zKzY8zsBOBfAQPe\nVPQNzOxIYA7Qk/59jbzPVuDvgN8s2XcRkTLeRlKU3uDJuacjbiL5Jfv3mto/A+w9V9fdXwD+jeSA\nRlGPufuG5o1N8+7kdN79GvA6M5tc4n1OB04C/tIz1zS4+z8Dm9k3owRKBa2Mp0+SfHxf5lzaX0q/\nPpyzbwA4wcyOado+2PT9T9OvP1fwvfcAP27eaGanpOfY7iCZyJ8kmVgBykysJ5KcwrC/jEcAv1ji\ndUVEysidd9NC9QeZ/SN+7O7etO2nFJ9zITldbB9mdpaZbTSzZ4GdJPPun6e7y8y7B/q/ZTP7ZpRA\n6RxaGTfu/gMzu4XkKO2nW/CWL+5nuxV8neebjk6MHE39CjAFWEky8T0L/ALJaQr6ZVBEDjfjNedC\nciHY6Bcx+y/AV0nm248CPyJZfusdwGVo3pUDUEEr4+2TwPkUvzDrh+nX6Tn7Xk9ybuy4LKV1iH4N\n+BXg/e7+tyMbzWx2TtvmIxb78yTJBWf7y/gSyQQuItIK2Xl372lQZnYUcCrJhWKt9C6Sdc0b7r73\nEzgz++2ctoc672Yz3tW0b3pmvwROv+3IuHL3R0nOsfoQMK3A87aSrGf7/uxyMWb2qyTnnd4xzl09\nmJEjEXuPPKRrJX4kp+2z6f7jc/bt5e4vAr3A75vZazOvOxU4D9jk7rvG1GsRkUO3keQI6CXp/Dbi\nApKP9/+5xf3Jm3cnA505bZ8FDjjnph4gWf3mQjM7OvO6vwvMoPUZpSI6QitV+DOSJaymkyy/cqiW\nkizbda+ZdfHysl1Pk38XriptBh4FrknXKtwF/AH554r1pV//t5ltAF5093X7ed0rSJb42mRmf0ly\n/u6HSI5KFF7PUUSkLHd/0sw+RbJs151m9gWSefvDJEtctfpmDb0kBfYXzexvgGOBPyEpSJsvNu4D\nLjKzK4DvA0+4e/MRWNz9Z2b2pyTLdn3NzLp5edmux4HrK8oiLaYjtDLu3P37lJgI3X0jMJdkzder\nSNaEvQ84y91zLyCoiidrL76L5Kjx5SQT/iMkt5lsdjvJmohzSVZ66D7A634X+C2SJWdGXveHwG+7\n1qAVkRZz9xXAEuAUkuLuj4AbgTk+eg3aVvTlYZK1bR24Brgw7ctncppfRfLJ3TKSOfcTB3jdm4F5\nJCs6rCQ5iPBPwG+61qCNhu17waKIiIiISDgKH6E1s19Ibxn3VHoLuf8ws9Mz+83MrjKzren+jWZ2\nWtNrTExvrfdUeuvP9el5hNk2U8zs1vT2oDvNrMvMji0fVURERERiVOgIrZn9HMndnP6FZCH6J4HT\nSO7g8Wja5k9JPkp9P8k6c1eTXDH+hpFFjc3sr0gWM15Icn7kGuAldz8r815fJjln5kMkd/r4LPBN\ndz+vfFypQ3pS/ysP1Mbdt4X+niIi7cLMTgSOPECTF7I3sQn1PUVGFC1oP01yPuNv7We/AVuAa939\nmnTbZJLbzi1093Xp908C57n7P6ZtXk+ysPwsd7/PzGYADwFnuPsDaZu5JOfLnOzuW8rFlTqY2c0k\nv+Dsl7uXWcewrd5TRKRdmNnjHPimAV9z97NDf0+REUVXOWgAG8zsH4C3Aj8huZ3cTen+U0mWatq7\ndp27P21m9wOzgHVAB8kR12ybzWY2mLa5L/26c6SYTW0kWafzzSQnc0s4VtH6q2XreE8RkXaxgAN/\nSvXTA+wL6T1FgOIF7euAi4DrSG5FdwbJUkUvuPvneHnd0e1Nz9ue2TeN5GOH5isLm9s8kd3p7nvS\nW5Ae8tqm0h7c/SGSI+5Rv6eISLtw928cDu8pMqJoQXsE8IC7/4/0+2+lC99fCHxuXHtWkJn9PPB2\nknXldtfZFxEJxkTgtcAGd3+q5r4ER/OuiJRQybxbtKDdyr5HvQZIFpwHGLnIZmralsz33860OcrM\njm86Sjs18/xtwEnZNzGzCcCUTJtmbwduPbQYIiKjLAD+ru5OBEjzroiUNa7zbtGC9hvsex/6X+Hl\neyE/RlJw/g5pAWtmryI57/Wv0jZ9wM/SNuvTNtNJFnW+N21zL3C8mXW4+8hdmP47yRHi/S0+/zjA\nLbfcwowZMwrGam+XXXYZ118f381MlCssMeYaGBjg/PPPh3T+kMIeB827IVGusMSYq6p5t2hBez3w\nr2b2P4C/B34D+GD6wN3dzG4ArjCzR3h52a4tQE/a5un0tqbXpefE7iK5y9K97n5f2mbAzO4EbjKz\nC0kuIlsDrDvACge7AWbMmMHMmTMLxmpvkydPji4TKFdoYs2V0sfl5WjeDYxyhSXWXKlxnXcLFbTu\n/k0zew/wKZLbzD0GXOru2Y+cVgHHkNyu7nhgEzB3ZA3a1GUkKxasJ7mH/QaSe0dnLSApYr+aaXtJ\nkf7G4vvf/37dXaiEcoUl1lwieWId78oVllhzVaHoEVrc/UvAlw6w30mK3QPdV3k3sDh97K/NDkA3\nUQBefPHFurtQCeUKS6y5RPLEOt6VKyyx5qpC4VvfSutNn9582nIclCssseYSyRPreFeusMSaqwoq\naAMwf/78urtQCeUKS6y5RPLEOt6VKyyx5qpCoVvftjMzmwn09fX1xXwCtYiMo/7+fjo6OgA63L2/\n7v6ERvOuiBRV1byrI7QBGBoaqrsLlVCusMSaSyRPrONducISa64qqKANwKJFi+ruQiWUKyyx5hLJ\nE+t4V66wxJqrCipoA7BixYq6u1AJ5QpLrLlE8sQ63pUrLLHmqoIK2gDEem6acoUl1lwieWId78oV\nllhzVUEFrYiIiIgETQWtiIiIiARNBW0Aurq66u5CJZQrLLHmEskT63hXrrDEmqsKKmgD0N8f5/KY\nyhWWWHOJ5Il1vCtXWGLNVQXdWEFEDlu6scLYaN4VkaJ0YwURERERkRwqaEVEREQkaCpoRURERCRo\nKmgD0Gg06u5CJZQrLLHmEskT63hXrrDEmqsKKmgDsGTJkrq7UAnlCkusuUTyxDrelSssseaqglY5\nEJHDllY5GBvNuyJSlFY5EBERERHJoYJWRERERIKmgjYAPT09dXehEsoVllhzieSJdbwrV1hizVUF\nFbQB6O7urrsLlVCusMSaSyRPrONducISa64q6KIwETls6aKwsdG8KyJF6aIwEREREZEcKmhFRERE\nJGgqaEVEREQkaCpoA9DZ2Vl3FyqhXGGJNZdInljHu3KFJdZcVZhQdwfG28DAQOHnnHDCCZxyyikV\n9GZ8zJkzp+4uVEK5whJrLpE8sY535QpLrLmqUGiVAzNbASxv2vywu78+3W/AlcCfAMcD3wAucvdH\nMq8xEbgWOBc4GtgAfNjdt2faTAFWA+8CXgLWAx9x92cO0LeZQN8hh8mYOHESDz880NZFrYiMP61y\nMDZa5UBEiqpq3i1zhPa7wNsy3+/J/HkZcAnwfuAx4Gpgg5m9wd13p22uB34POAd4GlgD3A6clXmd\nW4FXA7OBVwCfBW4Ezjt4964G3lEgzgC7d5/P0NCQCloRERGRAJUpaPe4+7bmjenR2UuBT7r759Nt\n7wO2A+8G1pnZZOAC4Dx3vytt0wkMmNmZ7n6fmc0A5gJnuPsDaZuLgTvM7GPuvuXA3TsV0JECERER\nkcNFmYvCTjOzLWb2AzO71cxGDmueCkwDNo40dPengfuBWemmDpIjrtk2m4HBTJtZwM6RYja1keTU\ngzeX6G/wNm3aVHcXKqFcYYk1l0ieWMe7coUl1lxVKFrQ3g8sJDmCehFJEXuPmR1HUsxCckQ2a3tm\n3zTgBXffeZA2T2R3uvseYEemzWFl1apVdXehEsoVllhzieSJdbwrV1hizVWFQqccuPuXM99+x8zu\nB34I/BFQfHkBOSTr1q2ruwuVUK6wxJpLJE+s4125whJrriqMaR3a9Ejr94BfBkbOq53a1GxqZt82\n4CgzO/4gbU7K7jSzCcCUTJsD6CQ5kNsBNNLHLKCnqV1vum+0xYsX09XVNWpbf38/jUaDoaGhUduX\nL1/OypUrR20bHByk0WiwefPmUdtXr17N0qVLR20bHh6m0Wjs85FCd3f3qLXnJk2aBMC8efPo6Rmd\no7e3l0YjjBwjRnKM5Ao9R1Zvby/nnntuFDmafx6TJk0KOkd3dzeNRoNGo8GJJ57I5MmTmT179j6v\nI8XNmjWLadOm0dHRsffveNasWUHPV5p32ztHlubd9s3Rynm30LJd+zzZ7FjgRyRLea0GtgDXuPu1\n6f5XkZw+sNDdRy4KexKY7+7r0zbTgc3ArMxFYQ8Bp7t7X9pmDnAncPL+Lgp7edmuW4AFBVL0Ax1o\n2RmRw4+W7RobLdslIkW1xbJdZnYN8EWS0wxeQ7Lm7B6g293dzG4ArjCzR3h52a4tpIdH3f1pM+sC\nrjOzHcAukkL4Xne/L20zYGZ3AjeZ2YUkF5GtAdYdfIUDERERETncFD3l4GSgG3gY+HvgKeBMd38y\n3b+KpEC9EfgmcCwwN7MGLcBlwJdIbpbwdZLTCN7b9D4LSI7afhW4A9gEfLBgX6PR/JFALJQrLLHm\nEskT63hXrrDEmqsKRS8K2/ckldH7HfhE+thfm93A4vSxvzY7OKSbKBweYr3hg3KFJdZcInliHe/K\nFZZYc1VhTOfQthOdQysiRekc2rHRObQiUlRbnEMrIiLSbGCg+KqNJ5xwgo4+ici4UUErIiJjcv75\n5xd+zsSJk3j44QEVtSIyLsa0Dq20RvPacbFQrrDEmkvGw9VAX4HHLezePbzPGprtJNbxrlxhiTVX\nFVTQBmDZsmV1d6ESyhWWWHPJeDgVmFngMaOebhYQ63hXrrDEmqsKKmgDsGbNmrq7UAnlCkusuUTy\nxDrelSssseaqggraAMR6jplyhSXWXCJ5Yh3vyhWWWHNVQQWtiIiIiARNBa2IiIiIBE0FbQBWrlxZ\ndxcqoVxhiTWXSJ5Yx7tyhSXWXFVQQRuA4eHhurtQCeUKS6y5RPLEOt6VKyyx5qqCbn2rW9+KHLZ0\n69ux0bwrIkVVNe/qCK2IiIiIBE0FrYiIiIgETQVtANr59pBjoVxhiTWXSJ5Yx7tyhSXWXFVQQRuA\nRYsW1d2FSihXWGLNJZIn1vGuXGGJNVcVVNAGYMWKFXV3oRLKFZZYc4nkiXW8K1dYYs1VBRW0AYj1\nKmDlCkusuUTyxDrelSssseaqggpaEREREQmaCloRERERCZoK2gB0dXXV3YVKKFdYYs0lkifW8a5c\nYYk1VxVU0Aagvz/OGxgpV1hizSWSJ9bxrlxhiTVXFXTrW92CUeSwpVvfjo3mXREpSre+FRERERHJ\noYJWRERERIKmglZEREREgqaCNgCNRqPuLlRCucISay6RPLGOd+UKS6y5qjCmgtbMPm5mbmY3ZLaZ\nmV1lZlvN7Dkz22hmpzU9b6KZrTWzp8zsGTNbb2ZTm9pMMbNbzWyXme00sy4zO3Ys/Q3VkiVL6u5C\nJZQrLLHmEskT63hXrrDEmqsKpQtaMzsD+BDwnaZdy4BLgAuBNwPPAhvMbGKmzfXAu4BzgLcCrwFu\nb3qdW4E3ArOBdwJvAW4s29+QzZkzp+4uVEK5whJrLpE8sY535QpLrLmqUKqgTY+U3gr8CfDTzHYD\nLgU+6e6fd/fvAO8jKVjfnbaZDFwAfNTd73L3PqAT+G9mdmbaZgYwF/iAu9/v7puAi4Fzzew15aKK\niIiISIzKHqFdC/yzu29s2n4qMA3Yu93dnwbuB2almzqAVzS12QwMZtrMAna6+wOZ194IvERy1FdE\nREREBChR0JrZucBM4PKc3dPSr9ubtm/P7JsGvODuOw/S5onsTnffA+zItDls9PT01N2FSihXWGLN\nJZIn1vGuXGGJNVcVChW0ZvaLwGeABe6+u5ouSbPu7u66u1AJ5QpLrLlE8sQ63pUrLLHmqkLRI7Qd\nwElAv5ntMbM9JBd1XZL+eeTI7NSm500FtqV/3gYcZWbHH6TNSdmdZjYBmJJpsx+dJAdxO4BG+pgF\nNP+W05vuG23x4sV0dXWN2tbf30+j0WBoaGjU9uXLl7Ny5cpR2wYHB2k0GmzevHnU9tWrV7N06dJR\n24aHh2k0GmzatGnU9u7ubjo7O/d+f9tttwEwb968fX5b6+3tzV3Wox1zjBjJMZIr9BxZvb29PPfc\nc1HkaP553HbbbUHn6O7uptFo0Gg0OPHEE5k8eTKzZ8/e53WkDM277ZpjhObdMHJo3i3P3P3QG5sd\nB/xS0+bPApuBlcB3gS3ANe5+bfqcV5GcPrDQ3delF4U9Ccx39/Vpm+npa8xy9/vSi8IeAk5PLxrD\nzOYAdwInu/uWnL7pnuIiUkhV9xQ/XGjeFZGiqpp3JxRp7O7/CTyY3WZmzwJPufuD6fc3AFeY2SPA\nY8DVJEVuT/oaT5tZF3Cdme0AdgGrgXvd/b60zYCZ3QncZGYXklxEtgZYl1fMioiIiMjhq1BBe4hW\nAceQrBl7PLAJmNt0zu1lJCsWrAeOBjYAH256nQUkRexXM20vqaC/IiIiIhKwMd/61t3PdvdLM9+7\nu3/C3ae5+0R3f5u7f6/pObvdfbG7T3H3Y9z9ve6+ranNDnc/z92Pc/fJ7r7I3Z8Za39DlHfeSgyU\nKyyx5hLJE+t4V66wxJqrCmMuaKV6sd4pRLnCEmsukTyxjnflCkusuaqggjYA8+fPr7sLlVCusMSa\nSyRPrONducISa64qqKAVERERkaCpoBURERGRoKmgDUDzAsexUK6wxJpLJE+s4125whJrriqooA3A\nqlWr6u5CJZQrLLHmEskT63hXrrDEmqsKKmgDsG7durq7UAnlCkusuUTyxDrelSssseaqggraAEya\nNKnuLlRCucISay6RPLGOd+UKS6y5qqCCVkRERESCpoJWRERERIKmgjYAS5curbsLlVCusMSaSyRP\nrONducISa64qqKANwCmnnFJ3FyqhXGGJNZdInljHu3KFJdZcVTB3r7sP48LMZgJ9cAuwoMAz+4EO\n+vr6mDlzZjWdE5G21N/fT0dHB0CHu/fX3Z/QaN4VkaKqmnd1hFZEREREgqaCVkRERESCpoI2AJs3\nb667C5VQrrDEmkskT6zjXbnCEmuuKqigDcCyZcvq7kIllCssseYSyRPreFeusMSaqwoqaAOwZs2a\nurtQCeUKS6y5RPLEOt6VKyyx5qqCCtoAxLpsh3KFJdZcInliHe/KFZZYc1VBBa2IiIiIBE0FrYiI\niIgETQVtAFauXFl3FyqhXGGJNZdInljHu3KFJdZcVVBBG4Dh4eG6u1AJ5QpLrLlE8sQ63pUrLLHm\nqoJufatbMIoctnTr27HRvCsiRenWtyIiIiIiOVTQioiIiEjQVNAGYGhoqO4uVEK5whJrLpE8sY53\n5QpLrLmqoII2AIsWLaq7C5VQrrDEmkskT6zjXbnCEmuuKhQqaM3sIjP7jpntSh/3mtnvZvabmV1l\nZlvN7Dkz22hmpzW9xkQzW2tmT5nZM2a23symNrWZYma3pu+x08y6zOzYsUUN14oVK+ruQiWUKyyx\n5hLJE+t4V66wxJqrCkWP0P4Y+DjQAZwO3AV83szemO5fBlwCXAi8GXgW2GBmEzOvcT3wLuAc4K3A\na4Dbm97nVuCNwGzgncBbgBsL9jUasV4FrFxhiTWXSJ5Yx7tyhSXWXFWYUKSxu3+xadP/NLOLgDPN\n7CHgUuCT7v55ADN7H7AdeDewzswmAxcA57n7XWmbTmDAzM509/vMbAYwFzjD3R9I21wM3GFmH3P3\nLaXTioiIiEh0Sp9Da2ZHmtm5wDHAvcCpwDRg40gbd38auB+YlW7qAF7R1GYzMJhpMwvYOVLMpjYC\nL5Ec9RURERER2atwQWtmv2ZmzwDPA38NvMfdHyIpZiE5Ipu1PbNvGvCCu+88SJsnsjvdfQ+wI9Pm\nsNLV1VV3FyqhXGGJNZdInljHu3KFJdZcVShzhPZh4NdJjpb+FfA5M3vDuPZKRunvj/MGRsoVllhz\nieSJdbwrV1hizVWFwgWtu7/g7t939z53vxz4d+AjwLa0ydSmp0zN7NsGHGVmxx+kzUnZnWY2AZiS\naXMAnSQHcjuARvqYBfQ0tetN9422ePHifX4j6u/vp9Fo7LMe3PLly1m5cuWobYODgzQaDTZv3jxq\n++rVq1m6dOmobcPDwzQaDTZt2jRqe3d3N52dnXu/X7t2LQDz5s2jp2d0jt7eXhqNMHKMGMkxkiv0\nHFm9vb386Ec/iiJH889j7dq1Qefo7u6m0WjQaDQ48cQTmTx5MrNnz97ndaQMzbvtmmOE5t0wcmje\nLc/cfWwvYHYXyTmwncAW4Bp3vzbd9yqS0wcWuvvIRWFPAvPdfX3aZjqwGZiVuSjsIeB0d+9L28wB\n7gRO3t9FYbqnuIgUVdU9xQ8XmndFpKiq5t1CqxyY2aeAL5MUsMcB5wFnA293dzezG4ArzOwR4DHg\napIitweSi8TMrAu4zsx2ALuA1cC97n5f2mbAzO4EbjKzC0kuIlsDrNMKByIiIiLSrFBBS3IqwN8C\nrwaeBr5DUsx+Jd2/imTVgxuB44FNwFx33515jctIVixYDxwNbAA+3PQ+C0iK2K9m2l5SsK8iIiIi\nchgodA6tu1/g7q9196Pd/SR3f1ummMUTn3D3ae4+Md3/vabX2O3ui919irsf4+7vdfdtTW12uPt5\n7n6cu09290Xu/szYooYr7xynGChXWGLNJZIn1vGuXGGJNVcVSq9DK62zZMmSurtQCeUKS6y5RPLE\nOt6VKyyx5qqCCtoAzJkzp+4uVEK5whJrLpE8sY535QpLrLmqoIJWRERERIKmglZEREREgqaCNgDN\nixbHQrnCEmsukTyxjnflCkusuaqggjYA3d3ddXehEsoVllhzieSJdbwrV1hizVUFFbQBuO222+ru\nQiWUKyyx5hLJE+t4V66wxJqrCipoRURERCRoKmhFREREJGgqaEVEREQkaCpoA9DZ2Vl3FyqhXGGJ\nNZdInljHu3KFJdZcVVBBG4BY7xSiXGGJNZdInljHu3KFJdZcVVBBG4D58+fX3YVKKFdYYs0lkifW\n8a5cYYk1VxVU0IqIiIhI0FTQioiIiEjQVNAGYNOmTXV3oRLKFZZYc4nkiXW8K1dYYs1VBRW0AVi1\nalXdXaiEcoUl1lwieWId78oVllhzVUEFbQDWrVtXdxcqoVxhiTWXSJ5Yx7tyhSXWXFVQQRuASZMm\n1d2FSihXWGLNJZIn1vGuXGGJNVcVVNCKiIiISNBU0IqIiIhI0FTQBmDp0qV1d6ESyhWWWHOJ5Il1\nvCtXWGLNVQUVtAE45ZRT6u5CJZQrLLHmEskT63hXrrDEmqsK5u5192FcmNlMoA9uARYUeGY/0EFf\nXx8zZ86spnMi0pb6+/vp6OgA6HD3/rr7ExrNuyJSVFXzro7QioiIiEjQVNCKiIiISNBU0AZg8+bN\ndXehEsoVllhzieSJdbwrV1hizVUFFbQBWLZsWd1dqIRyhSXWXCJ5Yh3vyhWWWHNVoVBBa2aXm9k3\nzew/zey82NvWAAAVsUlEQVQJM+sxs+lNbczMrjKzrWb2nJltNLPTmtpMNLO1ZvaUmT1jZuvNbGpT\nmylmdquZ7TKznWbWZWbHlo8arjVr1tTdhUooV1hizSWSJ9bxrlxhiTVXFYoeoX0rsBY4E5gNvALo\nNbNjMm2WAZcAFwJvBp4FNpjZxEyb64F3Aeekr/ka4Pam97oVeGP6Pu8E3gLcWLC/UYh12Q7lCkus\nuUTyxDrelSssseaqwoQijd19bvZ7M1sIPAF0AF83MwMuBT7p7p9P27wP2A68G1hnZpOBC4Dz3P2u\ntE0nMGBmZ7r7fWY2A5gLnOHuD6RtLgbuMLOPufuW0olFREREJCpjPYd2cvp1R/r1VGAasHGkgbs/\nDdwPzEo3dZAc2c222QwMZtrMAnaOFLOpjcBLJEd9RURERESAMRS0ZnYEcAPwDXd/MN08Lf26van5\n9sy+acAL7r7zIG2eyO509z0khfM0DjMrV66suwuVUK6wxJpLJE+s4125whJrrioUOuWgyVrgV4Hf\nHKe+yH4MDw/X3YVKKFdYYs0lkifW8a5cYYk1VxVKHaE1szUkF2r9trv/OLNrW/p1atNTpmb2bQOO\nMrPjD9LmpKb3nABMybTZj06Sg7gdQCN9zAJ6mtr1pvtGW7x4MV1dXaO29ff302g0GBoaGrV9+fLl\n+/z2NDg4SKPR2GftuNWrV7N06dJR24aHh2k0GmzatGnU9u7ubjo7O/d+f+WVVwIwb948enpG5+jt\n7aXRCCPHiJEcI7lCz5HV29vLt771rShyNP88rrzyyqBzdHd302g0aDQanHjiiUyePJnZs2fv8zpS\nhubdds0xQvNuGDk075Zn7n7ojZOLvlYD7wHOdvdHcvZvAa5x92vTba8iOX1gobuPXBT2JDDf3den\nbaYDm4FZmYvCHgJOd/e+tM0c4E7g5LyLwnRPcREpqqp7ih8uNO+KSFFVzbtFTzlYC5wH/D7wn2Y2\ncj7r0+7+nLu7md0AXGFmjwCPAVeTFLk9kFwkZmZdwHVmtgPYRVIk3+vu96VtBszsTuAmM7uQ5CKy\nNcA6rXAgIiIiIllFTzm4iGRlg7uBrZnHvEybVSQF6o3AN4FjgbnuvjvT5jLgS8B64OskpxG8t+m9\nFpActf0qcAewCfhgwf5GofkjhVgoV1hizSWSJ9bxrlxhiTVXFQoVtO5u+3ncnGnj7v4Jd5/m7hPd\n/W3u/r2m19nt7ovdfYq7H+Pu73X3bU1tdrj7ee5+nLtPdvdF7v7MmNIGatGiRXV3oRLKFZZYc4nk\niXW8K1dYYs1VhbGuQystsGLFirq7UAnlCkusuUTyxDrelSssseaqggraAMR60YRyhSXWXCJ5Yh3v\nyhWWWHNVQQWtiIiIiARNBa2IiIiIBE0FbQCaF1SOhXKFJdZcInliHe/KFZZYc1VBBW0A+vvjXO9d\nucISay6RPLGOd+UKS6y5qlDoTmHtTHesEZGidKewsdG8KyJFVTXv6gitiIiIiARNBa2IiIiIBE0F\nrYiIiIgETQVtABqNRt1dqIRyhSXWXCJ5Yh3vyhWWWHNVQQVtAJYsWVJ3FyqhXGGJNZdInljHu3KF\nJdZcVVBBG4A5c+bU3YVKKFdYYs0lkifW8a5cYYk1VxVU0IqIiIhI0FTQioiIiEjQVNAGoKenp+4u\nVEK5whJrLpE8sY535QpLrLmqoII2AN3d3XV3oRLKFZZYc4nkiXW8K1dYYs1VBRW0Abjtttvq7kIl\nlCssseYSyRPreFeusMSaqwoqaEVEREQkaCpoRURERCRoKmhFREREJGgqaAPQ2dlZdxcqoVxhiTWX\nSJ5Yx7tyhSXWXFVQQRuAWO8UolxhiTWXSJ5Yx7tyhSXWXFVQQRuA+fPn192FSihXWGLNJZIn1vGu\nXGGJNVcVVNCKiIiISNBU0IqIiIhI0FTQBmDTpk11d6ESyhWWWHOJ5Il1vCtXWGLNVQUVtAFYtWpV\n3V2ohHKFJdZcInliHe/KFZZYc1WhcEFrZm8xsy+a2RYzczN7d9N+M7OrzGyrmT1nZhvN7LSmNhPN\nbK2ZPWVmz5jZejOb2tRmipndama7zGynmXWZ2bHlYoZt3bp1dXehEsoVllhzieSJdbwrV1hizVWF\nMkdojwH+HVi8n/3LgEuAC4E3A88CG8xsYqbN9cC7gHOAtwKvAW5vep1bgTcCs4F3Am8BbizR3+BN\nmjSp7i5UQrnCEmsukTyxjnflCkusuaowoegT3P3LwJcBzGzUPks2XAp80t0/n257H7AdeDewzswm\nAxcA57n7XWmbTmDAzM509/vMbAYwFzjD3R9I21wM3GFmH3P3LaXSioiIiEh0Che0B3EqMA3YOLLB\n3Z82s/uBWcA6oAN4RVObzWY2mLa5L/26c6SYTW0EXiI56vtP49xvEQnc4OAgQ0NDhZ4zMDBQUW9E\nRKSVxrugnZZ+3d60fXtm3zTgBXffeZA2T2R3uvseM9uRaXPYWLp0KX/xF39RdzfGnXKFpZ1zDQ4O\nMn36DHbvHq67KxKJdh7vY6FcYQkhV9GDCVUdSBjvglYqcMopp9TdhUooV1jaOdfQ0FBazN4CzCjw\nzDuA/1VNpyRo7Tzex0K5wtLuudrpYMJ4L9u1Lf06tWn71My+bcBRZnb8QdqclN1pZhOAKZk2+9FJ\nchC3A2ikj1lAT1O73nTfaIsXL6arq2vUtv7+fhqNxj6/gSxfvpyVK1eO2jY4OEij0WDz5s2jtq9e\nvZqlS5eO2jY8PEyj0dhnnbnu7m46Ozv3fn/xxRcDMG/ePHp6Rufo7e2l0Qgjx4iRHCO5Qs+R1dvb\ny1e+8pUocjT/PC6++OK2zpHYBszMPIaAFemfH07/vAJ4O/DbQHsf+QiH5t12zTFC824YOUKbdz/w\ngQ9kDib0pY81JNfx9wF/lv75LcDxJOsKHLPP64wHc/fyTzZz4D3u3pN+b8AW4Bp3vzbd9iqS0wcW\nuvvIRWFPAvPdfX3aZjqwGZiVuSjsIeB0d+9L28wB7gROzrsozMxmAn3JX+qCAin6gQ76+vqYOXNm\nib8FEalbf38/HR0dJBNokX/HtwLnA3S4e38VfYuZ5l2Rw1u5ubeaebfwKQfpWrC/nNl0qpn9OrDD\n3QfN7AbgCjN7BHgMuJqkyO2BvReJdQHXpefE7gJWA/e6+31pmwEzuxO4ycwuJLmIbA2wTisciIiI\niEhWmVMOTge+lT4Arkv/fFX6/SqSAvVG4JvAscBcd9+deY3LgC8B64Gvk3xO+N6m91lActT2qyQn\num0CPliiv8Fr/pggFsoVllhzieSJdbwrV1hizVWFwgWtu9/t7pbzWJjud3f/hLtPc/eJ7v42d/9e\n02vsdvfF7j7F3Y9x9/e6+7amNjvc/Tx3P87dJ7v7Ind/ZkxpA7Vs2bK6u1AJ5QpLrLlE8sQ63pUr\nLLHmqoJWOUgVXUbihBNOaNnVh2vWrGnJ+7SacoUl1lwieWId78oVllhzVUEFLVuBIzj//PMLPWvi\nxEk8/PBAS4radl+2oyzlCkusuUTyxDrelSssseaqggpadpLcgKzI+pUD7N59PkNDQxpsIiIltfMn\nYyISFhW0e82g2HI/IiJSTvt/MiYiYRnvGytIBZoXSY6FcoUl1lxSh+wnY32H+LiF3buHC91icyxi\nHe/KFZZYc1VBR2gDMDxc/y3lqqBcYYk1l9SpfT8Zi3W8K1dYYs1VBRW0Abjyyivr7kIllCssseaS\nsBQ97xbKnXsb63hXrrDEmqsKKmhFRCQA5c67BZ17K3I4UEErIiIBKLMiDWhVGpHDgwraMWjVkjND\nQ0OccMIJhZ/X7pQrLLHmktC05rzbWMe7coUl1lxV0CoHpbz80VdHR8chP6ZPn8Hg4GDhd1u0aNH4\nR2gDyhWWWHOJ5Il1vCtXWGLNVQUdoS2ltTdjWLFiRbHuBUK5whJrLpE8sY535QpLrLmqoIJ2TFrz\n0dfMme25rM1YKVdYYs0lkifW8a5cYYk1VxV0yoGIiIiIBE1HaEVEJHqtuohXROqhgrbFyiwKvnHj\nRpYtW1ZBb+rV1dXFBRdcUHc3xp1yibSTcuvXTphwFI8++kh0RW2s/46VS1TQtkz5RcGPPHIC5557\nbnQTa39/f5T/UJVLpJ2Uu4h3z544166N9d+xcokK2pYpvyj4iy/GObGuXbu27i5UQrlE2lFrLuJt\nd7H+O1YuUUHbcppURURERMaTCloREZEcZa550MVkIvVQQRuIohPr888/z9FHH134fTQZi4iUv+bh\n6KMnsn79P/LqV7/6kJ+jeVdk7FTQtr2tACUm1iOBFwu/28SJk3j44YGWTK6zZ89m5cqVhZ4TwsTf\naDT4whe+UHc3xl2suUT2Vfaah3t4/vmP8s53vrPQu5UpgqHcgYuPfOQjfOYznyn0HGj/uTfW+SnW\nXFVQQdv2dqZfi0ysdwD/q+BzYCy35y1qcHCQr33tHjo6Ogo9r2zBPTg4yNDQUKHnlJ3AlyxZUvg5\nIWhlrqI/rzIfDYscXNFrHgYoXgiXK4ITZQ5cHFF43oX2P/KseVdU0AajyMQ68p97uQvQWnF6w8DA\nAD/72fMUXUqnTME9ODjI9Okz2L17uFAfyxbPc+bMKdQ+FK3KVfbnJdI+is7XZY4GlzlwUfZgR2uP\nPJcphDXvjl2ZAz/tdDBBBa1klD1vrNzpDYnqV30YGhpKi6PixfM999zDjBlFJv72/2iu3ZX7eY38\nRy0SqjJHg4s+r+zBjtYeeW73o8ExiuFAggpayShz3ljZ3/jrKECKTOLlLwpp5XnIcSvzH7WIVKcV\nR57b/2hwWa089a2ocgcSoJ0OJqiglRyt+o2/nKIfcZT7SKT8jTDKHNlt5aoUZSZVgN7e3sIff+mo\nicjhrlVHgy9r2dHgMvP11q1b+YM/OIfnn3+u0PMmTHgFPT3/VPkFgy//P9m6/8vHW9sXtGa2GFgK\nTAP+HbjY3f+t3l5JPcofNS2v6D/u1q5KUXQyLjupjrj88ssLtS/zn0U7nZMlInUpemDFadXR4LGd\nZlesUN+z59IWXjAYtrYuaM1sHnAdcCFwP3ApsMHMprv7E7V2TmpQ9qhpKz8SaeWqFGO5OroVF6CM\npX8iIkW14mjwWE+zK/MJaCsvGAxXWxe0wEeBm9z9swBmdiHwe8Ai4NN1dkzqFMJHIu16oUaZSXXk\nvSj4vLFevS0i0grtfZpday8YDFfbFrRmdhTQAXxqZJu7v2RmG4FZtXVMpO20+6QVwi8gIiISsiPq\n7sABnEByEsj2pu3bSc6nFRERERFp3yO0JUxMvnyj4NNG2t/BoR8ZatVz9F56L71Xa95rZP6Qgtp8\n3tV76b30Xu37XuM775q7j+frjZv0lINh4A/dvSez/XPA8e7++03tzwNubW0vRSQSC9z97+ruRGg0\n74rIGIzrvNu2R2jd/QUz6wN+B+gBMLMj0u/X5DxlA7AAeBzY3aJuikjYJgKvJZk/pDjNuyJSVCXz\nbtseoYW9y3Z9DvgQ8G8ky3b9EfB6d28+t1ZEREREDkNte4QWwN1vM7MTgatILgT7NjBXxayIiIiI\njGjrI7QiIiIiIgfTzst2iYiIiIgclApaEREREQlaUAWtmS02s8fNbLeZ3W9mv3GQ9mebWb+ZPW9m\n3zezhS3qaiFFcpnZe83sK2b2pJntMrN7zeztrezvoSr688o87ywz22Nm3666j2WUGIdHm9mfmdkP\n07H4uJktalV/D1WJXAvM7N/NbNjMtprZ/zWzn29Vfw/GzN5iZl80sy1m5mb27kN4ThBzRitp3tW8\n2w407+5t39bzLtQ497p7EA9gHvA80Am8AbgR+Clw0n7anwo8C1xLcu/NJcAe4O11ZxljrhuAZcAZ\nwGnAnwMvAG+qO8tYcmWedzzwKMlyHt+uO8d45AI+D9wHvI1kqZJZwFl1ZxnjODwLeBG4JP239pvA\ng8DtdWfJ9PF3gU8C7wEcePdB2gcxZ7T5uAji71Dz7j7P07wbQK4Q5t20n7XMvbUHL/AXdD+wJvP9\nEcBPgI/vp/1K4MGmbeuAO+vOMpZc+3mN7wKfqDvLeORKf0ZXAyvadGItOg7nAjuBKXX3fZxzfQx4\ntGnbxcCP686yn/4eyqQaxJzR5uMiiL9Dzbv7PE/zbhi5gpp30/61bO4N4pQDS+4a1gFsHNnm7i+l\n38/az9NmZdunNhygfcuVzNX8GkcAxwE7quhjGWVzmVkn8Drgyqr7WEbJXA3gAWCZmf3EzL5nZteY\n2Ssr7/AhKpnrXuAXzewdlpgKnENy/8NQtf2c0Uqadw/4Gpp3W0Tz7igxzrswTvNGEAUtcAJwJNC8\n/ux2kvVp80zbT/tXtdGgLpOr2ceAY4G/H8d+jVXhXGZ2GvBp4Hx331Nt90or8/N6HcnHQr9K8vHL\npcAfAn9ZUR/LKJzL3b9Bcoeo20g+et1GckRkcXXdrFwIc0Yrad7dP827raN5NxXpvAvjNG+EUtBK\nDkvuo74c+CN3f6Lu/pRlZkcCfwcsd/fv1d2fcXYEyUcuC9z939z9DuCjwPvb6D/4wszsDcBnSG56\n0kHyEd9rgb+usVsildO8GwTNu4ehtr5TWMYQyYnQU5u2TyX5DSXPtv203+Xuz41v90orkwsAMzsX\n+D/AOe7efKi+bkVzHQecDrzJzNak244AzMz2AHPc/a6qOltAmZ/XVuAn7v50ZtsAYMDJwCPj3ckS\nyuS6HPhXd/+L9PvvmNmzwD1mdoW7b62mq5UKYc5oJc27TTTv1kLz7stinHdhnOaNII7QuvsLQB/w\nOyPb0nOYfofknJI892bbp2YfoH3LlcyFmc0HPgvMd/d/rrqfRZXItQv4NeDXM4+/Bh5O/3x/xV0+\nJCV/Xt8AXmNmx2a2/QrwEvDjirpaSMlck0iuQs16ceTp493HFmn7OaOVNO+Opnm3Hpp3R4lx3oXx\nmjfqvgKuwJVy84DdwPtJlnX4G5LlLaam+z8F/G2m/cgyEKuA1wMfpn2XjymS6zzgZ2meaZnH5Lqz\njCVXzvNX0J5X2xb9eR0L/Aj4B5JlWd4CfA+4qe4sY8y1MB2HF5Gcr3YW8E3g/rqzNP3dj/xH7cBl\n6Z9P2U+mIOaMNh8XQfwdat7d7/M177Z3rrafdzN//y2fe2sPXvAvaQnwQ5J12+4H3pzZdzNwd1P7\ns4Fvpe0fBRbWnWGsuYC70wHS/Li57hxj/Xk1PbctJ9aS4/D1wFeA4XSSvRZ4Zd05xiHXxSRLFw0D\nW4BbgF+oO0emf2cf6N9KyHNGm4+LIP4ONe/mPlfzbvvnaut5N+1jLXOvpS8kIiIiIhKkIM6hFRER\nERHZHxW0IiIiIhI0FbQiIiIiEjQVtCIiIiISNBW0IiIiIhI0FbQiIiIiEjQVtCIiIiISNBW0IiIi\nIhI0FbQiIiIiEjQVtCIiIiISNBW0IiIiIhI0FbQiIiIiErT/DwNbthSr2LnsAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f8f427a3320>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# distribution of on_ratio in N/T sample groups\n",
    "_ = raw_pval_df[['N_on_ratio', 'T_on_ratio']].hist(\n",
    "    bins=np.arange(0, 1.01, 0.05), figsize=(8, 3), sharex=True, sharey=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remarkably similar overall."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pval_df = raw_pval_df.reset_index().rename(\n",
    "    columns={'level_2': 'mkid', 'level_3': 'mclv', 'level_4':'strand', 'level_5': 'aclv'}\n",
    ").drop('level_6', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mkid</th>\n",
       "      <th>mclv</th>\n",
       "      <th>strand</th>\n",
       "      <th>aclv</th>\n",
       "      <th>N_on</th>\n",
       "      <th>T_on</th>\n",
       "      <th>N_off</th>\n",
       "      <th>T_off</th>\n",
       "      <th>fisher_exact_p</th>\n",
       "      <th>N_base</th>\n",
       "      <th>T_base</th>\n",
       "      <th>N_on_ratio</th>\n",
       "      <th>T_on_ratio</th>\n",
       "      <th>N2T_ratio_diff</th>\n",
       "      <th>diff_is_significant</th>\n",
       "      <th>N2T_ratio_change</th>\n",
       "      <th>N2T_ratio_change_sig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>chr9|ABL1|+|133589972</td>\n",
       "      <td>133589972</td>\n",
       "      <td>+</td>\n",
       "      <td>133763062</td>\n",
       "      <td>1.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>382.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>19</td>\n",
       "      <td>405</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.05679</td>\n",
       "      <td>0.004159</td>\n",
       "      <td>False</td>\n",
       "      <td>up</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>chr9|ABL1|+|133763062</td>\n",
       "      <td>133763062</td>\n",
       "      <td>+</td>\n",
       "      <td>133763062</td>\n",
       "      <td>19.0</td>\n",
       "      <td>405.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>19</td>\n",
       "      <td>405</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>down</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  disease gene_name                   mkid       mclv strand       aclv  N_on   T_on  N_off  T_off  fisher_exact_p  N_base  T_base  N_on_ratio  T_on_ratio  N2T_ratio_diff  diff_is_significant N2T_ratio_change N2T_ratio_change_sig\n",
       "0  BLCA    ABL1      chr9|ABL1|+|133589972  133589972  +      133763062  1.0   23.0   18.0   382.0  1.0             19      405     0.052632    0.05679     0.004159        False                up               None               \n",
       "1  BLCA    ABL1      chr9|ABL1|+|133763062  133763062  +      133763062  19.0  405.0  0.0    0.0    1.0             19      405     1.000000    1.00000     0.000000        False                down             None               "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pval_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for col in ['N_on', 'N_off', 'T_on', 'T_off']:\n",
    "    pval_df[col] = pval_df[col].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4.5 s, sys: 66.7 ms, total: 4.57 s\n",
      "Wall time: 4.49 s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "# aca (alternative cleavage) is an old name used to mean APA\n",
    "aca_df = pval_df.groupby(['disease', 'gene_name'])\\\n",
    "    .apply(has_diff_aca)\\\n",
    "    .to_frame(name='has_diff_aca')\\\n",
    "    .reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "aca_df = pval_df.merge(aca_df, on=['disease', 'gene_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mkid</th>\n",
       "      <th>mclv</th>\n",
       "      <th>strand</th>\n",
       "      <th>aclv</th>\n",
       "      <th>N_on</th>\n",
       "      <th>T_on</th>\n",
       "      <th>N_off</th>\n",
       "      <th>T_off</th>\n",
       "      <th>fisher_exact_p</th>\n",
       "      <th>N_base</th>\n",
       "      <th>T_base</th>\n",
       "      <th>N_on_ratio</th>\n",
       "      <th>T_on_ratio</th>\n",
       "      <th>N2T_ratio_diff</th>\n",
       "      <th>diff_is_significant</th>\n",
       "      <th>N2T_ratio_change</th>\n",
       "      <th>N2T_ratio_change_sig</th>\n",
       "      <th>has_diff_aca</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>chr9|ABL1|+|133589972</td>\n",
       "      <td>133589972</td>\n",
       "      <td>+</td>\n",
       "      <td>133763062</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>18</td>\n",
       "      <td>382</td>\n",
       "      <td>1.0</td>\n",
       "      <td>19</td>\n",
       "      <td>405</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.05679</td>\n",
       "      <td>0.004159</td>\n",
       "      <td>False</td>\n",
       "      <td>up</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  disease gene_name                   mkid       mclv strand       aclv  N_on  T_on  N_off  T_off  fisher_exact_p  N_base  T_base  N_on_ratio  T_on_ratio  N2T_ratio_diff  diff_is_significant N2T_ratio_change N2T_ratio_change_sig  has_diff_aca\n",
       "0  BLCA    ABL1      chr9|ABL1|+|133589972  133589972  +      133763062  1     23    18     382    1.0             19      405     0.052632    0.05679     0.004159        False                up               None                 False       "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aca_df.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33 77\n"
     ]
    }
   ],
   "source": [
    "num_reported_genes = aca_df.query('has_diff_aca == True').gene_name.unique().shape[0]\n",
    "num_reported_cases = aca_df.query('has_diff_aca == True')[['gene_name', 'disease']].drop_duplicates().shape[0]\n",
    "print(num_reported_genes, num_reported_cases)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Estimate FDR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Obtain gene-cancer-pair-level (gc-level) p value upper bound by multiplying the lowest two p-values correspond to a significant increase and a significant decrease for each pairs.\n",
    "1. for those `has_diff_aca` set to false, their gc-level p-value set to 0, since they are not reported anyway. But we need them to do Benjamini Hochberg adjustment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def calc_pc_level_p(grp):\n",
    "    assert len(set(grp.has_diff_aca.values)) == 1 # make sure there is only one possibility\n",
    "    if grp.has_diff_aca.all():\n",
    "        sig_inc_p = grp.query('N2T_ratio_diff > 0').fisher_exact_p.min()\n",
    "        sig_dec_p = grp.query('N2T_ratio_diff < 0').fisher_exact_p.min()\n",
    "        return sig_inc_p * sig_dec_p\n",
    "    else:\n",
    "        return 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gc_level_pvals = aca_df.groupby(['gene_name', 'disease']).apply(calc_pc_level_p).sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# sanity check\n",
    "assert gc_level_pvals[gc_level_pvals < 1].shape[0] == num_reported_cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# num_genes * num_diseases tested, actually it should've been three cases less \n",
    "# because the three gene-cancer type pairs have not CSs detected in their normal\n",
    "# sample groups, thus cannot do tumor-specific APA analysis\n",
    "# but it makes little difference in terms of the FDR estimation\n",
    "total_num_tests = 114 * 14"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gc_level_pvals_df = gc_level_pvals.to_frame(name='gc_level_pval')\n",
    "gc_level_pvals_df['rank'] = np.arange(1, gc_level_pvals_df.shape[0] + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8.324790830353091e-05"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc_level_pvals[gc_level_pvals < 1].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# q_star as defined in the BH paper, https://www.jstor.org/stable/2346101\n",
    "q_star = (gc_level_pvals_df.gc_level_pval * total_num_tests / gc_level_pvals_df['rank'])[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# cap at 1 as probability cannot exceed one\n",
    "q_star = np.minimum(np.ones_like(q_star), q_star)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "q_star = np.minimum.accumulate(q_star)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gc_level_pvals_df['fdr_upper_bound'] = q_star"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    7.700000e+01\n",
       "mean     9.578885e-05\n",
       "std      2.696823e-04\n",
       "min      6.451290e-36\n",
       "25%      3.559475e-10\n",
       "50%      2.450215e-06\n",
       "75%      5.104204e-05\n",
       "max      1.725502e-03\n",
       "Name: fdr_upper_bound, dtype: float64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc_level_pvals_df.query('gc_level_pval < 1').fdr_upper_bound.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The estimated FDR is ~0.0017, less than 0.002, so we obtain an upper bound FDR of 0.002, sufficiently small."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Back to characterizing the identified gene-cancer type pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# number of significantly changed CSs within a gene-cancer type pair\n",
    "num_sig_change_clv = aca_df\\\n",
    "    .query('has_diff_aca == True')\\\n",
    "    .groupby(['disease', 'gene_name'])\\\n",
    "    .apply(lambda g: g.query('diff_is_significant').shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(77,)"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_sig_change_clv.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAEXCAYAAACjyo8UAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAAPYQAAD2EBqD+naQAAHYVJREFUeJzt3X+UXHWZ5/H3g7iTExBQAwGiWRlhQkYcOR0IZI6CHseA\nmeEO7lkNUZYhYVU4CUjWTRaUIR1gRhNUokmPI55eZPjRyc6e2QArTDJRHA2TRegSWDSNoNFeJARK\nncjQtAp59o97I9XVndSt2131rf7ez+ucezr1vbduP09X8umb+9PcHRERidchoQsQEZHWUtCLiERO\nQS8iEjkFvYhI5BT0IiKRU9CLiEROQS8iEjkFvYhI5BT0IiKRU9CLiEQueNCbWbeZed00UDPfzOw6\nM9ttZi+Z2TYzOylkzSIik0nwoM98HziuZnpnzbyVwBXApcAZwIvAFjOb0u4iRUQmo0NDF5B52d2f\nrR80MwOuBG5w97uysYuAPcD5wMa2VikiMgl1yhb9SWb2jJn92MzuMLOZ2fgJwLHAtv0Luvte4EFg\nXoA6RUQmnU4I+geBi4FzgctIw/07ZvY60pCHdAu+1p6aeSIichDBd924+301Lx8zsweBnwIfAnYW\nWaeZvRE4B/gJMDzeGkVEOsAU4C3AFnf/eTNvDB709dz9X83sh8CJwP3Z8HRgd81i04FHDrKac4A7\nWlOhiEhQHwHubOYNHRf0ZnY4acjfBuwCngXeSxbsZnYE6dk3Xz7Ian4CcPvttzN79uxWlttxli9f\nzk033RS6jLZT3+VSxr537tzJhRdeCFm+NSN40JvZ54B7SHfXHA+sBl4G+tzdzWwdcI2ZPUka/NcD\nzwCbD7LaYYDZs2fT1dXVyvI7zpFHHlm6nkF9l01Z+840vTs6eNADbwL6gDcCzwPbgTPd/fls/lrg\nMOBm4Khs/rnurn3vY3jqqadClxCE+i6XsvZdVPCgd/cLGsx34NpskgZeeeWV0CUEob7Lpax9F9UJ\np1fKBJo1a1boEoJQ3+VS1r6LUtBHZtGiRaFLCEJ9l0tZ+y7K0j0jcTGzLqC/v7+/zAdsRCQilUqF\nOXPmAMxx90oz79UWfWSq1WroEoJQ3+VS1r6LUtBHZsmSJaFLCEJ9l0tZ+y5KQR+Z7u7u0CUEob7L\npax9F6Wgj0xZj0mo73Ipa99FKehFRCKnoBcRiZyCPjK9vb2hSwhCfZdLWfsuSkEfmUqlqdNro6G+\ny6WsfRelC6ZERCYBXTAlIiIHFPzulTKxBgcHW37V4LRp05g5c2bjBUWkIyjoIzI4OMisWbMZHh5q\n6feZMmUqTzyxU2EvMkko6CNSrVazkL8daNUjFHcyPHwh1Wq1o4I+SRLuvvvu0GW0nfqWPBT0UZoN\nlOsg9LJly0KXEIT6ljx0MFaiMH/+/NAlBKG+JQ8FvYhI5BT0IiKRU9BLFDZv3hy6hCDUt+ShoJco\n9PX1hS4hCPUteSjoJQqbNm0KXUIQ6lvyUNCLiEROQS8iEjkFvYhI5BT0EoXFixeHLiEI9S15KOgl\nCmW9UlJ9Sx4KeonCokWLQpcQhPqWPBT0IiKRU9CLiEROQS9R2L59e+gSglDfkoeCXqKwdu3a0CUE\nob4lDwW9RGHjxo2hSwhCfUseCnqJwtSpU0OXEIT6ljwU9CIikeu4oDezq8zMzWxdzZiZ2XVmttvM\nXjKzbWZ2Usg6RUQmi44KejM7Hfg48FjdrJXAFcClwBnAi8AWM5vS3gqlU61YsSJ0CUGob8mjY4Le\nzA4H7gA+CvyyZtyAK4Eb3P0ud38MuAg4Hjg/RK3SeWbOnBm6hCDUt+TRMUEP9ABfd/dtdeMnAMcC\nvxt3973Ag8C89pUnnezyyy8PXUIQ6lvyODR0AQBmdgHQBZw+xuxjs6976sb31MwTEZEDCL5Fb2Zv\nBr4IfMTdhydy3QsWLCBJkhHTvHnzRj1YeOvWrSRJMur9S5cupbe3d8RYpVIhSRKq1eqI8VWrVrFm\nzZoRY4ODgyRJwsDAwIjx9evXj9rHODQ0RJIko6746+vrG/OWrAsXLhzVx44dO0Ytl3UC9NaNVYAE\nqNaNrwLW1I0NZsu2p49YPg/1oT6K9tHX10eSJMyYMYO5c+eSJAnLly8fVXtu7h50It3P7sDLNZMD\n+7I/vzV7fWrd+/4Z+OIB1tkFeH9/v5dJf3+/pz+rfgdv0ZR+j0772e7cuTN0CUGo7/J49d83Xd5k\nzgbfoge+AbwdOLVmepj0wOypwI+BZ4H37n+DmR1BevbNgTZhpWRWrlwZuoQg1LfkEXwfvbu/ADxe\nO2ZmLwI/d/fHs9frgGvM7ElgF3A98AywGRFgw4YNoUsIQn1LHsGDPqe1wGHAzcBRwHbgXJ/gffoy\neZX1dDv1LXl0ZNC7+7vrXjtwbTaJiEgTOmEfvYiItJCCXqJQf8pbWahvyUNBL1EYGhoKXUIQ6lvy\nUNBLFFavXh26hCDUt+ShoBcRiZyCXkQkcgp6iUL9PUnKQn1LHgp6icKSJUtClxCE+pY8FPQShe7u\n7tAlBKG+JQ8FvUShq6srdAlBqG/JQ0EvIhI5Bb2ISOQU9BKF+icElYX6ljwU9BKFSqUSuoQg1Lfk\noaCXKPT09IQuIQj1LXko6EVEIqegFxGJnIJeRCRyCnqJQpIkoUsIQn1LHgp6icKyZctClxCE+pY8\nFPQShfnz54cuIQj1LXko6EVEIqegFxGJnIJeorB58+bQJQShviUPBb1Eoa+vL3QJQahvyUNBL1HY\ntGlT6BKCUN+Sh4JeRCRyCnoRkcgp6EVEIqeglygsXrw4dAlBqG/JQ0EvUSjrlZLqW/IoFPRm9p/M\nbMpEFyNS1KJFi0KXEIT6ljyKbtHfBDxrZl8xs7kTWZCIiEysokF/PPBR4E3AA2b2uJl90syOnrjS\nRERkIhQKenf/jbv/vbv/KTATuA24BHjazP7BzP7UzGwiCxU5mO3bt4cuIQj1LXmM+2Csu+8GtgH3\nAw6cBvQBT5rZuxq938wuM7PHzOxX2bTDzN5fM9/M7Doz221mL5nZNjM7abx1S1zWrl0buoQg1Lfk\nUTjozWyamV1pZo8CDwDHAOcD/x6YAWwG/i7Hqp4GrgLmkP6S+CZwl5m9LZu/ErgCuBQ4A3gR2KKD\nwVJr48aNoUsIQn1LHkXPuvlfwM9Iw/c24M3u/kF3/0dPvQCsJQ39g3L3e9z9Xnd/0t1/6O6fBv4N\nODPb/XMlcIO73+XujwEXkR4jOL9I7RKnqVOnhi4hCPUteRxa8H2/Av7E3b9zkGWeB5raxWJmrwE+\nCBwG7ABOAI4l3TUEgLvvNbMHgXmAfq2LiDRQKOjd/S9yLOPAj/Ksz8zeThrsU0i35j/g7j8wsz/O\nFtlT95Y9pL8ARESkgaK7bm4ys6VjjC81s88XWOUTwKmk++C/DNxqZn9YpDYppxUrVoQuIQj1LXkU\nPRj7QeBfxhj/P8DCZleWna75lLv3u/vVwKPAJ4Bns0Wm171les28A1qwYAFJkoyY5s2bN+rpNFu3\nbiVJklHvX7p0Kb29vSPGKpUKSZJQrVZHjK9atYo1a9aMGBscHCRJEgYGBkaMr1+/ftRf1KGhIZIk\nGXXaWF9f35j39Vi4cOGoPnbs2DFquawToLdurAIkQLVufBWwpm5sMFu2PX0U+Tze8IY3jOyiAz6P\ndvy9mjlzZhR91MrTx/6+J3sf+9X30dfXR5IkzJgxg7lz55IkCcuXLx9Ve16W7mFp8k1mw8Db3P1H\ndeMnAo+7+7jOiDGzb5Kmy2LgGeBz7v75bN4RwHPAxe4+5j56M+sC+vv7++nq6hpPKZNKpVJhzpw5\nQD/Qqr4rwBzK9rMVCe3Vf9/McfdKM+8tejD2R8A5wN/UjZ8D7GpmRWb2GeA+0mB/HfBh4N3AOe7u\nZrYOuMbMnszWfT1p+OuhkSIiORQN+nXAOjN7I+l57wDvJT3n/b82ua5jSM+3Pw7YCzxGGvL/lM1f\nS3oWzs3AUcB24Fx3Hy5Yu4hIqRQ96+ar2QVLnwJWZ8NPA1e4+39vcl2XNJjvwLXZJDKmgYEBTj75\n5NBltJ36ljwKXxnr7uvd/TjSq2Df4O4zmw15kYmycuXK0CUEob4lj6K7bn4nu9eNSFAbNmwIXUIQ\n6lvyKHoe/dFmdouZDZrZsJn9pnaa6CJFGqk9zbBM1LfkUXSL/mvAW4Ebgd2kd60UEZEOVDTozwLO\ncvfvTWQxIiIy8YoejH0abcVLB6m/WrEs1LfkUTTolwOfMbM3TWQxIkUNDQ2FLiEI9S15FN11cxvp\nVaw/NbNfAb+tnenux4y3MJFmrF69uvFCEVLfkkfRoL9qQqsQEZGWKXplbP2tEEVEpEON55mxbzGz\nbjO7zcyOycbmm9nsiStPJJ/628mWhfqWPIpeMPUu4PvA2cCHgMOzWXOA6yamNJH8lixZErqEINS3\n5FF0i34N0O3u7wFqr4T9BnDmuKsSaVJ3d3foEoJQ35JH0aD/I+B/jjH+HHB08XJEiinrQ1DUt+RR\nNOj3MvbDud8B/Kx4OSIiMtGKBv0m4LNmdjTZFbJmdgbweeD2CapNREQmQNGgvxr4Mekj/Q4HfkD6\nsPCHSB/1J9JW9Q93Lgv1LXkUCnp3/7W7Lwb+ADgfWEL6sPBF7v7yRBYokkel0tSzkqOhviWPcT14\nxN130eTDwEVaoaenJ3QJQahvyaNQ0JvZzQeb7+4fK1aOiIhMtKJb9MfVvX4t8DbSG519e1wViYjI\nhCp6r5vz6sfM7FDgb0kPzIqISIcofK+betlB2BuBFRO1TpG8kiQJXUIQ6lvymLCgz5xAuhtHpK2W\nLVsWuoQg1LfkUfRg7Nr6IdL99gm6YEoCmD9/fugSglDfkkfRg7Hz6l7vA54nfSDJV8dVkYiITKii\nB2PfNdGFiIhIa0z0PnqRIDZv3hy6hCDUt+RR9MEjD5nZd/NME12wyFj6+vpClxCE+pY8iu6jvx/4\nOPBDYEc2diYwC/gK8OvxlyaS36ZNm0KXEIT6ljyKBv1RQI+7f6p20Mz+Cpju7v953JWJiMiEKLqP\n/kPALWOMfw34YOFqRERkwhUN+l8z9rNhz0S7bUREOkrRoP8S8BUz+4KZXZBNNwFfBr44ceWJ5LN4\n8eLQJQShviWPoufR/5WZ7QI+AezfH78T+Ji73zlRxYnkVdYrJdW35FH4PHp3v9Pdz3D3I7LpjCIh\nb2ZXZ6drvmBmz5nZZjObVbeMmdl1ZrbbzF4ys21mdlLR2iU+ixYtCl1CEOpb8igc9GZ2hJldnAXw\n67Oxd5hZ/b3qGzkb6CHdv/8+0puibTWzw2qWWQlcAVwKnAG8CGwxsylF6xcRKYuiNzU7BdgGDAFv\nJj3b5pfAQmAG8Bd51+Xu59at+2LgOWAO8G0zM+BK4AZ3vytb5iJgD+nzajcW6UFEpCyKbtHfBNwJ\nvBUYrhn/OnDWOGs6Mvv6i+zrCcCxpL9YAHD3vcCDjL65mpTU9u3bQ5cQhPqWPIoG/enA37i7143/\njNGPGczNzA4B1gEPuPvj2fCx2dc9dYvvqZknJbd2bf2ds8tBfUseRa+M/S1w+BjjJwLV4uXQA5wC\nvHMc65AS2rixnHvw1LfkUXSL/h7gL7PnxAK4mc0APgv8Q5EVmtkG4M+A97j70zWzns2+Tq97y/Sa\neWNasGABSZKMmObNmzfqzndbt24d89FkS5cupbe3d8RYpVIhSRKq1ZG/z1atWsWaNWtGjA0ODpIk\nCQMDAyPG169fz4oVI5+4ODQ0RJIko/5L2tfXN+Y5wwsXLhzVx44dO0Ytl3UC9NaNVUifE1P/e3kV\nsKZubDBbtj19FPk8hoaGRnbRAZ9HO/5eTZ06NYo+auXpY3/fk72P/er76OvrI0kSZsyYwdy5c0mS\nhOXLl4+qPTd3b3oCXk96Y7Mq8DKwi/SK2O3A4U2uy4ANpLt9TjrA/N3AJ2vGjiA9NnDBAdbZBXh/\nf7+XSX9/vwMO/Q7eoin9HmX72YqE9uq/b7q8ycwuesHUL4H3mNnZwDtId+NUgC3uo/bbN9IDfBj4\nc+AFM9u/332vu7/k7m5m64BrzOxJ0l8q1wPPALoptYhIA03vujGz15rZFjM7yd3/2d2/5O5/7e7/\nWCDkAS4jPdPmW6Rb7vunhTXLrAXWAzcDD5H+YjnX3YcRgVH/RS4L9S15NL1F7+6/NbM5pP+FGDd3\ntxzLOHBtNkkJDA4Ojto/ejBmRqVSaep7TJs2jZkzZzZbWkeZ7PUXVda+iyp61s0dwGLg0xNYiwiQ\nhvysWbMZHh5qvHCNG2+8sanlp0yZyhNP7JzUoXH55ZeHLiGIsvZdVNGgd2CZmf0J8DDpLQlenem+\ncryFSXlVq9Us5G8HZrfou+xkePhCqtXqpA56kTyKBv0c4LHsz39UN29CdumIpCHfFboIkUmvqaA3\ns98Hdrn7u1pUj0hBA8DJoYtou4GBAU4+WX3LwTV71s2TwNH7X5jZJjOrv5BJJIBy7i1cuVJ9S2PN\nBn39GTILgMPGWlCkvTaELiCIDRvUtzRW+H70Ip2lnAdUy3oguax9F9Vs0O+/BLd+TEREOlSzZ90Y\n8DUz+3X2egrwt2ZWf3rlf5iI4kREZPya3aK/lfTpT3uz6XbSe87srZtE2qz+jpvlUH8nxbIoa99F\nNbVF7+6j7wcq0hGau4o2FvW3Zy6LsvZdlA7GSiRWhy4giNWr1bc0pqAXEYmcgl5EJHIKeonEeB5V\nPHk1cyvnmJS176IU9BKJJaELCGLJEvUtjSnoJRLdoQsIoru7O3QJQZS176IU9BKJct7OuKtLfUtj\nCnoRkcgp6EVEIqegl0j0hi4giN5e9S2NKeglEpXQBQRRqahvaUxBL5HoCV1AED096lsaU9CLiERO\nQS8iEjkFvYhI5BT0EokkdAFBJIn6lsYU9BKJZaELCGLZMvUtjSnoJRLzQxcQxPz56lsaU9CLiERO\nQS8iEjkFvURic+gCgti8WX1LYwp6iURf6AKC6OtT39KYgl4isSl0AUFs2qS+pTEFvYhI5BT0IiKR\n64igN7OzzOweM3vGzNzMzq+bb2Z2nZntNrOXzGybmZ0Uql4RkcmkI4IeOAx4FFh6gPkrgSuAS4Ez\ngBeBLWY2pT3lSedbHLqAIBYvVt/S2KGhCwBw9/uA+wDMbMQ8SweuBG5w97uysYuAPcD5wMa2Fisd\nqpxXSpb1CtGy9l1Up2zRH8wJwLHAtv0D7r4XeBCYF6oo6TSLQhcQxKJF6lsamwxBf2z2dU/d+J6a\neSIicgCTIehFRGQcJkPQP5t9nV43Pr1m3pgWLFhAkiQjpnnz5o26fHrr1q1j3t966dKlo542X6lU\nSJKEarU6YnzVqlWsWbNmxNjg4CBJkjAwMDBifP369axYsWLE2NDQEEmSsH379hHjfX19Yx54Wrhw\n4ag+duzYMWq5rBOgt26sQnoP92rd+CpgTd3YYLZsJ/fxvzuuj3b8vaqtbzL3UStPH/u/TvY+9qvv\no6+vjyRJmDFjBnPnziVJEpYvXz6q9tzcvaMmwIHza14bsBv4ZM3YEcAwcMEB1tEFeH9/v5dJf3+/\npz+/fgdv0ZR+j1b+bIv1cV7H9dEO5513XugSgihj36/+u6DLm8zVjjjrxswOB06sGTrBzE4FfuHu\ng2a2DrjGzJ4EdgHXA89Q1jtZyRjKefLVxo3qWxrriKAHTgPur3n9hezrrcDFwFrSc+1vBo4CtgPn\nuvtwG2uUjjY1dAFBTJ2qvqWxjgh6d/8W6S6aA8134NpsEhGRJkyGg7EiIjIOCnqJxIrGi0So/qyO\nsihr30Up6CUSM0MXEMTMmepbGlPQSyQuD11AEJdfrr6lMQW9iEjkFPQiIpFT0EskBhovEqH6y+7L\noqx9F6Wgl0isDF1AECtXqm9pTEEvkdgQuoAgNmxQ39KYgl4iUc7T7cp6mmFZ+y5KQS8iEjkFvYhI\n5BT0Eon6h4yUQ/1DMMqirH0XpaCXSAyFLiCIoSH1LY0p6CUSq0MXEMTq1epbGlPQi4hETkEvIhK5\njnjClMj4VYFpoYtou0ceeYR9+/a19HtMmzat485br1arTJtWvs+7KAW9RGIJcHfoItpqcHCQOXNO\nY9++V1r6faZMmcoTT+zsqLBfsmQJd99drs97PBT0Eonu0AW0XbVazUL+dmB2i77LToaHL6RarXZU\n0Hd3d4cuYVJR0EskukIXENBsytZ/V1e5+h0vHYwVEYmctuhFRCbA4OAg1Wq1ZevfuXNn4fcq6CUS\nvcAloYuQNunt7eWSSzrn8x4cHGTWrNkMD3fmFbsKeolEBQV9eVQqlY4K+mq1moV8Kw+M3wv8ZaF3\nKuglEj2hC5A26unp1M+7lQfGtetGRCaxVu/fhs688KtdFPQiElS79m934oVf7aKgF5Gg2rN/uzMv\n/GoXBb1EIqFst0CITzP7t/V5N0MXTEkkloUuQNpKn3czFPQSifmhC5C20ufdDAW9iEjkFPQiIpFT\n0EskNocuQNpKn3czFPQSiTWhC5C20ufdjEkV9Ga21Mx+YmbDZvagmc0NXZN0iqNDFyBtpc+7GZMm\n6M1sIfAFYDXpybaPAlvM7JighYmIdLhJE/TAfwG+6u63uPsPgEuBIdKHhYqIyAFMiqA3s38HzAG2\n7R9z933Z63mh6hIRmQwmyy0QpgGvAfbUje8BTh5j+SkwvieyTEav9nsv47ml6cHtSr/Dvfe27Oe7\na9eu7E/N9PEAcEcz3yX9Di3s45BDDmHfvn0tWTcU/Tk1/V3S76DP++Brb8tn8cD+P0xp9p3m7hNb\nSwuY2fHAz4A/dvcdNeNrgbPd/Yy65T9Mc38LREQmi4+4+53NvGGybNFXgVeA6XXj04Fnx1h+C/AR\n4CfAcEsrExFpjynAW0jzrSmTYosewMweBL7r7pdnrw8BBoEN7v7ZoMWJiHSwybJFD+mplbea2cPA\nd4ErgcOAW4JWJSLS4SZN0Lv7JjM7GrgOOBZ4BDjX3esP0IqISI1Js+tGRESKmRTn0YuISHFRBb2Z\nnWVm95jZM2bmZnZ+6JrawcyuNrOHzOwFM3vOzDab2azQdbWamV1mZo+Z2a+yaYeZvT90Xe1kZldl\nf9fXha6l1cysO+u1dhoIXVc7mNkMM7vdzH5uZi+Z2f81s9Pyvj+qoCc9OPsosDR0IW12NtADnAm8\nD3gtsNXMDgtaVes9DVxFetX0acA3gbvM7G1Bq2oTMzsd+DjwWOha2uj7wHE10zvDltN6ZvZ60qul\nfgu8H/hD4JPAL/OuY9IcjM3D3e8D7gMws8DVtI+7n1v72swuBp4jDcBvh6ipHdz9nrqhT5vZZaS/\n8L4foKS2MbPDSS8K/ChwTeBy2ulldx/r2pmY/Tfg/7n74pqxXQdaeCyxbdFL6sjs6y+CVtFGZvYa\nM7uA9H91OxotH4Ee4Ovuvq3hknE5Kds1+2Mzu8PMZoYuqA0S4GEz+/ts1+z3zOyjzawgqi16+d2F\nZOuAB9z98dD1tJqZvZ002KcA/wZ8ILu7abSyX2hdwOmha2mzB4GLgSdId9usAr5jZqe4+wshC2ux\n3wcuI72W6K9JP/cvmdlv3P3WPCtQ0MenBziFEuy7zDwBnEr6v5j/SHpR3dmxhr2ZvRn4IvA+dy/V\n7T2yXbP7PZZdLf9T4ENAb5iq2uIQ4GF3/1T2+ntmdgrprdpzBb123UTEzDYAfwa8x92fDl1PO7j7\nb9z9KXfvd/erSQ/GfyJ0XS00BzgGqJjZy2b2MunB+Cuy168JW177uPu/Aj8ETgxdS4vtBuo3XHYC\nuXdbaYs+ApYeeV4PfAB4t7s3daAmMocAvxe6iBb6BvD2urFbgAFgjbu/0v6SwsgOSJ8I3Ba6lhZ7\nAKg/XfoPSP83k0tUQV/zwe93gpmdCvzC3QcDldUOPcCHgT8HXjCzY7Pxve7+UriyWsvMPkN6ltUg\n8DrSn8G7gXMCltVS2b7oEcdezOxF4OexH5Mxs88B95AG3PGkjxV9GegLWVcb3AT8i5l9CvgfwFzg\nY9mUS1RBT3ou9f01r7+Qfb2V9CBOrC7Lvn6rbnwx8LW2VtJexwB/R3pgbi/p+eTnuPs/Ba1KWuVN\npKH+RuB5YDtwprs/H7SqFnP3h8zsA8BngGtJT6280t1zP3ND97oREYmcDsaKiEROQS8iEjkFvYhI\n5BT0IiKRU9CLiEROQS8iEjkFvYhI5BT0IiKRU9CLiEROQS8iEjkFvYhI5BT0IiKR+/97K3Dty8rL\ntQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f8f42bd8b38>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ax = num_sig_change_clv.plot.hist(figsize=(4, 3))\n",
    "_ = ax.set_xticks(np.arange(1, 7, 1))\n",
    "ax.grid()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nothing at one because of the requirement of co-occurrence for reporting tumor-specific APA cases,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_sig_change_clv[num_sig_change_clv > 2].shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.4025974025974026"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "31 / 77"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So 31/77 cases report here have more than 2 CSs that undergo significant change cases "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyze recurrence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cases_df = aca_df.query('has_diff_aca == True')[['disease', 'gene_name']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDKN2A</th>\n",
       "      <th>EZH2</th>\n",
       "      <th>CCNE1</th>\n",
       "      <th>FGFR2</th>\n",
       "      <th>WT1</th>\n",
       "      <th>FGF2</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>MITF</th>\n",
       "      <th>AKT2</th>\n",
       "      <th>NFE2L2</th>\n",
       "      <th>CHURC1</th>\n",
       "      <th>KRAS</th>\n",
       "      <th>HNF1A</th>\n",
       "      <th>GNAS</th>\n",
       "      <th>FLT3</th>\n",
       "      <th>DRAM1</th>\n",
       "      <th>RNF43</th>\n",
       "      <th>MAX</th>\n",
       "      <th>MET</th>\n",
       "      <th>PTCH1</th>\n",
       "      <th>KIT</th>\n",
       "      <th>NFE2L3</th>\n",
       "      <th>CDKN2C</th>\n",
       "      <th>MYCL</th>\n",
       "      <th>APC</th>\n",
       "      <th>MDM4</th>\n",
       "      <th>RB1</th>\n",
       "      <th>ATM</th>\n",
       "      <th>SUFU</th>\n",
       "      <th>SOX2</th>\n",
       "      <th>AMER1</th>\n",
       "      <th>NF1</th>\n",
       "      <th>ALK</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           CDKN2A  EZH2  CCNE1  FGFR2  WT1  FGF2  BRCA1  MITF  AKT2  NFE2L2  CHURC1  KRAS  HNF1A  GNAS  FLT3  DRAM1  RNF43  MAX  MET  PTCH1  KIT  NFE2L3  CDKN2C  MYCL  APC  MDM4  RB1  ATM  SUFU  SOX2  AMER1  NF1  ALK\n",
       "gene_name  8       7     6      4      4    4     4      3     3     3       2       2     2      2     2     2      2      2    1    1      1    1       1       1     1    1     1    1    1     1     1      1    1  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cases_df.gene_name.value_counts().to_frame().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18,)"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_count = cases_df.gene_name.value_counts()\n",
    "_count[_count > 1].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So 18 genes have recurrence across different cancer types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(77, 2)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cases_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33,)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cases_df.gene_name.unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13,)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cases_df.disease.unique().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Of 14 `NT_diseases`, no tumor-specific APA cases are found in STAD"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Check a few genes in more detail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1064</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>FGF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6294</th>\n",
       "      <td>LUAD</td>\n",
       "      <td>FGF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7150</th>\n",
       "      <td>LUSC</td>\n",
       "      <td>FGF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7972</th>\n",
       "      <td>PRAD</td>\n",
       "      <td>FGF2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     disease gene_name\n",
       "1064  BRCA    FGF2    \n",
       "6294  LUAD    FGF2    \n",
       "7150  LUSC    FGF2    \n",
       "7972  PRAD    FGF2    "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cases_df.query('gene_name ==\"FGF2\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chr19|CCNE1|+|30314898', 'chr19|CCNE1|+|30315134',\n",
       "       'chr19|CCNE1|+|30315215', 'chr19|CCNE1|+|30314486'], dtype=object)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aca_df.query('gene_name == \"CCNE1\"').mkid.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "lean_aca_df = aca_df[['gene_name', 'has_diff_aca', 'disease']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33,)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lean_aca_df.query('has_diff_aca == True').gene_name.unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(77, 2)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lean_aca_df.query('has_diff_aca == True')[['gene_name', 'disease']].drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>251</th>\n",
       "      <th>1911</th>\n",
       "      <th>2658</th>\n",
       "      <th>3403</th>\n",
       "      <th>4061</th>\n",
       "      <th>4838</th>\n",
       "      <th>5537</th>\n",
       "      <th>8926</th>\n",
       "      <th>10046</th>\n",
       "      <th>10867</th>\n",
       "      <th>1064</th>\n",
       "      <th>6294</th>\n",
       "      <th>7150</th>\n",
       "      <th>7972</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "      <td>FGF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>has_diff_aca</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>disease</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>COAD</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>KICH</td>\n",
       "      <td>KIRC</td>\n",
       "      <td>KIRP</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>THCA</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>LUSC</td>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                251   1911   2658   3403   4061   4838   5537   8926  10046  10867  1064  6294  7150  7972\n",
       "gene_name     FGF2   FGF2   FGF2   FGF2   FGF2   FGF2   FGF2   FGF2   FGF2   FGF2   FGF2  FGF2  FGF2  FGF2\n",
       "has_diff_aca  False  False  False  False  False  False  False  False  False  False  True  True  True  True\n",
       "disease       BLCA   COAD   HNSC   KICH   KIRC   KIRP   LIHC   STAD   THCA   UCEC   BRCA  LUAD  LUSC  PRAD"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lean_aca_df.query('gene_name == \"FGF2\"').sort_values('has_diff_aca').T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mkid</th>\n",
       "      <th>mclv</th>\n",
       "      <th>strand</th>\n",
       "      <th>aclv</th>\n",
       "      <th>N_on</th>\n",
       "      <th>T_on</th>\n",
       "      <th>N_off</th>\n",
       "      <th>T_off</th>\n",
       "      <th>fisher_exact_p</th>\n",
       "      <th>N_base</th>\n",
       "      <th>T_base</th>\n",
       "      <th>N_on_ratio</th>\n",
       "      <th>T_on_ratio</th>\n",
       "      <th>N2T_ratio_diff</th>\n",
       "      <th>diff_is_significant</th>\n",
       "      <th>N2T_ratio_change</th>\n",
       "      <th>N2T_ratio_change_sig</th>\n",
       "      <th>has_diff_aca</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4707</th>\n",
       "      <td>KIRP</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>chr19|CCNE1|+|30314898</td>\n",
       "      <td>30314898</td>\n",
       "      <td>+</td>\n",
       "      <td>30314918</td>\n",
       "      <td>4</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "      <td>198</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>28</td>\n",
       "      <td>235</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.157447</td>\n",
       "      <td>0.014590</td>\n",
       "      <td>False</td>\n",
       "      <td>up</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4708</th>\n",
       "      <td>KIRP</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>chr19|CCNE1|+|30315134</td>\n",
       "      <td>30315134</td>\n",
       "      <td>+</td>\n",
       "      <td>30315119</td>\n",
       "      <td>21</td>\n",
       "      <td>108</td>\n",
       "      <td>7</td>\n",
       "      <td>127</td>\n",
       "      <td>0.004592</td>\n",
       "      <td>28</td>\n",
       "      <td>235</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.459574</td>\n",
       "      <td>-0.290426</td>\n",
       "      <td>True</td>\n",
       "      <td>down</td>\n",
       "      <td>down</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4709</th>\n",
       "      <td>KIRP</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>chr19|CCNE1|+|30315215</td>\n",
       "      <td>30315215</td>\n",
       "      <td>+</td>\n",
       "      <td>30315214</td>\n",
       "      <td>12</td>\n",
       "      <td>167</td>\n",
       "      <td>16</td>\n",
       "      <td>68</td>\n",
       "      <td>0.004500</td>\n",
       "      <td>28</td>\n",
       "      <td>235</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.710638</td>\n",
       "      <td>0.282067</td>\n",
       "      <td>True</td>\n",
       "      <td>up</td>\n",
       "      <td>up</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     disease gene_name                    mkid      mclv strand      aclv  N_on  T_on  N_off  T_off  fisher_exact_p  N_base  T_base  N_on_ratio  T_on_ratio  N2T_ratio_diff  diff_is_significant N2T_ratio_change N2T_ratio_change_sig  has_diff_aca\n",
       "4707  KIRP    CCNE1     chr19|CCNE1|+|30314898  30314898  +      30314918  4     37    24     198    1.000000        28      235     0.142857    0.157447    0.014590        False                up               None                 True        \n",
       "4708  KIRP    CCNE1     chr19|CCNE1|+|30315134  30315134  +      30315119  21    108   7      127    0.004592        28      235     0.750000    0.459574   -0.290426        True                 down             down                 True        \n",
       "4709  KIRP    CCNE1     chr19|CCNE1|+|30315215  30315215  +      30315214  12    167   16     68     0.004500        28      235     0.428571    0.710638    0.282067        True                 up               up                   True        "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aca_df.query('gene_name == \"CCNE1\"').query('disease == \"KIRP\"').sort_values('mclv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>106</th>\n",
       "      <th>1781</th>\n",
       "      <th>3295</th>\n",
       "      <th>3918</th>\n",
       "      <th>7823</th>\n",
       "      <th>8682</th>\n",
       "      <th>9889</th>\n",
       "      <th>10746</th>\n",
       "      <th>913</th>\n",
       "      <th>2505</th>\n",
       "      <th>4707</th>\n",
       "      <th>5410</th>\n",
       "      <th>6133</th>\n",
       "      <th>6994</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "      <td>CCNE1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>has_diff_aca</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>disease</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>COAD</td>\n",
       "      <td>KICH</td>\n",
       "      <td>KIRC</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>STAD</td>\n",
       "      <td>THCA</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>KIRP</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>LUSC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                106   1781   3295   3918   7823   8682   9889  10746    913   2505   4707   5410   6133   6994\n",
       "gene_name     CCNE1  CCNE1  CCNE1  CCNE1  CCNE1  CCNE1  CCNE1  CCNE1  CCNE1  CCNE1  CCNE1  CCNE1  CCNE1  CCNE1\n",
       "has_diff_aca  False  False  False  False  False  False  False  False  True   True   True   True   True   True \n",
       "disease       BLCA   COAD   KICH   KIRC   PRAD   STAD   THCA   UCEC   BRCA   HNSC   KIRP   LIHC   LUAD   LUSC "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lean_aca_df.query('gene_name == \"CCNE1\"').sort_values('has_diff_aca').T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>682</th>\n",
       "      <th>1558</th>\n",
       "      <th>2290</th>\n",
       "      <th>3086</th>\n",
       "      <th>3731</th>\n",
       "      <th>5219</th>\n",
       "      <th>5924</th>\n",
       "      <th>6755</th>\n",
       "      <th>7603</th>\n",
       "      <th>8389</th>\n",
       "      <th>9594</th>\n",
       "      <th>10515</th>\n",
       "      <th>4494</th>\n",
       "      <th>11257</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "      <td>RNF43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>has_diff_aca</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>disease</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>COAD</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>KICH</td>\n",
       "      <td>KIRP</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>LUSC</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>STAD</td>\n",
       "      <td>THCA</td>\n",
       "      <td>KIRC</td>\n",
       "      <td>UCEC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                682   1558   2290   3086   3731   5219   5924   6755   7603   8389   9594  10515   4494  11257\n",
       "gene_name     RNF43  RNF43  RNF43  RNF43  RNF43  RNF43  RNF43  RNF43  RNF43  RNF43  RNF43  RNF43  RNF43  RNF43\n",
       "has_diff_aca  False  False  False  False  False  False  False  False  False  False  False  False  True   True \n",
       "disease       BLCA   BRCA   COAD   HNSC   KICH   KIRP   LIHC   LUAD   LUSC   PRAD   STAD   THCA   KIRC   UCEC "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lean_aca_df.query('gene_name == \"RNF43\"').sort_values('has_diff_aca').T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>126</th>\n",
       "      <th>939</th>\n",
       "      <th>6157</th>\n",
       "      <th>7016</th>\n",
       "      <th>8708</th>\n",
       "      <th>10763</th>\n",
       "      <th>1795</th>\n",
       "      <th>2528</th>\n",
       "      <th>3308</th>\n",
       "      <th>3937</th>\n",
       "      <th>4724</th>\n",
       "      <th>5426</th>\n",
       "      <th>7847</th>\n",
       "      <th>9915</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>has_diff_aca</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>disease</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>LUSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>COAD</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>KICH</td>\n",
       "      <td>KIRC</td>\n",
       "      <td>KIRP</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>THCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 126     939    6157    7016    8708   10763    1795    2528    3308    3937    4724    5426    7847    9915\n",
       "gene_name     CDKN2A  CDKN2A  CDKN2A  CDKN2A  CDKN2A  CDKN2A  CDKN2A  CDKN2A  CDKN2A  CDKN2A  CDKN2A  CDKN2A  CDKN2A  CDKN2A\n",
       "has_diff_aca  False   False   False   False   False   False   True    True    True    True    True    True    True    True  \n",
       "disease       BLCA    BRCA    LUAD    LUSC    STAD    UCEC    COAD    HNSC    KICH    KIRC    KIRP    LIHC    PRAD    THCA  "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lean_aca_df.query('gene_name == \"CDKN2A\"').sort_values('has_diff_aca').T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>232</th>\n",
       "      <th>1893</th>\n",
       "      <th>2640</th>\n",
       "      <th>3388</th>\n",
       "      <th>7130</th>\n",
       "      <th>8889</th>\n",
       "      <th>10853</th>\n",
       "      <th>1046</th>\n",
       "      <th>4042</th>\n",
       "      <th>4819</th>\n",
       "      <th>5518</th>\n",
       "      <th>6273</th>\n",
       "      <th>7954</th>\n",
       "      <th>10026</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>EZH2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>has_diff_aca</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>disease</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>COAD</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>KICH</td>\n",
       "      <td>LUSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>KIRC</td>\n",
       "      <td>KIRP</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>THCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                232   1893   2640   3388   7130   8889  10853  1046  4042  4819  5518  6273  7954 10026\n",
       "gene_name     EZH2   EZH2   EZH2   EZH2   EZH2   EZH2   EZH2   EZH2  EZH2  EZH2  EZH2  EZH2  EZH2  EZH2\n",
       "has_diff_aca  False  False  False  False  False  False  False  True  True  True  True  True  True  True\n",
       "disease       BLCA   COAD   HNSC   KICH   LUSC   STAD   UCEC   BRCA  KIRC  KIRP  LIHC  LUAD  PRAD  THCA"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lean_aca_df.query('gene_name == \"EZH2\"').sort_values('has_diff_aca').T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add stop codon info to aca_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# clv_sc_df = pd.read_csv('/projects/btl/zxue/tasrkleat-TCGA-results/hexamer-patch/extracted-annotated-clv-sc-mapping-window-20bp-PC-NMD-v4.csv')\n",
    "clv_sc_df = pd.read_csv('./reference_data/annotated-clv-sc-mapping.csv.gz', compression='gzip')\n",
    "# 642 => 632: removed transcripts from ['HG122_PATCH', 'HG1257_PATCH']\n",
    "assert clv_sc_df.shape[0] == 632"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "aca_genes_df = clv_sc_df.query('gene_name in {0}'.format(aca_df.query('has_diff_aca == True').gene_name.unique().tolist()))\n",
    "num_sc_df = aca_genes_df.groupby('gene_name').apply(lambda g: g.sc.unique().shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_sc_df[num_sc_df > 1].shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8484848484848485"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_sc_df[num_sc_df > 1].shape[0] / num_sc_df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "clv_sc_df.drop(['transcript_id', 'clv'], axis=1, inplace=True)\n",
    "clv_sc_df.drop_duplicates(inplace=True)\n",
    "clv_sc_df.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# window = 15 bp\n",
    "assert clv_sc_df.shape[0] == 561"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>sc</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>is_cds_end_NF</th>\n",
       "      <th>is_cds_start_NF</th>\n",
       "      <th>utr_beg</th>\n",
       "      <th>cluster_idx</th>\n",
       "      <th>aclv</th>\n",
       "      <th>clv0</th>\n",
       "      <th>hexamer</th>\n",
       "      <th>hexamer_id</th>\n",
       "      <th>hexamer_loc0</th>\n",
       "      <th>hexamer_loc</th>\n",
       "      <th>sdist_aclv2ahex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>133761070</td>\n",
       "      <td>+</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>133761071</td>\n",
       "      <td>1</td>\n",
       "      <td>133763062</td>\n",
       "      <td>133763061</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>1</td>\n",
       "      <td>133763038</td>\n",
       "      <td>133763039</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name seqname          source         sc strand     gene_source transcript_source  is_cds_end_NF  is_cds_start_NF    utr_beg  cluster_idx       aclv       clv0 hexamer  hexamer_id  hexamer_loc0  hexamer_loc  sdist_aclv2ahex\n",
       "0  ABL1      9       protein_coding  133761070  +      ensembl_havana  ensembl_havana    False          False            133761071  1            133763062  133763061  AATAAA  1           133763038     133763039    23.0           "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clv_sc_df.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein_coding             457\n",
       "nonsense_mediated_decay    104\n",
       "Name: source, dtype: int64"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clv_sc_df.source.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "assert clv_sc_df.gene_name.unique().shape[0] == 114"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# ANNOT_GENE2CLV_DD = clv_sc_df.groupby(['gene_name']).apply(lambda g: g.aclv.unique())\n",
    "ANNOT_CLV2SC_DD = clv_sc_df.groupby(['gene_name', 'aclv']).apply(\n",
    "    lambda g: np.sort(g.sc.unique())).to_dict()\n",
    "ANNOT_CLV2SRC_DD = clv_sc_df.groupby(['gene_name', 'aclv']).apply(\n",
    "    lambda g: np.sort(g.source.unique())).to_dict()\n",
    "\n",
    "# def map_mclv2aclv(row):\n",
    "#     \"\"\"map KLEAT predicted clv to a closest annotated clv\"\"\"\n",
    "#     gene = row.gene_name\n",
    "#     mclv = row['mclv']\n",
    "#     poss_aclvs = ANNOT_GENE2CLV_DD[gene]\n",
    "#     return poss_aclvs[np.argmin(np.abs(poss_aclvs - mclv))]\n",
    "# cdf['aclv2'] = cdf.apply(map_mclv2aclv, axis=1)\n",
    "# cdf['dist_mclv2aclv'] = np.abs(cdf['mclv'].values - cdf['aclv'].values)\n",
    "\n",
    "def map_aclv2sc(row):\n",
    "    \"\"\"with aclv assigned, now add the corresponding annotated stop codon information\"\"\"\n",
    "    # ('gene_name', 'aclv') acts like a shorter kid (skid)\n",
    "    skid = (row.gene_name, row.aclv)\n",
    "    scs = ANNOT_CLV2SC_DD[skid]\n",
    "    return tuple(scs.tolist())\n",
    "\n",
    "def map_aclv2src(row):\n",
    "    \"\"\"with aclv assigned, now add the corresponding annotated stop codon information\"\"\"\n",
    "    # ('gene_name', 'aclv') acts like a shorter kid (skid)\n",
    "    skid = (row.gene_name, row.aclv)\n",
    "    srcs = ANNOT_CLV2SRC_DD[skid]\n",
    "    return tuple(srcs.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mkid</th>\n",
       "      <th>mclv</th>\n",
       "      <th>strand</th>\n",
       "      <th>aclv</th>\n",
       "      <th>N_on</th>\n",
       "      <th>T_on</th>\n",
       "      <th>N_off</th>\n",
       "      <th>T_off</th>\n",
       "      <th>fisher_exact_p</th>\n",
       "      <th>N_base</th>\n",
       "      <th>T_base</th>\n",
       "      <th>N_on_ratio</th>\n",
       "      <th>T_on_ratio</th>\n",
       "      <th>N2T_ratio_diff</th>\n",
       "      <th>diff_is_significant</th>\n",
       "      <th>N2T_ratio_change</th>\n",
       "      <th>N2T_ratio_change_sig</th>\n",
       "      <th>has_diff_aca</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>chr9|ABL1|+|133589972</td>\n",
       "      <td>133589972</td>\n",
       "      <td>+</td>\n",
       "      <td>133763062</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>18</td>\n",
       "      <td>382</td>\n",
       "      <td>1.0</td>\n",
       "      <td>19</td>\n",
       "      <td>405</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.05679</td>\n",
       "      <td>0.004159</td>\n",
       "      <td>False</td>\n",
       "      <td>up</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  disease gene_name                   mkid       mclv strand       aclv  N_on  T_on  N_off  T_off  fisher_exact_p  N_base  T_base  N_on_ratio  T_on_ratio  N2T_ratio_diff  diff_is_significant N2T_ratio_change N2T_ratio_change_sig  has_diff_aca\n",
       "0  BLCA    ABL1      chr9|ABL1|+|133589972  133589972  +      133763062  1     23    18     382    1.0             19      405     0.052632    0.05679     0.004159        False                up               None                 False       "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aca_df.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "aca_with_sc_df = aca_df.copy()\n",
    "# sc: stop codons\n",
    "aca_with_sc_df['sc_list'] = aca_with_sc_df[['gene_name', 'aclv']].apply(map_aclv2sc, axis=1)\n",
    "aca_with_sc_df['src_list'] = aca_with_sc_df[['gene_name', 'aclv']].apply(map_aclv2src, axis=1)\n",
    "aca_with_sc_df['num_sc'] = aca_with_sc_df['sc_list'].apply(len)\n",
    "\n",
    "aca_with_sc_df['sc_list'] = aca_with_sc_df.sc_list.apply(lambda v: ', '.join(map(str, v)))\n",
    "aca_with_sc_df['src_list'] = aca_with_sc_df.src_list.apply(lambda v: ', '.join(map(str, v)))\n",
    "\n",
    "aca_with_sc_df.query('gene_name == \"FGF2\"').mclv.sort_values().unique()\n",
    "\n",
    "aca_with_sc_df.query('gene_name == \"FGF2\"').aclv.sort_values().unique()\n",
    "\n",
    "aca_with_sc_df.head(1)\n",
    "\n",
    "aca_with_sc_df.to_csv('./results/aca_pval_with_sc_info.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_df = aca_with_sc_df.query('has_diff_aca')[['mkid', 'mclv', 'num_sc']].drop_duplicates()\n",
    "_df.query('num_sc > 1').shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "297"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.1111111111111111"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "33 / 297"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So 11% of the predicted CSs in 33 genes are mapped to more than one stop codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "_df = aca_with_sc_df.query('has_diff_aca')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "_sr = _df.groupby('gene_name').apply(\n",
    "    lambda g: g.src_list.str.contains('nonsense_mediated_decay').any())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_sr[_sr].shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.48484848484848486"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "16 / 33."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So about half of the reported genes have stop codons involved in NMD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multi-trend gene search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "comp_between_pairs = lambda row: row.dropna().unique().shape[0] > 1\n",
    "\n",
    "# if there is at least one cleavage site that shifts differently and significantly between any pair of cancer types,\n",
    "# then it counts as having multiple shift patterns\n",
    "def has_multi_trends(grp):\n",
    "#     print(grp[['disease', 'gene_name']]).drop_duplicates().values\n",
    "    # check if a gene show different types of significant freq shift trends in different cancer types\n",
    "    change_df = grp.pivot(index='mclv', columns='disease', values='N2T_ratio_change_sig')\n",
    "    \n",
    "    # if a change is not significant for a mclv site in one sstype, then don't use it for comparison at all\n",
    "    diff = change_df.apply(comp_between_pairs, axis=1)\n",
    "    return diff.values.any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['disease', 'gene_name', 'mkid', 'mclv', 'strand', 'aclv', 'N_on', 'T_on', 'N_off', 'T_off', 'fisher_exact_p', 'N_base', 'T_base', 'N_on_ratio', 'T_on_ratio', 'N2T_ratio_diff', 'diff_is_significant', 'N2T_ratio_change', 'N2T_ratio_change_sig', 'has_diff_aca'], dtype='object')"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aca_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# mt: multi-trend\n",
    "mt_df = aca_with_sc_df\\\n",
    "    .query('has_diff_aca == True')\\\n",
    "    .groupby('gene_name')\\\n",
    "    .apply(has_multi_trends)\\\n",
    "    .to_frame(name='has_mt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>CDKN2A</th>\n",
       "      <th>FLT3</th>\n",
       "      <th>HNF1A</th>\n",
       "      <th>NFE2L2</th>\n",
       "      <th>RNF43</th>\n",
       "      <th>WT1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>has_mt</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name  CDKN2A  FLT3  HNF1A  NFE2L2  RNF43   WT1\n",
       "has_mt     True    True  True   True    True   True"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt_df.query('has_mt == True').T"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
