{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import datetime\n",
    "\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 250)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2016-05-04 17:08:41.903406\n"
     ]
    }
   ],
   "source": [
    "import datetime\n",
    "print(datetime.datetime.now())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "read_lengths.csv.gz is exported from https://bigquery.cloud.google.com/table/isb-cgc:tcga_seq_metadata.RNAseq_FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df0 = pd.read_csv('https://cghub.ucsc.edu/reports/SUMMARY_STATS/LATEST_MANIFEST.tsv', low_memory=False, sep='\\t')\n",
    "df = pd.read_csv('read_lengths.csv.gz', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(22408, 174)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CGHub_analysisID</th>\n",
       "      <th>AliquotBarcode</th>\n",
       "      <th>fastqFilename</th>\n",
       "      <th>FastQC_html_url</th>\n",
       "      <th>FastQC_zip_url</th>\n",
       "      <th>BasicStats</th>\n",
       "      <th>numTotalSeq</th>\n",
       "      <th>numPoorSeq</th>\n",
       "      <th>minLength</th>\n",
       "      <th>maxLength</th>\n",
       "      <th>delLength</th>\n",
       "      <th>pctGC</th>\n",
       "      <th>PerBaseSeqQual</th>\n",
       "      <th>PBSQ10_0</th>\n",
       "      <th>PBSQ50_0</th>\n",
       "      <th>PBSQ10_1</th>\n",
       "      <th>PBSQ50_1</th>\n",
       "      <th>PBSQ10_2</th>\n",
       "      <th>PBSQ50_2</th>\n",
       "      <th>PBSQ10_3</th>\n",
       "      <th>PBSQ50_3</th>\n",
       "      <th>PBSQ10_4</th>\n",
       "      <th>PBSQ50_4</th>\n",
       "      <th>PBSQ10_5</th>\n",
       "      <th>PBSQ50_5</th>\n",
       "      <th>PBSQ10_6</th>\n",
       "      <th>PBSQ50_6</th>\n",
       "      <th>PBSQ10_7</th>\n",
       "      <th>PBSQ50_7</th>\n",
       "      <th>PBSQ10_8</th>\n",
       "      <th>PBSQ50_8</th>\n",
       "      <th>PBSQ10_9</th>\n",
       "      <th>PBSQ50_9</th>\n",
       "      <th>PBSQ10_10</th>\n",
       "      <th>PBSQ50_10</th>\n",
       "      <th>PerTileSeqQual</th>\n",
       "      <th>PerSeqQualScore</th>\n",
       "      <th>PSQC_0</th>\n",
       "      <th>PSQC_1</th>\n",
       "      <th>PSQC_2</th>\n",
       "      <th>PSQC_3</th>\n",
       "      <th>PSQC_4</th>\n",
       "      <th>PSQC_5</th>\n",
       "      <th>PSQC_6</th>\n",
       "      <th>PSQC_7</th>\n",
       "      <th>PSQC_8</th>\n",
       "      <th>PSQC_9</th>\n",
       "      <th>PSQC_10</th>\n",
       "      <th>PerBaseSeqCont</th>\n",
       "      <th>PBSC_G_0</th>\n",
       "      <th>PBSC_A_0</th>\n",
       "      <th>PBSC_T_0</th>\n",
       "      <th>PBSC_C_0</th>\n",
       "      <th>PBSC_G_1</th>\n",
       "      <th>PBSC_A_1</th>\n",
       "      <th>PBSC_T_1</th>\n",
       "      <th>PBSC_C_1</th>\n",
       "      <th>PBSC_G_2</th>\n",
       "      <th>PBSC_A_2</th>\n",
       "      <th>PBSC_T_2</th>\n",
       "      <th>PBSC_C_2</th>\n",
       "      <th>PBSC_G_3</th>\n",
       "      <th>PBSC_A_3</th>\n",
       "      <th>PBSC_T_3</th>\n",
       "      <th>PBSC_C_3</th>\n",
       "      <th>PBSC_G_4</th>\n",
       "      <th>PBSC_A_4</th>\n",
       "      <th>PBSC_T_4</th>\n",
       "      <th>PBSC_C_4</th>\n",
       "      <th>PBSC_G_5</th>\n",
       "      <th>PBSC_A_5</th>\n",
       "      <th>PBSC_T_5</th>\n",
       "      <th>PBSC_C_5</th>\n",
       "      <th>PBSC_G_6</th>\n",
       "      <th>PBSC_A_6</th>\n",
       "      <th>PBSC_T_6</th>\n",
       "      <th>PBSC_C_6</th>\n",
       "      <th>PBSC_G_7</th>\n",
       "      <th>PBSC_A_7</th>\n",
       "      <th>PBSC_T_7</th>\n",
       "      <th>PBSC_C_7</th>\n",
       "      <th>PBSC_G_8</th>\n",
       "      <th>PBSC_A_8</th>\n",
       "      <th>PBSC_T_8</th>\n",
       "      <th>PBSC_C_8</th>\n",
       "      <th>PBSC_G_9</th>\n",
       "      <th>PBSC_A_9</th>\n",
       "      <th>PBSC_T_9</th>\n",
       "      <th>PBSC_C_9</th>\n",
       "      <th>PBSC_G_10</th>\n",
       "      <th>PBSC_A_10</th>\n",
       "      <th>PBSC_T_10</th>\n",
       "      <th>PBSC_C_10</th>\n",
       "      <th>PerSeqGCcont</th>\n",
       "      <th>PSGCC_0</th>\n",
       "      <th>PSGCC_1</th>\n",
       "      <th>PSGCC_2</th>\n",
       "      <th>PSGCC_3</th>\n",
       "      <th>PSGCC_4</th>\n",
       "      <th>PSGCC_5</th>\n",
       "      <th>PSGCC_6</th>\n",
       "      <th>PSGCC_7</th>\n",
       "      <th>PSGCC_8</th>\n",
       "      <th>PSGCC_9</th>\n",
       "      <th>PSGCC_10</th>\n",
       "      <th>PerBaseNcont</th>\n",
       "      <th>PBNC_1</th>\n",
       "      <th>PBNC_2</th>\n",
       "      <th>PBNC_3</th>\n",
       "      <th>PBNC_4</th>\n",
       "      <th>PBNC_5</th>\n",
       "      <th>PBNC_6</th>\n",
       "      <th>PBNC_7</th>\n",
       "      <th>PBNC_8</th>\n",
       "      <th>PBNC_9</th>\n",
       "      <th>PBNC_10</th>\n",
       "      <th>SeqLenDist</th>\n",
       "      <th>SLD_modeVal</th>\n",
       "      <th>SLD_neg10</th>\n",
       "      <th>SLD_neg9</th>\n",
       "      <th>SLD_neg8</th>\n",
       "      <th>SLD_neg7</th>\n",
       "      <th>SLD_neg6</th>\n",
       "      <th>SLD_neg5</th>\n",
       "      <th>SLD_neg4</th>\n",
       "      <th>SLD_neg3</th>\n",
       "      <th>SLD_neg2</th>\n",
       "      <th>SLD_neg1</th>\n",
       "      <th>SLD_mode</th>\n",
       "      <th>SLD_pos1</th>\n",
       "      <th>SLD_pos2</th>\n",
       "      <th>SLD_pos3</th>\n",
       "      <th>SLD_pos4</th>\n",
       "      <th>SLD_pos5</th>\n",
       "      <th>SLD_pos6</th>\n",
       "      <th>SLD_pos7</th>\n",
       "      <th>SLD_pos8</th>\n",
       "      <th>SLD_pos9</th>\n",
       "      <th>SLD_pos10</th>\n",
       "      <th>SeqDupLev</th>\n",
       "      <th>Total_Dedup_Pct</th>\n",
       "      <th>Dedup_relPct_1</th>\n",
       "      <th>Dedup_relPct_2</th>\n",
       "      <th>Dedup_relPct_3</th>\n",
       "      <th>Dedup_relPct_4</th>\n",
       "      <th>Dedup_relPct_5</th>\n",
       "      <th>Dedup_relPct_6</th>\n",
       "      <th>Dedup_relPct_7</th>\n",
       "      <th>Dedup_relPct_8</th>\n",
       "      <th>Dedup_relPct_9</th>\n",
       "      <th>Dedup_relPct_ge10</th>\n",
       "      <th>Dedup_relPct_ge50</th>\n",
       "      <th>Dedup_relPct_ge100</th>\n",
       "      <th>Dedup_relPct_ge500</th>\n",
       "      <th>Dedup_relPct_ge1k</th>\n",
       "      <th>Dedup_relPct_ge5k</th>\n",
       "      <th>Dedup_relPct_ge10k</th>\n",
       "      <th>OverrepSeq</th>\n",
       "      <th>OverRepSeq_1</th>\n",
       "      <th>OverRepSeq_1_Pct</th>\n",
       "      <th>OverRepSeq_2</th>\n",
       "      <th>OverRepSeq_2_Pct</th>\n",
       "      <th>OverRepSeq_3</th>\n",
       "      <th>OverRepSeq_3_Pct</th>\n",
       "      <th>AdaptCont</th>\n",
       "      <th>KmerCont</th>\n",
       "      <th>Kmer_Seq_1</th>\n",
       "      <th>Kmer_Ratio_1</th>\n",
       "      <th>Kmer_Seq_2</th>\n",
       "      <th>Kmer_Ratio_2</th>\n",
       "      <th>Kmer_Seq_3</th>\n",
       "      <th>Kmer_Ratio_3</th>\n",
       "      <th>Kmer_Ratio_Min</th>\n",
       "      <th>Kmer_Num_Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00065a62-5e18-4223-a884-12fca053a109</td>\n",
       "      <td>TCGA-2Z-A9J3-01A-12R-A38C-07</td>\n",
       "      <td>140516_UNC12-SN629_0369_AC4GGKACXX_GGCTAC_L002...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>pass</td>\n",
       "      <td>78492984</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>pass</td>\n",
       "      <td>31</td>\n",
       "      <td>34</td>\n",
       "      <td>35.0</td>\n",
       "      <td>37</td>\n",
       "      <td>35.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>31</td>\n",
       "      <td>38</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>2</td>\n",
       "      <td>34</td>\n",
       "      <td>36</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>fail</td>\n",
       "      <td>27.530030</td>\n",
       "      <td>12.408029</td>\n",
       "      <td>17.071692</td>\n",
       "      <td>42.990249</td>\n",
       "      <td>23.629694</td>\n",
       "      <td>33.048944</td>\n",
       "      <td>20.767008</td>\n",
       "      <td>22.554355</td>\n",
       "      <td>23.770993</td>\n",
       "      <td>34.710779</td>\n",
       "      <td>21.849800</td>\n",
       "      <td>19.668428</td>\n",
       "      <td>24.409745</td>\n",
       "      <td>24.892885</td>\n",
       "      <td>25.559626</td>\n",
       "      <td>25.137745</td>\n",
       "      <td>25.024649</td>\n",
       "      <td>24.871529</td>\n",
       "      <td>25.208490</td>\n",
       "      <td>24.895332</td>\n",
       "      <td>24.597984</td>\n",
       "      <td>25.573245</td>\n",
       "      <td>25.093488</td>\n",
       "      <td>24.735283</td>\n",
       "      <td>25.257113</td>\n",
       "      <td>25.034442</td>\n",
       "      <td>24.996694</td>\n",
       "      <td>24.711751</td>\n",
       "      <td>25.149930</td>\n",
       "      <td>24.964309</td>\n",
       "      <td>24.835695</td>\n",
       "      <td>25.050066</td>\n",
       "      <td>24.531676</td>\n",
       "      <td>25.426731</td>\n",
       "      <td>25.086860</td>\n",
       "      <td>24.954734</td>\n",
       "      <td>24.619681</td>\n",
       "      <td>25.010791</td>\n",
       "      <td>25.409735</td>\n",
       "      <td>24.959794</td>\n",
       "      <td>24.438681</td>\n",
       "      <td>25.514561</td>\n",
       "      <td>25.489523</td>\n",
       "      <td>24.557236</td>\n",
       "      <td>pass</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>39</td>\n",
       "      <td>43</td>\n",
       "      <td>47</td>\n",
       "      <td>50</td>\n",
       "      <td>54</td>\n",
       "      <td>57</td>\n",
       "      <td>61</td>\n",
       "      <td>66</td>\n",
       "      <td>100</td>\n",
       "      <td>pass</td>\n",
       "      <td>0.074186</td>\n",
       "      <td>0.002505</td>\n",
       "      <td>0.002519</td>\n",
       "      <td>0.012485</td>\n",
       "      <td>0.007138</td>\n",
       "      <td>0.006976</td>\n",
       "      <td>0.038672</td>\n",
       "      <td>0.018324</td>\n",
       "      <td>0.000837</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>pass</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fail</td>\n",
       "      <td>34.622958</td>\n",
       "      <td>24.206718</td>\n",
       "      <td>9.243894</td>\n",
       "      <td>5.809638</td>\n",
       "      <td>4.151374</td>\n",
       "      <td>3.126133</td>\n",
       "      <td>2.535093</td>\n",
       "      <td>2.090047</td>\n",
       "      <td>1.773165</td>\n",
       "      <td>1.452206</td>\n",
       "      <td>17.491370</td>\n",
       "      <td>5.383170</td>\n",
       "      <td>10.216731</td>\n",
       "      <td>3.750783</td>\n",
       "      <td>7.068727</td>\n",
       "      <td>1.041764</td>\n",
       "      <td>0.659186</td>\n",
       "      <td>warn</td>\n",
       "      <td>GATCGGAAGAGCACACGTCTGAACTCCAGTCACGGCTACATCTCGTAT</td>\n",
       "      <td>0.246763</td>\n",
       "      <td>AGATCGGAAGAGCACACGTCTGAACTCCAGTCACGGCTACATCTCGTA</td>\n",
       "      <td>0.16580</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>CTCGTAT</td>\n",
       "      <td>34.167786</td>\n",
       "      <td>AGATCGG</td>\n",
       "      <td>28.136084</td>\n",
       "      <td>TCTCGTA</td>\n",
       "      <td>20.962118</td>\n",
       "      <td>16.753263</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00065a62-5e18-4223-a884-12fca053a109</td>\n",
       "      <td>TCGA-2Z-A9J3-01A-12R-A38C-07</td>\n",
       "      <td>140516_UNC12-SN629_0369_AC4GGKACXX_GGCTAC_L002...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>pass</td>\n",
       "      <td>78492984</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>1</td>\n",
       "      <td>49</td>\n",
       "      <td>pass</td>\n",
       "      <td>30</td>\n",
       "      <td>34</td>\n",
       "      <td>35.0</td>\n",
       "      <td>37</td>\n",
       "      <td>35.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>30</td>\n",
       "      <td>39</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>2</td>\n",
       "      <td>33</td>\n",
       "      <td>36</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>fail</td>\n",
       "      <td>27.198083</td>\n",
       "      <td>12.039404</td>\n",
       "      <td>17.773835</td>\n",
       "      <td>42.988678</td>\n",
       "      <td>23.579986</td>\n",
       "      <td>32.822984</td>\n",
       "      <td>20.767820</td>\n",
       "      <td>22.829210</td>\n",
       "      <td>23.657261</td>\n",
       "      <td>34.578154</td>\n",
       "      <td>21.961956</td>\n",
       "      <td>19.802629</td>\n",
       "      <td>24.309275</td>\n",
       "      <td>24.579144</td>\n",
       "      <td>25.885254</td>\n",
       "      <td>25.226328</td>\n",
       "      <td>24.919756</td>\n",
       "      <td>25.061318</td>\n",
       "      <td>25.278590</td>\n",
       "      <td>24.740336</td>\n",
       "      <td>24.768200</td>\n",
       "      <td>25.537878</td>\n",
       "      <td>25.216845</td>\n",
       "      <td>24.477077</td>\n",
       "      <td>25.049790</td>\n",
       "      <td>25.100404</td>\n",
       "      <td>25.117851</td>\n",
       "      <td>24.731955</td>\n",
       "      <td>24.935348</td>\n",
       "      <td>25.240007</td>\n",
       "      <td>24.965208</td>\n",
       "      <td>24.859436</td>\n",
       "      <td>24.380360</td>\n",
       "      <td>25.269378</td>\n",
       "      <td>25.442460</td>\n",
       "      <td>24.907803</td>\n",
       "      <td>24.727565</td>\n",
       "      <td>25.058307</td>\n",
       "      <td>25.433388</td>\n",
       "      <td>24.780740</td>\n",
       "      <td>24.549014</td>\n",
       "      <td>25.306096</td>\n",
       "      <td>25.350874</td>\n",
       "      <td>24.794016</td>\n",
       "      <td>pass</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>39</td>\n",
       "      <td>43</td>\n",
       "      <td>47</td>\n",
       "      <td>50</td>\n",
       "      <td>54</td>\n",
       "      <td>57</td>\n",
       "      <td>61</td>\n",
       "      <td>66</td>\n",
       "      <td>100</td>\n",
       "      <td>pass</td>\n",
       "      <td>0.264213</td>\n",
       "      <td>0.294763</td>\n",
       "      <td>0.292768</td>\n",
       "      <td>0.256810</td>\n",
       "      <td>0.375943</td>\n",
       "      <td>0.663836</td>\n",
       "      <td>0.610065</td>\n",
       "      <td>0.564528</td>\n",
       "      <td>0.348678</td>\n",
       "      <td>0.506099</td>\n",
       "      <td>pass</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fail</td>\n",
       "      <td>35.836806</td>\n",
       "      <td>25.363791</td>\n",
       "      <td>9.484503</td>\n",
       "      <td>5.751943</td>\n",
       "      <td>4.031005</td>\n",
       "      <td>3.070190</td>\n",
       "      <td>2.499580</td>\n",
       "      <td>2.103125</td>\n",
       "      <td>1.831038</td>\n",
       "      <td>1.533931</td>\n",
       "      <td>17.151544</td>\n",
       "      <td>5.299548</td>\n",
       "      <td>9.916259</td>\n",
       "      <td>3.639022</td>\n",
       "      <td>6.870589</td>\n",
       "      <td>1.087557</td>\n",
       "      <td>0.366375</td>\n",
       "      <td>warn</td>\n",
       "      <td>AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCG</td>\n",
       "      <td>0.174365</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>AGATCGG</td>\n",
       "      <td>28.025505</td>\n",
       "      <td>TGGTCGC</td>\n",
       "      <td>23.397738</td>\n",
       "      <td>GCGTCGT</td>\n",
       "      <td>23.207323</td>\n",
       "      <td>17.234850</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0009f27d-a053-49b0-a030-ba9c1469224f</td>\n",
       "      <td>TCGA-OR-A5KX-01A-11R-A29S-07</td>\n",
       "      <td>130806_UNC13-SN749_0298_BC2E3KACXX_AGTTCC_L002...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>pass</td>\n",
       "      <td>64552799</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>pass</td>\n",
       "      <td>30</td>\n",
       "      <td>33</td>\n",
       "      <td>33.0</td>\n",
       "      <td>37</td>\n",
       "      <td>33.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>30</td>\n",
       "      <td>39</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>2</td>\n",
       "      <td>32</td>\n",
       "      <td>35</td>\n",
       "      <td>36</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>fail</td>\n",
       "      <td>29.265817</td>\n",
       "      <td>13.475011</td>\n",
       "      <td>15.248408</td>\n",
       "      <td>42.010763</td>\n",
       "      <td>25.023976</td>\n",
       "      <td>32.116618</td>\n",
       "      <td>20.732930</td>\n",
       "      <td>22.126476</td>\n",
       "      <td>24.534986</td>\n",
       "      <td>34.704568</td>\n",
       "      <td>21.444782</td>\n",
       "      <td>19.315664</td>\n",
       "      <td>23.899588</td>\n",
       "      <td>25.462132</td>\n",
       "      <td>24.689230</td>\n",
       "      <td>25.949050</td>\n",
       "      <td>24.451229</td>\n",
       "      <td>24.343288</td>\n",
       "      <td>25.394174</td>\n",
       "      <td>25.811309</td>\n",
       "      <td>24.167171</td>\n",
       "      <td>26.249006</td>\n",
       "      <td>24.302454</td>\n",
       "      <td>25.281370</td>\n",
       "      <td>25.876342</td>\n",
       "      <td>25.741469</td>\n",
       "      <td>24.095793</td>\n",
       "      <td>24.286395</td>\n",
       "      <td>24.393565</td>\n",
       "      <td>25.709042</td>\n",
       "      <td>23.968601</td>\n",
       "      <td>25.928792</td>\n",
       "      <td>24.107040</td>\n",
       "      <td>24.484223</td>\n",
       "      <td>24.325847</td>\n",
       "      <td>27.082891</td>\n",
       "      <td>24.273119</td>\n",
       "      <td>25.777725</td>\n",
       "      <td>25.554144</td>\n",
       "      <td>24.395012</td>\n",
       "      <td>25.564380</td>\n",
       "      <td>24.535373</td>\n",
       "      <td>25.610403</td>\n",
       "      <td>24.289844</td>\n",
       "      <td>pass</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>39</td>\n",
       "      <td>44</td>\n",
       "      <td>48</td>\n",
       "      <td>51</td>\n",
       "      <td>54</td>\n",
       "      <td>58</td>\n",
       "      <td>62</td>\n",
       "      <td>66</td>\n",
       "      <td>100</td>\n",
       "      <td>pass</td>\n",
       "      <td>0.104547</td>\n",
       "      <td>0.009420</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.001190</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>pass</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fail</td>\n",
       "      <td>31.594789</td>\n",
       "      <td>21.192824</td>\n",
       "      <td>9.506226</td>\n",
       "      <td>5.660822</td>\n",
       "      <td>3.983366</td>\n",
       "      <td>3.007602</td>\n",
       "      <td>2.416336</td>\n",
       "      <td>2.018633</td>\n",
       "      <td>1.672463</td>\n",
       "      <td>1.455059</td>\n",
       "      <td>17.541818</td>\n",
       "      <td>5.644347</td>\n",
       "      <td>12.414002</td>\n",
       "      <td>4.660810</td>\n",
       "      <td>5.919057</td>\n",
       "      <td>0.506184</td>\n",
       "      <td>2.400449</td>\n",
       "      <td>fail</td>\n",
       "      <td>AGATCGGAAGAGCACACGTCTGAACTCCAGTCACAGTTCCGTATCTCG</td>\n",
       "      <td>1.257402</td>\n",
       "      <td>GATCGGAAGAGCACACGTCTGAACTCCAGTCACAGTTCCGTATCTCGT</td>\n",
       "      <td>1.08258</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>AGATCGG</td>\n",
       "      <td>39.204740</td>\n",
       "      <td>ATCTCGT</td>\n",
       "      <td>39.093940</td>\n",
       "      <td>TATCTCG</td>\n",
       "      <td>21.762442</td>\n",
       "      <td>20.323175</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0009f27d-a053-49b0-a030-ba9c1469224f</td>\n",
       "      <td>TCGA-OR-A5KX-01A-11R-A29S-07</td>\n",
       "      <td>130806_UNC13-SN749_0298_BC2E3KACXX_AGTTCC_L002...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>pass</td>\n",
       "      <td>64552799</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>pass</td>\n",
       "      <td>28</td>\n",
       "      <td>31</td>\n",
       "      <td>31.0</td>\n",
       "      <td>37</td>\n",
       "      <td>31.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>23</td>\n",
       "      <td>38</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>2</td>\n",
       "      <td>29</td>\n",
       "      <td>34</td>\n",
       "      <td>36</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>fail</td>\n",
       "      <td>29.269218</td>\n",
       "      <td>13.213105</td>\n",
       "      <td>15.178619</td>\n",
       "      <td>42.339058</td>\n",
       "      <td>25.067976</td>\n",
       "      <td>32.082061</td>\n",
       "      <td>20.690003</td>\n",
       "      <td>22.159960</td>\n",
       "      <td>24.848891</td>\n",
       "      <td>34.408552</td>\n",
       "      <td>21.426579</td>\n",
       "      <td>19.315977</td>\n",
       "      <td>24.235270</td>\n",
       "      <td>24.243227</td>\n",
       "      <td>26.207930</td>\n",
       "      <td>25.313573</td>\n",
       "      <td>24.903245</td>\n",
       "      <td>25.029965</td>\n",
       "      <td>25.784518</td>\n",
       "      <td>24.282273</td>\n",
       "      <td>26.615324</td>\n",
       "      <td>25.159660</td>\n",
       "      <td>24.348506</td>\n",
       "      <td>23.876510</td>\n",
       "      <td>25.743032</td>\n",
       "      <td>25.968272</td>\n",
       "      <td>24.113834</td>\n",
       "      <td>24.174862</td>\n",
       "      <td>25.512936</td>\n",
       "      <td>26.107308</td>\n",
       "      <td>24.044047</td>\n",
       "      <td>24.335709</td>\n",
       "      <td>24.214731</td>\n",
       "      <td>24.623978</td>\n",
       "      <td>26.084916</td>\n",
       "      <td>25.076375</td>\n",
       "      <td>25.619535</td>\n",
       "      <td>24.369447</td>\n",
       "      <td>25.747289</td>\n",
       "      <td>24.263729</td>\n",
       "      <td>26.024164</td>\n",
       "      <td>24.587494</td>\n",
       "      <td>24.367782</td>\n",
       "      <td>25.020561</td>\n",
       "      <td>warn</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>39</td>\n",
       "      <td>43</td>\n",
       "      <td>47</td>\n",
       "      <td>51</td>\n",
       "      <td>54</td>\n",
       "      <td>58</td>\n",
       "      <td>62</td>\n",
       "      <td>67</td>\n",
       "      <td>100</td>\n",
       "      <td>pass</td>\n",
       "      <td>0.070632</td>\n",
       "      <td>0.198232</td>\n",
       "      <td>0.544576</td>\n",
       "      <td>0.835779</td>\n",
       "      <td>0.620532</td>\n",
       "      <td>0.790429</td>\n",
       "      <td>0.947627</td>\n",
       "      <td>0.907143</td>\n",
       "      <td>0.749264</td>\n",
       "      <td>0.988286</td>\n",
       "      <td>pass</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fail</td>\n",
       "      <td>34.266587</td>\n",
       "      <td>23.933363</td>\n",
       "      <td>9.527153</td>\n",
       "      <td>5.590441</td>\n",
       "      <td>3.925953</td>\n",
       "      <td>2.982969</td>\n",
       "      <td>2.327996</td>\n",
       "      <td>2.032910</td>\n",
       "      <td>1.711178</td>\n",
       "      <td>1.429472</td>\n",
       "      <td>17.103376</td>\n",
       "      <td>5.604830</td>\n",
       "      <td>11.973742</td>\n",
       "      <td>4.370839</td>\n",
       "      <td>5.629063</td>\n",
       "      <td>0.439424</td>\n",
       "      <td>1.417291</td>\n",
       "      <td>fail</td>\n",
       "      <td>AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCG</td>\n",
       "      <td>1.215249</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>AGATCGG</td>\n",
       "      <td>39.016556</td>\n",
       "      <td>GTGGTCG</td>\n",
       "      <td>35.261444</td>\n",
       "      <td>GCGTCGT</td>\n",
       "      <td>35.131214</td>\n",
       "      <td>32.124573</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>000f476b-7d55-4fd1-8f72-622ab00e009b</td>\n",
       "      <td>TCGA-EJ-7784-01A-11R-2118-07</td>\n",
       "      <td>120514_UNC16-SN851_0159_BC0TARACXX_CGATGT_L003...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>pass</td>\n",
       "      <td>81024858</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>1</td>\n",
       "      <td>49</td>\n",
       "      <td>pass</td>\n",
       "      <td>31</td>\n",
       "      <td>34</td>\n",
       "      <td>35.0</td>\n",
       "      <td>37</td>\n",
       "      <td>37.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>33</td>\n",
       "      <td>40</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>2</td>\n",
       "      <td>35</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>fail</td>\n",
       "      <td>26.671582</td>\n",
       "      <td>12.819763</td>\n",
       "      <td>16.458082</td>\n",
       "      <td>44.050573</td>\n",
       "      <td>21.676697</td>\n",
       "      <td>31.990909</td>\n",
       "      <td>21.598513</td>\n",
       "      <td>24.733881</td>\n",
       "      <td>23.768912</td>\n",
       "      <td>34.553707</td>\n",
       "      <td>21.231209</td>\n",
       "      <td>20.446172</td>\n",
       "      <td>24.110908</td>\n",
       "      <td>24.593077</td>\n",
       "      <td>25.931231</td>\n",
       "      <td>25.364784</td>\n",
       "      <td>24.883602</td>\n",
       "      <td>25.240159</td>\n",
       "      <td>25.071155</td>\n",
       "      <td>24.805084</td>\n",
       "      <td>24.545708</td>\n",
       "      <td>25.786209</td>\n",
       "      <td>25.324149</td>\n",
       "      <td>24.343934</td>\n",
       "      <td>24.922397</td>\n",
       "      <td>25.261460</td>\n",
       "      <td>25.143413</td>\n",
       "      <td>24.672729</td>\n",
       "      <td>24.830888</td>\n",
       "      <td>25.317945</td>\n",
       "      <td>25.130380</td>\n",
       "      <td>24.720786</td>\n",
       "      <td>24.404201</td>\n",
       "      <td>25.418330</td>\n",
       "      <td>25.567852</td>\n",
       "      <td>24.609617</td>\n",
       "      <td>24.641606</td>\n",
       "      <td>25.244639</td>\n",
       "      <td>25.349600</td>\n",
       "      <td>24.764156</td>\n",
       "      <td>24.432072</td>\n",
       "      <td>25.467349</td>\n",
       "      <td>25.525071</td>\n",
       "      <td>24.575508</td>\n",
       "      <td>pass</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>39</td>\n",
       "      <td>43</td>\n",
       "      <td>47</td>\n",
       "      <td>50</td>\n",
       "      <td>53</td>\n",
       "      <td>57</td>\n",
       "      <td>60</td>\n",
       "      <td>65</td>\n",
       "      <td>100</td>\n",
       "      <td>warn</td>\n",
       "      <td>7.971947</td>\n",
       "      <td>0.000278</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pass</td>\n",
       "      <td>48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fail</td>\n",
       "      <td>40.687403</td>\n",
       "      <td>27.466543</td>\n",
       "      <td>12.432203</td>\n",
       "      <td>7.642441</td>\n",
       "      <td>5.257081</td>\n",
       "      <td>4.031188</td>\n",
       "      <td>2.936302</td>\n",
       "      <td>2.394908</td>\n",
       "      <td>1.966194</td>\n",
       "      <td>1.621288</td>\n",
       "      <td>16.949951</td>\n",
       "      <td>4.511024</td>\n",
       "      <td>8.714685</td>\n",
       "      <td>2.254745</td>\n",
       "      <td>1.727041</td>\n",
       "      <td>0.067535</td>\n",
       "      <td>0.026871</td>\n",
       "      <td>pass</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>CTCGTAT</td>\n",
       "      <td>19.006138</td>\n",
       "      <td>TATCTCG</td>\n",
       "      <td>13.624064</td>\n",
       "      <td>GATCGGA</td>\n",
       "      <td>13.038096</td>\n",
       "      <td>9.642544</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       CGHub_analysisID                AliquotBarcode  \\\n",
       "0  00065a62-5e18-4223-a884-12fca053a109  TCGA-2Z-A9J3-01A-12R-A38C-07   \n",
       "1  00065a62-5e18-4223-a884-12fca053a109  TCGA-2Z-A9J3-01A-12R-A38C-07   \n",
       "2  0009f27d-a053-49b0-a030-ba9c1469224f  TCGA-OR-A5KX-01A-11R-A29S-07   \n",
       "3  0009f27d-a053-49b0-a030-ba9c1469224f  TCGA-OR-A5KX-01A-11R-A29S-07   \n",
       "4  000f476b-7d55-4fd1-8f72-622ab00e009b  TCGA-EJ-7784-01A-11R-2118-07   \n",
       "\n",
       "                                       fastqFilename  \\\n",
       "0  140516_UNC12-SN629_0369_AC4GGKACXX_GGCTAC_L002...   \n",
       "1  140516_UNC12-SN629_0369_AC4GGKACXX_GGCTAC_L002...   \n",
       "2  130806_UNC13-SN749_0298_BC2E3KACXX_AGTTCC_L002...   \n",
       "3  130806_UNC13-SN749_0298_BC2E3KACXX_AGTTCC_L002...   \n",
       "4  120514_UNC16-SN851_0159_BC0TARACXX_CGATGT_L003...   \n",
       "\n",
       "                                     FastQC_html_url  \\\n",
       "0  https://storage.cloud.google.com/isb-cgc-open/...   \n",
       "1  https://storage.cloud.google.com/isb-cgc-open/...   \n",
       "2  https://storage.cloud.google.com/isb-cgc-open/...   \n",
       "3  https://storage.cloud.google.com/isb-cgc-open/...   \n",
       "4  https://storage.cloud.google.com/isb-cgc-open/...   \n",
       "\n",
       "                                      FastQC_zip_url BasicStats  numTotalSeq  \\\n",
       "0  https://storage.cloud.google.com/isb-cgc-open/...       pass     78492984   \n",
       "1  https://storage.cloud.google.com/isb-cgc-open/...       pass     78492984   \n",
       "2  https://storage.cloud.google.com/isb-cgc-open/...       pass     64552799   \n",
       "3  https://storage.cloud.google.com/isb-cgc-open/...       pass     64552799   \n",
       "4  https://storage.cloud.google.com/isb-cgc-open/...       pass     81024858   \n",
       "\n",
       "   numPoorSeq  minLength  maxLength  delLength  pctGC PerBaseSeqQual  \\\n",
       "0           0         48         48          1     50           pass   \n",
       "1           0         48         48          1     49           pass   \n",
       "2           0         48         48          1     50           pass   \n",
       "3           0         48         48          1     50           pass   \n",
       "4           0         48         48          1     49           pass   \n",
       "\n",
       "   PBSQ10_0  PBSQ50_0  PBSQ10_1  PBSQ50_1  PBSQ10_2  PBSQ50_2  PBSQ10_3  \\\n",
       "0        31        34      35.0        37      35.0      39.0      36.0   \n",
       "1        30        34      35.0        37      35.0      39.0      34.0   \n",
       "2        30        33      33.0        37      33.0      39.0      34.0   \n",
       "3        28        31      31.0        37      31.0      39.0      31.0   \n",
       "4        31        34      35.0        37      37.0      39.0      37.0   \n",
       "\n",
       "   PBSQ50_3  PBSQ10_4  PBSQ50_4  PBSQ10_5  PBSQ50_5  PBSQ10_6  PBSQ50_6  \\\n",
       "0      40.0      35.0      40.0      35.0      40.0      35.0      40.0   \n",
       "1      40.0      34.0      40.0      34.0      40.0      33.0      40.0   \n",
       "2      40.0      34.0      40.0      34.0      40.0      33.0      40.0   \n",
       "3      40.0      31.0      40.0      30.0      40.0      30.0      40.0   \n",
       "4      41.0      37.0      41.0      36.0      41.0      36.0      41.0   \n",
       "\n",
       "   PBSQ10_7  PBSQ50_7  PBSQ10_8  PBSQ50_8  PBSQ10_9  PBSQ50_9  PBSQ10_10  \\\n",
       "0      34.0      40.0      34.0      40.0      33.0      40.0         31   \n",
       "1      33.0      40.0      33.0      40.0      32.0      40.0         30   \n",
       "2      32.0      40.0      32.0      40.0      31.0      39.0         30   \n",
       "3      29.0      40.0      28.0      39.0      26.0      39.0         23   \n",
       "4      35.0      41.0      35.0      40.0      35.0      40.0         33   \n",
       "\n",
       "   PBSQ50_10 PerTileSeqQual PerSeqQualScore  PSQC_0  PSQC_1  PSQC_2  PSQC_3  \\\n",
       "0         38           pass            pass       2      34      36      37   \n",
       "1         39           pass            pass       2      33      36      37   \n",
       "2         39           fail            pass       2      32      35      36   \n",
       "3         38           warn            pass       2      29      34      36   \n",
       "4         40           fail            pass       2      35      37      38   \n",
       "\n",
       "   PSQC_4  PSQC_5  PSQC_6  PSQC_7  PSQC_8  PSQC_9  PSQC_10 PerBaseSeqCont  \\\n",
       "0      38      38      39      39      39      39       39           fail   \n",
       "1      38      38      39      39      39      39       39           fail   \n",
       "2      37      38      39      39      39      39       39           fail   \n",
       "3      37      38      38      39      39      39       39           fail   \n",
       "4      38      39      39      39      39      39       39           fail   \n",
       "\n",
       "    PBSC_G_0   PBSC_A_0   PBSC_T_0   PBSC_C_0   PBSC_G_1   PBSC_A_1  \\\n",
       "0  27.530030  12.408029  17.071692  42.990249  23.629694  33.048944   \n",
       "1  27.198083  12.039404  17.773835  42.988678  23.579986  32.822984   \n",
       "2  29.265817  13.475011  15.248408  42.010763  25.023976  32.116618   \n",
       "3  29.269218  13.213105  15.178619  42.339058  25.067976  32.082061   \n",
       "4  26.671582  12.819763  16.458082  44.050573  21.676697  31.990909   \n",
       "\n",
       "    PBSC_T_1   PBSC_C_1   PBSC_G_2   PBSC_A_2   PBSC_T_2   PBSC_C_2  \\\n",
       "0  20.767008  22.554355  23.770993  34.710779  21.849800  19.668428   \n",
       "1  20.767820  22.829210  23.657261  34.578154  21.961956  19.802629   \n",
       "2  20.732930  22.126476  24.534986  34.704568  21.444782  19.315664   \n",
       "3  20.690003  22.159960  24.848891  34.408552  21.426579  19.315977   \n",
       "4  21.598513  24.733881  23.768912  34.553707  21.231209  20.446172   \n",
       "\n",
       "    PBSC_G_3   PBSC_A_3   PBSC_T_3   PBSC_C_3   PBSC_G_4   PBSC_A_4  \\\n",
       "0  24.409745  24.892885  25.559626  25.137745  25.024649  24.871529   \n",
       "1  24.309275  24.579144  25.885254  25.226328  24.919756  25.061318   \n",
       "2  23.899588  25.462132  24.689230  25.949050  24.451229  24.343288   \n",
       "3  24.235270  24.243227  26.207930  25.313573  24.903245  25.029965   \n",
       "4  24.110908  24.593077  25.931231  25.364784  24.883602  25.240159   \n",
       "\n",
       "    PBSC_T_4   PBSC_C_4   PBSC_G_5   PBSC_A_5   PBSC_T_5   PBSC_C_5  \\\n",
       "0  25.208490  24.895332  24.597984  25.573245  25.093488  24.735283   \n",
       "1  25.278590  24.740336  24.768200  25.537878  25.216845  24.477077   \n",
       "2  25.394174  25.811309  24.167171  26.249006  24.302454  25.281370   \n",
       "3  25.784518  24.282273  26.615324  25.159660  24.348506  23.876510   \n",
       "4  25.071155  24.805084  24.545708  25.786209  25.324149  24.343934   \n",
       "\n",
       "    PBSC_G_6   PBSC_A_6   PBSC_T_6   PBSC_C_6   PBSC_G_7   PBSC_A_7  \\\n",
       "0  25.257113  25.034442  24.996694  24.711751  25.149930  24.964309   \n",
       "1  25.049790  25.100404  25.117851  24.731955  24.935348  25.240007   \n",
       "2  25.876342  25.741469  24.095793  24.286395  24.393565  25.709042   \n",
       "3  25.743032  25.968272  24.113834  24.174862  25.512936  26.107308   \n",
       "4  24.922397  25.261460  25.143413  24.672729  24.830888  25.317945   \n",
       "\n",
       "    PBSC_T_7   PBSC_C_7   PBSC_G_8   PBSC_A_8   PBSC_T_8   PBSC_C_8  \\\n",
       "0  24.835695  25.050066  24.531676  25.426731  25.086860  24.954734   \n",
       "1  24.965208  24.859436  24.380360  25.269378  25.442460  24.907803   \n",
       "2  23.968601  25.928792  24.107040  24.484223  24.325847  27.082891   \n",
       "3  24.044047  24.335709  24.214731  24.623978  26.084916  25.076375   \n",
       "4  25.130380  24.720786  24.404201  25.418330  25.567852  24.609617   \n",
       "\n",
       "    PBSC_G_9   PBSC_A_9   PBSC_T_9   PBSC_C_9  PBSC_G_10  PBSC_A_10  \\\n",
       "0  24.619681  25.010791  25.409735  24.959794  24.438681  25.514561   \n",
       "1  24.727565  25.058307  25.433388  24.780740  24.549014  25.306096   \n",
       "2  24.273119  25.777725  25.554144  24.395012  25.564380  24.535373   \n",
       "3  25.619535  24.369447  25.747289  24.263729  26.024164  24.587494   \n",
       "4  24.641606  25.244639  25.349600  24.764156  24.432072  25.467349   \n",
       "\n",
       "   PBSC_T_10  PBSC_C_10 PerSeqGCcont  PSGCC_0  PSGCC_1  PSGCC_2  PSGCC_3  \\\n",
       "0  25.489523  24.557236         pass        0       34       39       43   \n",
       "1  25.350874  24.794016         pass        0       33       39       43   \n",
       "2  25.610403  24.289844         pass        0       33       39       44   \n",
       "3  24.367782  25.020561         warn        0       33       39       43   \n",
       "4  25.525071  24.575508         pass        0       34       39       43   \n",
       "\n",
       "   PSGCC_4  PSGCC_5  PSGCC_6  PSGCC_7  PSGCC_8  PSGCC_9  PSGCC_10  \\\n",
       "0       47       50       54       57       61       66       100   \n",
       "1       47       50       54       57       61       66       100   \n",
       "2       48       51       54       58       62       66       100   \n",
       "3       47       51       54       58       62       67       100   \n",
       "4       47       50       53       57       60       65       100   \n",
       "\n",
       "  PerBaseNcont    PBNC_1    PBNC_2    PBNC_3    PBNC_4    PBNC_5    PBNC_6  \\\n",
       "0         pass  0.074186  0.002505  0.002519  0.012485  0.007138  0.006976   \n",
       "1         pass  0.264213  0.294763  0.292768  0.256810  0.375943  0.663836   \n",
       "2         pass  0.104547  0.009420  0.000005  0.000020  0.001190  0.000025   \n",
       "3         pass  0.070632  0.198232  0.544576  0.835779  0.620532  0.790429   \n",
       "4         warn  7.971947  0.000278  0.000000  0.000043  0.000000  0.000000   \n",
       "\n",
       "     PBNC_7    PBNC_8    PBNC_9   PBNC_10 SeqLenDist  SLD_modeVal  SLD_neg10  \\\n",
       "0  0.038672  0.018324  0.000837  0.000001       pass           48        NaN   \n",
       "1  0.610065  0.564528  0.348678  0.506099       pass           48        NaN   \n",
       "2  0.000000  0.000000  0.000000  0.000033       pass           48        NaN   \n",
       "3  0.947627  0.907143  0.749264  0.988286       pass           48        NaN   \n",
       "4  0.000000  0.000000  0.000020  0.000000       pass           48        NaN   \n",
       "\n",
       "   SLD_neg9  SLD_neg8  SLD_neg7  SLD_neg6  SLD_neg5  SLD_neg4  SLD_neg3  \\\n",
       "0       NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "1       NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "2       NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "3       NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "4       NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "\n",
       "   SLD_neg2  SLD_neg1  SLD_mode  SLD_pos1  SLD_pos2  SLD_pos3  SLD_pos4  \\\n",
       "0       NaN       NaN       1.0       NaN       NaN       NaN       NaN   \n",
       "1       NaN       NaN       1.0       NaN       NaN       NaN       NaN   \n",
       "2       NaN       NaN       1.0       NaN       NaN       NaN       NaN   \n",
       "3       NaN       NaN       1.0       NaN       NaN       NaN       NaN   \n",
       "4       NaN       NaN       1.0       NaN       NaN       NaN       NaN   \n",
       "\n",
       "   SLD_pos5  SLD_pos6  SLD_pos7  SLD_pos8  SLD_pos9  SLD_pos10 SeqDupLev  \\\n",
       "0       NaN       NaN       NaN       NaN       NaN        NaN      fail   \n",
       "1       NaN       NaN       NaN       NaN       NaN        NaN      fail   \n",
       "2       NaN       NaN       NaN       NaN       NaN        NaN      fail   \n",
       "3       NaN       NaN       NaN       NaN       NaN        NaN      fail   \n",
       "4       NaN       NaN       NaN       NaN       NaN        NaN      fail   \n",
       "\n",
       "   Total_Dedup_Pct  Dedup_relPct_1  Dedup_relPct_2  Dedup_relPct_3  \\\n",
       "0        34.622958       24.206718        9.243894        5.809638   \n",
       "1        35.836806       25.363791        9.484503        5.751943   \n",
       "2        31.594789       21.192824        9.506226        5.660822   \n",
       "3        34.266587       23.933363        9.527153        5.590441   \n",
       "4        40.687403       27.466543       12.432203        7.642441   \n",
       "\n",
       "   Dedup_relPct_4  Dedup_relPct_5  Dedup_relPct_6  Dedup_relPct_7  \\\n",
       "0        4.151374        3.126133        2.535093        2.090047   \n",
       "1        4.031005        3.070190        2.499580        2.103125   \n",
       "2        3.983366        3.007602        2.416336        2.018633   \n",
       "3        3.925953        2.982969        2.327996        2.032910   \n",
       "4        5.257081        4.031188        2.936302        2.394908   \n",
       "\n",
       "   Dedup_relPct_8  Dedup_relPct_9  Dedup_relPct_ge10  Dedup_relPct_ge50  \\\n",
       "0        1.773165        1.452206          17.491370           5.383170   \n",
       "1        1.831038        1.533931          17.151544           5.299548   \n",
       "2        1.672463        1.455059          17.541818           5.644347   \n",
       "3        1.711178        1.429472          17.103376           5.604830   \n",
       "4        1.966194        1.621288          16.949951           4.511024   \n",
       "\n",
       "   Dedup_relPct_ge100  Dedup_relPct_ge500  Dedup_relPct_ge1k  \\\n",
       "0           10.216731            3.750783           7.068727   \n",
       "1            9.916259            3.639022           6.870589   \n",
       "2           12.414002            4.660810           5.919057   \n",
       "3           11.973742            4.370839           5.629063   \n",
       "4            8.714685            2.254745           1.727041   \n",
       "\n",
       "   Dedup_relPct_ge5k  Dedup_relPct_ge10k OverrepSeq  \\\n",
       "0           1.041764            0.659186       warn   \n",
       "1           1.087557            0.366375       warn   \n",
       "2           0.506184            2.400449       fail   \n",
       "3           0.439424            1.417291       fail   \n",
       "4           0.067535            0.026871       pass   \n",
       "\n",
       "                                       OverRepSeq_1  OverRepSeq_1_Pct  \\\n",
       "0  GATCGGAAGAGCACACGTCTGAACTCCAGTCACGGCTACATCTCGTAT          0.246763   \n",
       "1  AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCG          0.174365   \n",
       "2  AGATCGGAAGAGCACACGTCTGAACTCCAGTCACAGTTCCGTATCTCG          1.257402   \n",
       "3  AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCG          1.215249   \n",
       "4                                               NaN               NaN   \n",
       "\n",
       "                                       OverRepSeq_2  OverRepSeq_2_Pct  \\\n",
       "0  AGATCGGAAGAGCACACGTCTGAACTCCAGTCACGGCTACATCTCGTA           0.16580   \n",
       "1                                               NaN               NaN   \n",
       "2  GATCGGAAGAGCACACGTCTGAACTCCAGTCACAGTTCCGTATCTCGT           1.08258   \n",
       "3                                               NaN               NaN   \n",
       "4                                               NaN               NaN   \n",
       "\n",
       "  OverRepSeq_3  OverRepSeq_3_Pct AdaptCont KmerCont Kmer_Seq_1  Kmer_Ratio_1  \\\n",
       "0          NaN               NaN      pass     fail    CTCGTAT     34.167786   \n",
       "1          NaN               NaN      pass     fail    AGATCGG     28.025505   \n",
       "2          NaN               NaN      pass     fail    AGATCGG     39.204740   \n",
       "3          NaN               NaN      pass     fail    AGATCGG     39.016556   \n",
       "4          NaN               NaN      pass     fail    CTCGTAT     19.006138   \n",
       "\n",
       "  Kmer_Seq_2  Kmer_Ratio_2 Kmer_Seq_3  Kmer_Ratio_3  Kmer_Ratio_Min  \\\n",
       "0    AGATCGG     28.136084    TCTCGTA     20.962118       16.753263   \n",
       "1    TGGTCGC     23.397738    GCGTCGT     23.207323       17.234850   \n",
       "2    ATCTCGT     39.093940    TATCTCG     21.762442       20.323175   \n",
       "3    GTGGTCG     35.261444    GCGTCGT     35.131214       32.124573   \n",
       "4    TATCTCG     13.624064    GATCGGA     13.038096        9.642544   \n",
       "\n",
       "   Kmer_Num_Seq  \n",
       "0            21  \n",
       "1            21  \n",
       "2            21  \n",
       "3            21  \n",
       "4            21  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ndf = df.merge(df0, left_on='CGHub_analysisID', right_on='analysis_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(22408, 211)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CGHub_analysisID', 'AliquotBarcode', 'fastqFilename',\n",
       "       'FastQC_html_url', 'FastQC_zip_url', 'BasicStats', 'numTotalSeq',\n",
       "       'numPoorSeq', 'minLength', 'maxLength', 'delLength', 'pctGC',\n",
       "       'PerBaseSeqQual', 'PBSQ10_0', 'PBSQ50_0', 'PBSQ10_1', 'PBSQ50_1',\n",
       "       'PBSQ10_2', 'PBSQ50_2', 'PBSQ10_3', 'PBSQ50_3', 'PBSQ10_4',\n",
       "       'PBSQ50_4', 'PBSQ10_5', 'PBSQ50_5', 'PBSQ10_6', 'PBSQ50_6',\n",
       "       'PBSQ10_7', 'PBSQ50_7', 'PBSQ10_8', 'PBSQ50_8', 'PBSQ10_9',\n",
       "       'PBSQ50_9', 'PBSQ10_10', 'PBSQ50_10', 'PerTileSeqQual',\n",
       "       'PerSeqQualScore', 'PSQC_0', 'PSQC_1', 'PSQC_2', 'PSQC_3', 'PSQC_4',\n",
       "       'PSQC_5', 'PSQC_6', 'PSQC_7', 'PSQC_8', 'PSQC_9', 'PSQC_10',\n",
       "       'PerBaseSeqCont', 'PBSC_G_0', 'PBSC_A_0', 'PBSC_T_0', 'PBSC_C_0',\n",
       "       'PBSC_G_1', 'PBSC_A_1', 'PBSC_T_1', 'PBSC_C_1', 'PBSC_G_2',\n",
       "       'PBSC_A_2', 'PBSC_T_2', 'PBSC_C_2', 'PBSC_G_3', 'PBSC_A_3',\n",
       "       'PBSC_T_3', 'PBSC_C_3', 'PBSC_G_4', 'PBSC_A_4', 'PBSC_T_4',\n",
       "       'PBSC_C_4', 'PBSC_G_5', 'PBSC_A_5', 'PBSC_T_5', 'PBSC_C_5',\n",
       "       'PBSC_G_6', 'PBSC_A_6', 'PBSC_T_6', 'PBSC_C_6', 'PBSC_G_7',\n",
       "       'PBSC_A_7', 'PBSC_T_7', 'PBSC_C_7', 'PBSC_G_8', 'PBSC_A_8',\n",
       "       'PBSC_T_8', 'PBSC_C_8', 'PBSC_G_9', 'PBSC_A_9', 'PBSC_T_9',\n",
       "       'PBSC_C_9', 'PBSC_G_10', 'PBSC_A_10', 'PBSC_T_10', 'PBSC_C_10',\n",
       "       'PerSeqGCcont', 'PSGCC_0', 'PSGCC_1', 'PSGCC_2', 'PSGCC_3',\n",
       "       'PSGCC_4', 'PSGCC_5', 'PSGCC_6', 'PSGCC_7', 'PSGCC_8', 'PSGCC_9',\n",
       "       'PSGCC_10', 'PerBaseNcont', 'PBNC_1', 'PBNC_2', 'PBNC_3', 'PBNC_4',\n",
       "       'PBNC_5', 'PBNC_6', 'PBNC_7', 'PBNC_8', 'PBNC_9', 'PBNC_10',\n",
       "       'SeqLenDist', 'SLD_modeVal', 'SLD_neg10', 'SLD_neg9', 'SLD_neg8',\n",
       "       'SLD_neg7', 'SLD_neg6', 'SLD_neg5', 'SLD_neg4', 'SLD_neg3',\n",
       "       'SLD_neg2', 'SLD_neg1', 'SLD_mode', 'SLD_pos1', 'SLD_pos2',\n",
       "       'SLD_pos3', 'SLD_pos4', 'SLD_pos5', 'SLD_pos6', 'SLD_pos7',\n",
       "       'SLD_pos8', 'SLD_pos9', 'SLD_pos10', 'SeqDupLev', 'Total_Dedup_Pct',\n",
       "       'Dedup_relPct_1', 'Dedup_relPct_2', 'Dedup_relPct_3',\n",
       "       'Dedup_relPct_4', 'Dedup_relPct_5', 'Dedup_relPct_6',\n",
       "       'Dedup_relPct_7', 'Dedup_relPct_8', 'Dedup_relPct_9',\n",
       "       'Dedup_relPct_ge10', 'Dedup_relPct_ge50', 'Dedup_relPct_ge100',\n",
       "       'Dedup_relPct_ge500', 'Dedup_relPct_ge1k', 'Dedup_relPct_ge5k',\n",
       "       'Dedup_relPct_ge10k', 'OverrepSeq', 'OverRepSeq_1',\n",
       "       'OverRepSeq_1_Pct', 'OverRepSeq_2', 'OverRepSeq_2_Pct',\n",
       "       'OverRepSeq_3', 'OverRepSeq_3_Pct', 'AdaptCont', 'KmerCont',\n",
       "       'Kmer_Seq_1', 'Kmer_Ratio_1', 'Kmer_Seq_2', 'Kmer_Ratio_2',\n",
       "       'Kmer_Seq_3', 'Kmer_Ratio_3', 'Kmer_Ratio_Min', 'Kmer_Num_Seq',\n",
       "       'study', 'barcode', 'disease', 'disease_name', 'sample_type',\n",
       "       'sample_type_name', 'analyte_type', 'library_type', 'center',\n",
       "       'center_name', 'platform', 'platform_name', 'assembly', 'filename',\n",
       "       'files_size', 'checksum', 'analysis_id', 'aliquot_id',\n",
       "       'participant_id', 'sample_id', 'tss_id', 'sample_accession',\n",
       "       'published', 'uploaded', 'modified', 'state', 'sample_type_code',\n",
       "       'analyte_type_code', 'platform_full_name', 'file_type', 'reason',\n",
       "       'reagent_vendor', 'reagent_name', 'catalog_number', 'is_custom',\n",
       "       'target_file_url', 'probe_file_url'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf.columns.values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Centers that produce RNA-Seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "UNC-LCCC    19238\n",
       "BCCAGSC      2830\n",
       "BI            340\n",
       "Name: center, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf[ndf.library_type == 'RNA-Seq'].center.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Platform full names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Illumina HiSeq 2000             20968\n",
       "Illumina Genome Analyzer IIx      736\n",
       "Illumina Genome Analyzer II       648\n",
       "Illumina HiSeq 2500                56\n",
       "Name: platform_full_name, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf.platform_full_name.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Some verification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCGA    22408\n",
      "Name: study, dtype: int64\n",
      "\n",
      "RNA-Seq    22408\n",
      "Name: library_type, dtype: int64\n",
      "\n",
      "Live    22408\n",
      "Name: state, dtype: int64\n",
      "\n",
      "fasta    22408\n",
      "Name: file_type, dtype: int64\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for col in ['study', 'library_type', 'state', 'file_type']:\n",
    "    print(ndf[col].value_counts())\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1     22407\n",
       "45        1\n",
       "Name: delLength, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf.delLength.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CGHub_analysisID</th>\n",
       "      <th>AliquotBarcode</th>\n",
       "      <th>fastqFilename</th>\n",
       "      <th>FastQC_html_url</th>\n",
       "      <th>FastQC_zip_url</th>\n",
       "      <th>BasicStats</th>\n",
       "      <th>numTotalSeq</th>\n",
       "      <th>numPoorSeq</th>\n",
       "      <th>minLength</th>\n",
       "      <th>maxLength</th>\n",
       "      <th>delLength</th>\n",
       "      <th>pctGC</th>\n",
       "      <th>PerBaseSeqQual</th>\n",
       "      <th>PBSQ10_0</th>\n",
       "      <th>PBSQ50_0</th>\n",
       "      <th>PBSQ10_1</th>\n",
       "      <th>PBSQ50_1</th>\n",
       "      <th>PBSQ10_2</th>\n",
       "      <th>PBSQ50_2</th>\n",
       "      <th>PBSQ10_3</th>\n",
       "      <th>PBSQ50_3</th>\n",
       "      <th>PBSQ10_4</th>\n",
       "      <th>PBSQ50_4</th>\n",
       "      <th>PBSQ10_5</th>\n",
       "      <th>PBSQ50_5</th>\n",
       "      <th>PBSQ10_6</th>\n",
       "      <th>PBSQ50_6</th>\n",
       "      <th>PBSQ10_7</th>\n",
       "      <th>PBSQ50_7</th>\n",
       "      <th>PBSQ10_8</th>\n",
       "      <th>PBSQ50_8</th>\n",
       "      <th>PBSQ10_9</th>\n",
       "      <th>PBSQ50_9</th>\n",
       "      <th>PBSQ10_10</th>\n",
       "      <th>PBSQ50_10</th>\n",
       "      <th>PerTileSeqQual</th>\n",
       "      <th>PerSeqQualScore</th>\n",
       "      <th>PSQC_0</th>\n",
       "      <th>PSQC_1</th>\n",
       "      <th>PSQC_2</th>\n",
       "      <th>PSQC_3</th>\n",
       "      <th>PSQC_4</th>\n",
       "      <th>PSQC_5</th>\n",
       "      <th>PSQC_6</th>\n",
       "      <th>PSQC_7</th>\n",
       "      <th>PSQC_8</th>\n",
       "      <th>PSQC_9</th>\n",
       "      <th>PSQC_10</th>\n",
       "      <th>PerBaseSeqCont</th>\n",
       "      <th>PBSC_G_0</th>\n",
       "      <th>PBSC_A_0</th>\n",
       "      <th>PBSC_T_0</th>\n",
       "      <th>PBSC_C_0</th>\n",
       "      <th>PBSC_G_1</th>\n",
       "      <th>PBSC_A_1</th>\n",
       "      <th>PBSC_T_1</th>\n",
       "      <th>PBSC_C_1</th>\n",
       "      <th>PBSC_G_2</th>\n",
       "      <th>PBSC_A_2</th>\n",
       "      <th>PBSC_T_2</th>\n",
       "      <th>PBSC_C_2</th>\n",
       "      <th>PBSC_G_3</th>\n",
       "      <th>PBSC_A_3</th>\n",
       "      <th>PBSC_T_3</th>\n",
       "      <th>PBSC_C_3</th>\n",
       "      <th>PBSC_G_4</th>\n",
       "      <th>PBSC_A_4</th>\n",
       "      <th>PBSC_T_4</th>\n",
       "      <th>PBSC_C_4</th>\n",
       "      <th>PBSC_G_5</th>\n",
       "      <th>PBSC_A_5</th>\n",
       "      <th>PBSC_T_5</th>\n",
       "      <th>PBSC_C_5</th>\n",
       "      <th>PBSC_G_6</th>\n",
       "      <th>PBSC_A_6</th>\n",
       "      <th>PBSC_T_6</th>\n",
       "      <th>PBSC_C_6</th>\n",
       "      <th>PBSC_G_7</th>\n",
       "      <th>PBSC_A_7</th>\n",
       "      <th>PBSC_T_7</th>\n",
       "      <th>PBSC_C_7</th>\n",
       "      <th>PBSC_G_8</th>\n",
       "      <th>PBSC_A_8</th>\n",
       "      <th>PBSC_T_8</th>\n",
       "      <th>PBSC_C_8</th>\n",
       "      <th>PBSC_G_9</th>\n",
       "      <th>PBSC_A_9</th>\n",
       "      <th>PBSC_T_9</th>\n",
       "      <th>PBSC_C_9</th>\n",
       "      <th>PBSC_G_10</th>\n",
       "      <th>PBSC_A_10</th>\n",
       "      <th>PBSC_T_10</th>\n",
       "      <th>PBSC_C_10</th>\n",
       "      <th>PerSeqGCcont</th>\n",
       "      <th>PSGCC_0</th>\n",
       "      <th>PSGCC_1</th>\n",
       "      <th>PSGCC_2</th>\n",
       "      <th>PSGCC_3</th>\n",
       "      <th>PSGCC_4</th>\n",
       "      <th>PSGCC_5</th>\n",
       "      <th>PSGCC_6</th>\n",
       "      <th>PSGCC_7</th>\n",
       "      <th>PSGCC_8</th>\n",
       "      <th>PSGCC_9</th>\n",
       "      <th>PSGCC_10</th>\n",
       "      <th>PerBaseNcont</th>\n",
       "      <th>PBNC_1</th>\n",
       "      <th>PBNC_2</th>\n",
       "      <th>PBNC_3</th>\n",
       "      <th>PBNC_4</th>\n",
       "      <th>PBNC_5</th>\n",
       "      <th>PBNC_6</th>\n",
       "      <th>PBNC_7</th>\n",
       "      <th>PBNC_8</th>\n",
       "      <th>PBNC_9</th>\n",
       "      <th>PBNC_10</th>\n",
       "      <th>SeqLenDist</th>\n",
       "      <th>SLD_modeVal</th>\n",
       "      <th>SLD_neg10</th>\n",
       "      <th>SLD_neg9</th>\n",
       "      <th>SLD_neg8</th>\n",
       "      <th>SLD_neg7</th>\n",
       "      <th>SLD_neg6</th>\n",
       "      <th>SLD_neg5</th>\n",
       "      <th>SLD_neg4</th>\n",
       "      <th>SLD_neg3</th>\n",
       "      <th>SLD_neg2</th>\n",
       "      <th>SLD_neg1</th>\n",
       "      <th>SLD_mode</th>\n",
       "      <th>SLD_pos1</th>\n",
       "      <th>SLD_pos2</th>\n",
       "      <th>SLD_pos3</th>\n",
       "      <th>SLD_pos4</th>\n",
       "      <th>SLD_pos5</th>\n",
       "      <th>SLD_pos6</th>\n",
       "      <th>SLD_pos7</th>\n",
       "      <th>SLD_pos8</th>\n",
       "      <th>SLD_pos9</th>\n",
       "      <th>SLD_pos10</th>\n",
       "      <th>SeqDupLev</th>\n",
       "      <th>Total_Dedup_Pct</th>\n",
       "      <th>Dedup_relPct_1</th>\n",
       "      <th>Dedup_relPct_2</th>\n",
       "      <th>Dedup_relPct_3</th>\n",
       "      <th>Dedup_relPct_4</th>\n",
       "      <th>Dedup_relPct_5</th>\n",
       "      <th>Dedup_relPct_6</th>\n",
       "      <th>Dedup_relPct_7</th>\n",
       "      <th>Dedup_relPct_8</th>\n",
       "      <th>Dedup_relPct_9</th>\n",
       "      <th>Dedup_relPct_ge10</th>\n",
       "      <th>Dedup_relPct_ge50</th>\n",
       "      <th>Dedup_relPct_ge100</th>\n",
       "      <th>Dedup_relPct_ge500</th>\n",
       "      <th>Dedup_relPct_ge1k</th>\n",
       "      <th>Dedup_relPct_ge5k</th>\n",
       "      <th>Dedup_relPct_ge10k</th>\n",
       "      <th>OverrepSeq</th>\n",
       "      <th>OverRepSeq_1</th>\n",
       "      <th>OverRepSeq_1_Pct</th>\n",
       "      <th>OverRepSeq_2</th>\n",
       "      <th>OverRepSeq_2_Pct</th>\n",
       "      <th>OverRepSeq_3</th>\n",
       "      <th>OverRepSeq_3_Pct</th>\n",
       "      <th>AdaptCont</th>\n",
       "      <th>KmerCont</th>\n",
       "      <th>Kmer_Seq_1</th>\n",
       "      <th>Kmer_Ratio_1</th>\n",
       "      <th>Kmer_Seq_2</th>\n",
       "      <th>Kmer_Ratio_2</th>\n",
       "      <th>Kmer_Seq_3</th>\n",
       "      <th>Kmer_Ratio_3</th>\n",
       "      <th>Kmer_Ratio_Min</th>\n",
       "      <th>Kmer_Num_Seq</th>\n",
       "      <th>study</th>\n",
       "      <th>barcode</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_name</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>sample_type_name</th>\n",
       "      <th>analyte_type</th>\n",
       "      <th>library_type</th>\n",
       "      <th>center</th>\n",
       "      <th>center_name</th>\n",
       "      <th>platform</th>\n",
       "      <th>platform_name</th>\n",
       "      <th>assembly</th>\n",
       "      <th>filename</th>\n",
       "      <th>files_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>analysis_id</th>\n",
       "      <th>aliquot_id</th>\n",
       "      <th>participant_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>published</th>\n",
       "      <th>uploaded</th>\n",
       "      <th>modified</th>\n",
       "      <th>state</th>\n",
       "      <th>sample_type_code</th>\n",
       "      <th>analyte_type_code</th>\n",
       "      <th>platform_full_name</th>\n",
       "      <th>file_type</th>\n",
       "      <th>reason</th>\n",
       "      <th>reagent_vendor</th>\n",
       "      <th>reagent_name</th>\n",
       "      <th>catalog_number</th>\n",
       "      <th>is_custom</th>\n",
       "      <th>target_file_url</th>\n",
       "      <th>probe_file_url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3462</th>\n",
       "      <td>261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13</td>\n",
       "      <td>61YETAAXX_1_ATCACG_2.fastq.gz</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>pass</td>\n",
       "      <td>83121778</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>50</td>\n",
       "      <td>45</td>\n",
       "      <td>47</td>\n",
       "      <td>pass</td>\n",
       "      <td>30</td>\n",
       "      <td>38</td>\n",
       "      <td>30.0</td>\n",
       "      <td>38</td>\n",
       "      <td>15.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>35</td>\n",
       "      <td>36</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>fail</td>\n",
       "      <td>12.495474</td>\n",
       "      <td>63.343577</td>\n",
       "      <td>9.489971</td>\n",
       "      <td>14.670979</td>\n",
       "      <td>60.915261</td>\n",
       "      <td>14.225736</td>\n",
       "      <td>13.044784</td>\n",
       "      <td>11.814219</td>\n",
       "      <td>24.198353</td>\n",
       "      <td>27.006609</td>\n",
       "      <td>25.681176</td>\n",
       "      <td>23.113862</td>\n",
       "      <td>23.447029</td>\n",
       "      <td>26.562247</td>\n",
       "      <td>26.857087</td>\n",
       "      <td>23.133636</td>\n",
       "      <td>23.456119</td>\n",
       "      <td>26.401801</td>\n",
       "      <td>26.787168</td>\n",
       "      <td>23.354911</td>\n",
       "      <td>23.428193</td>\n",
       "      <td>26.60321</td>\n",
       "      <td>26.603931</td>\n",
       "      <td>23.364666</td>\n",
       "      <td>23.439432</td>\n",
       "      <td>26.460188</td>\n",
       "      <td>26.654933</td>\n",
       "      <td>23.445447</td>\n",
       "      <td>23.526332</td>\n",
       "      <td>26.566157</td>\n",
       "      <td>26.542675</td>\n",
       "      <td>23.364837</td>\n",
       "      <td>23.523397</td>\n",
       "      <td>26.445991</td>\n",
       "      <td>26.512676</td>\n",
       "      <td>23.517937</td>\n",
       "      <td>23.541807</td>\n",
       "      <td>26.330577</td>\n",
       "      <td>26.670436</td>\n",
       "      <td>23.457181</td>\n",
       "      <td>23.558229</td>\n",
       "      <td>26.44273</td>\n",
       "      <td>26.512136</td>\n",
       "      <td>23.486905</td>\n",
       "      <td>fail</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>44</td>\n",
       "      <td>46</td>\n",
       "      <td>48</td>\n",
       "      <td>50</td>\n",
       "      <td>51</td>\n",
       "      <td>53</td>\n",
       "      <td>55</td>\n",
       "      <td>57</td>\n",
       "      <td>100</td>\n",
       "      <td>pass</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>warn</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>fail</td>\n",
       "      <td>0.410745</td>\n",
       "      <td>0.340027</td>\n",
       "      <td>0.054663</td>\n",
       "      <td>0.023789</td>\n",
       "      <td>0.015458</td>\n",
       "      <td>0.011768</td>\n",
       "      <td>0.01035</td>\n",
       "      <td>0.010206</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.008803</td>\n",
       "      <td>0.270195</td>\n",
       "      <td>0.342614</td>\n",
       "      <td>1.343957</td>\n",
       "      <td>0.395589</td>\n",
       "      <td>0.607665</td>\n",
       "      <td>0.342961</td>\n",
       "      <td>96.212956</td>\n",
       "      <td>fail</td>\n",
       "      <td>ATCACG</td>\n",
       "      <td>47.41739</td>\n",
       "      <td>.TCACG</td>\n",
       "      <td>0.441099</td>\n",
       "      <td>ATC.CG</td>\n",
       "      <td>0.421891</td>\n",
       "      <td>pass</td>\n",
       "      <td>warn</td>\n",
       "      <td>CG.GATT</td>\n",
       "      <td>44.000004</td>\n",
       "      <td>AC.GCGC</td>\n",
       "      <td>44.000004</td>\n",
       "      <td>ACACG.C</td>\n",
       "      <td>44.000004</td>\n",
       "      <td>44.0</td>\n",
       "      <td>21</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13</td>\n",
       "      <td>LAML</td>\n",
       "      <td>ACUTE MYELOID LEUKEMIA</td>\n",
       "      <td>TB</td>\n",
       "      <td>Primary Blood Derived Cancer - Peripheral Blood</td>\n",
       "      <td>Total RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13_rnaseq_fastq.tar</td>\n",
       "      <td>6.494423e+09</td>\n",
       "      <td>6924e520152b16158c3ea24ef20550af</td>\n",
       "      <td>261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990</td>\n",
       "      <td>77952b61-6b89-4285-9443-a41246a7d8f2</td>\n",
       "      <td>ab60901b-d2ad-401a-9742-52df35ce1de0</td>\n",
       "      <td>40168b34-bbfe-4ee2-b7b5-2e33fbf64a18</td>\n",
       "      <td>AB</td>\n",
       "      <td>SRS131131</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>Live</td>\n",
       "      <td>03</td>\n",
       "      <td>T</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>fasta</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          CGHub_analysisID                AliquotBarcode  \\\n",
       "3462  261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990  TCGA-AB-2909-03A-01T-0744-13   \n",
       "\n",
       "                      fastqFilename  \\\n",
       "3462  61YETAAXX_1_ATCACG_2.fastq.gz   \n",
       "\n",
       "                                        FastQC_html_url  \\\n",
       "3462  https://storage.cloud.google.com/isb-cgc-open/...   \n",
       "\n",
       "                                         FastQC_zip_url BasicStats  \\\n",
       "3462  https://storage.cloud.google.com/isb-cgc-open/...       pass   \n",
       "\n",
       "      numTotalSeq  numPoorSeq  minLength  maxLength  delLength  pctGC  \\\n",
       "3462     83121778           0          6         50         45     47   \n",
       "\n",
       "     PerBaseSeqQual  PBSQ10_0  PBSQ50_0  PBSQ10_1  PBSQ50_1  PBSQ10_2  \\\n",
       "3462           pass        30        38      30.0        38      15.0   \n",
       "\n",
       "      PBSQ50_2  PBSQ10_3  PBSQ50_3  PBSQ10_4  PBSQ50_4  PBSQ10_5  PBSQ50_5  \\\n",
       "3462      38.0      12.0      37.0       9.0      37.0       2.0      37.0   \n",
       "\n",
       "      PBSQ10_6  PBSQ50_6  PBSQ10_7  PBSQ50_7  PBSQ10_8  PBSQ50_8  PBSQ10_9  \\\n",
       "3462       2.0      37.0       2.0      37.0       2.0      37.0       2.0   \n",
       "\n",
       "      PBSQ50_9  PBSQ10_10  PBSQ50_10 PerTileSeqQual PerSeqQualScore  PSQC_0  \\\n",
       "3462      36.0          2         36            NaN            pass       2   \n",
       "\n",
       "      PSQC_1  PSQC_2  PSQC_3  PSQC_4  PSQC_5  PSQC_6  PSQC_7  PSQC_8  PSQC_9  \\\n",
       "3462      27      32      35      36      37      37      37      38      38   \n",
       "\n",
       "      PSQC_10 PerBaseSeqCont   PBSC_G_0   PBSC_A_0  PBSC_T_0   PBSC_C_0  \\\n",
       "3462       38           fail  12.495474  63.343577  9.489971  14.670979   \n",
       "\n",
       "       PBSC_G_1   PBSC_A_1   PBSC_T_1   PBSC_C_1   PBSC_G_2   PBSC_A_2  \\\n",
       "3462  60.915261  14.225736  13.044784  11.814219  24.198353  27.006609   \n",
       "\n",
       "       PBSC_T_2   PBSC_C_2   PBSC_G_3   PBSC_A_3   PBSC_T_3   PBSC_C_3  \\\n",
       "3462  25.681176  23.113862  23.447029  26.562247  26.857087  23.133636   \n",
       "\n",
       "       PBSC_G_4   PBSC_A_4   PBSC_T_4   PBSC_C_4   PBSC_G_5  PBSC_A_5  \\\n",
       "3462  23.456119  26.401801  26.787168  23.354911  23.428193  26.60321   \n",
       "\n",
       "       PBSC_T_5   PBSC_C_5   PBSC_G_6   PBSC_A_6   PBSC_T_6   PBSC_C_6  \\\n",
       "3462  26.603931  23.364666  23.439432  26.460188  26.654933  23.445447   \n",
       "\n",
       "       PBSC_G_7   PBSC_A_7   PBSC_T_7   PBSC_C_7   PBSC_G_8   PBSC_A_8  \\\n",
       "3462  23.526332  26.566157  26.542675  23.364837  23.523397  26.445991   \n",
       "\n",
       "       PBSC_T_8   PBSC_C_8   PBSC_G_9   PBSC_A_9   PBSC_T_9   PBSC_C_9  \\\n",
       "3462  26.512676  23.517937  23.541807  26.330577  26.670436  23.457181   \n",
       "\n",
       "      PBSC_G_10  PBSC_A_10  PBSC_T_10  PBSC_C_10 PerSeqGCcont  PSGCC_0  \\\n",
       "3462  23.558229   26.44273  26.512136  23.486905         fail        0   \n",
       "\n",
       "      PSGCC_1  PSGCC_2  PSGCC_3  PSGCC_4  PSGCC_5  PSGCC_6  PSGCC_7  PSGCC_8  \\\n",
       "3462       43       44       46       48       50       51       53       55   \n",
       "\n",
       "      PSGCC_9  PSGCC_10 PerBaseNcont  PBNC_1  PBNC_2  PBNC_3  PBNC_4  PBNC_5  \\\n",
       "3462       57       100         pass     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "      PBNC_6  PBNC_7  PBNC_8  PBNC_9  PBNC_10 SeqLenDist  SLD_modeVal  \\\n",
       "3462     0.0     0.0     0.0     0.0      0.0       warn            6   \n",
       "\n",
       "      SLD_neg10  SLD_neg9  SLD_neg8  SLD_neg7  SLD_neg6  SLD_neg5  SLD_neg4  \\\n",
       "3462        NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "\n",
       "      SLD_neg3  SLD_neg2  SLD_neg1  SLD_mode  SLD_pos1  SLD_pos2  SLD_pos3  \\\n",
       "3462       NaN       NaN       NaN       0.5       0.0       0.0       0.0   \n",
       "\n",
       "      SLD_pos4  SLD_pos5  SLD_pos6  SLD_pos7  SLD_pos8  SLD_pos9  SLD_pos10  \\\n",
       "3462       0.0       0.0       0.0       0.0       0.0       0.0        0.0   \n",
       "\n",
       "     SeqDupLev  Total_Dedup_Pct  Dedup_relPct_1  Dedup_relPct_2  \\\n",
       "3462      fail         0.410745        0.340027        0.054663   \n",
       "\n",
       "      Dedup_relPct_3  Dedup_relPct_4  Dedup_relPct_5  Dedup_relPct_6  \\\n",
       "3462        0.023789        0.015458        0.011768         0.01035   \n",
       "\n",
       "      Dedup_relPct_7  Dedup_relPct_8  Dedup_relPct_9  Dedup_relPct_ge10  \\\n",
       "3462        0.010206           0.009        0.008803           0.270195   \n",
       "\n",
       "      Dedup_relPct_ge50  Dedup_relPct_ge100  Dedup_relPct_ge500  \\\n",
       "3462           0.342614            1.343957            0.395589   \n",
       "\n",
       "      Dedup_relPct_ge1k  Dedup_relPct_ge5k  Dedup_relPct_ge10k OverrepSeq  \\\n",
       "3462           0.607665           0.342961           96.212956       fail   \n",
       "\n",
       "     OverRepSeq_1  OverRepSeq_1_Pct OverRepSeq_2  OverRepSeq_2_Pct  \\\n",
       "3462       ATCACG          47.41739       .TCACG          0.441099   \n",
       "\n",
       "     OverRepSeq_3  OverRepSeq_3_Pct AdaptCont KmerCont Kmer_Seq_1  \\\n",
       "3462       ATC.CG          0.421891      pass     warn    CG.GATT   \n",
       "\n",
       "      Kmer_Ratio_1 Kmer_Seq_2  Kmer_Ratio_2 Kmer_Seq_3  Kmer_Ratio_3  \\\n",
       "3462     44.000004    AC.GCGC     44.000004    ACACG.C     44.000004   \n",
       "\n",
       "      Kmer_Ratio_Min  Kmer_Num_Seq study                       barcode  \\\n",
       "3462            44.0            21  TCGA  TCGA-AB-2909-03A-01T-0744-13   \n",
       "\n",
       "     disease            disease_name sample_type  \\\n",
       "3462    LAML  ACUTE MYELOID LEUKEMIA          TB   \n",
       "\n",
       "                                     sample_type_name analyte_type  \\\n",
       "3462  Primary Blood Derived Cancer - Peripheral Blood    Total RNA   \n",
       "\n",
       "     library_type   center  center_name  platform platform_name   assembly  \\\n",
       "3462      RNA-Seq  BCCAGSC          NaN  ILLUMINA      Illumina  unaligned   \n",
       "\n",
       "                                           filename    files_size  \\\n",
       "3462  TCGA-AB-2909-03A-01T-0744-13_rnaseq_fastq.tar  6.494423e+09   \n",
       "\n",
       "                              checksum                           analysis_id  \\\n",
       "3462  6924e520152b16158c3ea24ef20550af  261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990   \n",
       "\n",
       "                                aliquot_id  \\\n",
       "3462  77952b61-6b89-4285-9443-a41246a7d8f2   \n",
       "\n",
       "                            participant_id  \\\n",
       "3462  ab60901b-d2ad-401a-9742-52df35ce1de0   \n",
       "\n",
       "                                 sample_id tss_id sample_accession  \\\n",
       "3462  40168b34-bbfe-4ee2-b7b5-2e33fbf64a18     AB        SRS131131   \n",
       "\n",
       "       published    uploaded    modified state sample_type_code  \\\n",
       "3462  2014-05-13  2014-05-13  2014-05-13  Live               03   \n",
       "\n",
       "     analyte_type_code            platform_full_name file_type reason  \\\n",
       "3462                 T  Illumina Genome Analyzer IIx     fasta    NaN   \n",
       "\n",
       "     reagent_vendor reagent_name catalog_number is_custom target_file_url  \\\n",
       "3462            NaN          NaN            NaN       NaN             NaN   \n",
       "\n",
       "     probe_file_url  \n",
       "3462            NaN  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf[ndf.delLength == 45]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://storage.cloud.google.com/isb-cgc-open/tcga-qc/LAML/Total_RNA/RNA-Seq/BCCAGSC/ILLUMINA/261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990-61YETAAXX_1_ATCACG_2_fastqc.html\n"
     ]
    }
   ],
   "source": [
    "print(ndf[ndf.delLength == 45].FastQC_html_url.values[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Maybe drop these recods with analysis_id 261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CGHub_analysisID</th>\n",
       "      <th>AliquotBarcode</th>\n",
       "      <th>fastqFilename</th>\n",
       "      <th>FastQC_html_url</th>\n",
       "      <th>FastQC_zip_url</th>\n",
       "      <th>BasicStats</th>\n",
       "      <th>numTotalSeq</th>\n",
       "      <th>numPoorSeq</th>\n",
       "      <th>minLength</th>\n",
       "      <th>maxLength</th>\n",
       "      <th>delLength</th>\n",
       "      <th>pctGC</th>\n",
       "      <th>PerBaseSeqQual</th>\n",
       "      <th>PBSQ10_0</th>\n",
       "      <th>PBSQ50_0</th>\n",
       "      <th>PBSQ10_1</th>\n",
       "      <th>PBSQ50_1</th>\n",
       "      <th>PBSQ10_2</th>\n",
       "      <th>PBSQ50_2</th>\n",
       "      <th>PBSQ10_3</th>\n",
       "      <th>PBSQ50_3</th>\n",
       "      <th>PBSQ10_4</th>\n",
       "      <th>PBSQ50_4</th>\n",
       "      <th>PBSQ10_5</th>\n",
       "      <th>PBSQ50_5</th>\n",
       "      <th>PBSQ10_6</th>\n",
       "      <th>PBSQ50_6</th>\n",
       "      <th>PBSQ10_7</th>\n",
       "      <th>PBSQ50_7</th>\n",
       "      <th>PBSQ10_8</th>\n",
       "      <th>PBSQ50_8</th>\n",
       "      <th>PBSQ10_9</th>\n",
       "      <th>PBSQ50_9</th>\n",
       "      <th>PBSQ10_10</th>\n",
       "      <th>PBSQ50_10</th>\n",
       "      <th>PerTileSeqQual</th>\n",
       "      <th>PerSeqQualScore</th>\n",
       "      <th>PSQC_0</th>\n",
       "      <th>PSQC_1</th>\n",
       "      <th>PSQC_2</th>\n",
       "      <th>PSQC_3</th>\n",
       "      <th>PSQC_4</th>\n",
       "      <th>PSQC_5</th>\n",
       "      <th>PSQC_6</th>\n",
       "      <th>PSQC_7</th>\n",
       "      <th>PSQC_8</th>\n",
       "      <th>PSQC_9</th>\n",
       "      <th>PSQC_10</th>\n",
       "      <th>PerBaseSeqCont</th>\n",
       "      <th>PBSC_G_0</th>\n",
       "      <th>PBSC_A_0</th>\n",
       "      <th>PBSC_T_0</th>\n",
       "      <th>PBSC_C_0</th>\n",
       "      <th>PBSC_G_1</th>\n",
       "      <th>PBSC_A_1</th>\n",
       "      <th>PBSC_T_1</th>\n",
       "      <th>PBSC_C_1</th>\n",
       "      <th>PBSC_G_2</th>\n",
       "      <th>PBSC_A_2</th>\n",
       "      <th>PBSC_T_2</th>\n",
       "      <th>PBSC_C_2</th>\n",
       "      <th>PBSC_G_3</th>\n",
       "      <th>PBSC_A_3</th>\n",
       "      <th>PBSC_T_3</th>\n",
       "      <th>PBSC_C_3</th>\n",
       "      <th>PBSC_G_4</th>\n",
       "      <th>PBSC_A_4</th>\n",
       "      <th>PBSC_T_4</th>\n",
       "      <th>PBSC_C_4</th>\n",
       "      <th>PBSC_G_5</th>\n",
       "      <th>PBSC_A_5</th>\n",
       "      <th>PBSC_T_5</th>\n",
       "      <th>PBSC_C_5</th>\n",
       "      <th>PBSC_G_6</th>\n",
       "      <th>PBSC_A_6</th>\n",
       "      <th>PBSC_T_6</th>\n",
       "      <th>PBSC_C_6</th>\n",
       "      <th>PBSC_G_7</th>\n",
       "      <th>PBSC_A_7</th>\n",
       "      <th>PBSC_T_7</th>\n",
       "      <th>PBSC_C_7</th>\n",
       "      <th>PBSC_G_8</th>\n",
       "      <th>PBSC_A_8</th>\n",
       "      <th>PBSC_T_8</th>\n",
       "      <th>PBSC_C_8</th>\n",
       "      <th>PBSC_G_9</th>\n",
       "      <th>PBSC_A_9</th>\n",
       "      <th>PBSC_T_9</th>\n",
       "      <th>PBSC_C_9</th>\n",
       "      <th>PBSC_G_10</th>\n",
       "      <th>PBSC_A_10</th>\n",
       "      <th>PBSC_T_10</th>\n",
       "      <th>PBSC_C_10</th>\n",
       "      <th>PerSeqGCcont</th>\n",
       "      <th>PSGCC_0</th>\n",
       "      <th>PSGCC_1</th>\n",
       "      <th>PSGCC_2</th>\n",
       "      <th>PSGCC_3</th>\n",
       "      <th>PSGCC_4</th>\n",
       "      <th>PSGCC_5</th>\n",
       "      <th>PSGCC_6</th>\n",
       "      <th>PSGCC_7</th>\n",
       "      <th>PSGCC_8</th>\n",
       "      <th>PSGCC_9</th>\n",
       "      <th>PSGCC_10</th>\n",
       "      <th>PerBaseNcont</th>\n",
       "      <th>PBNC_1</th>\n",
       "      <th>PBNC_2</th>\n",
       "      <th>PBNC_3</th>\n",
       "      <th>PBNC_4</th>\n",
       "      <th>PBNC_5</th>\n",
       "      <th>PBNC_6</th>\n",
       "      <th>PBNC_7</th>\n",
       "      <th>PBNC_8</th>\n",
       "      <th>PBNC_9</th>\n",
       "      <th>PBNC_10</th>\n",
       "      <th>SeqLenDist</th>\n",
       "      <th>SLD_modeVal</th>\n",
       "      <th>SLD_neg10</th>\n",
       "      <th>SLD_neg9</th>\n",
       "      <th>SLD_neg8</th>\n",
       "      <th>SLD_neg7</th>\n",
       "      <th>SLD_neg6</th>\n",
       "      <th>SLD_neg5</th>\n",
       "      <th>SLD_neg4</th>\n",
       "      <th>SLD_neg3</th>\n",
       "      <th>SLD_neg2</th>\n",
       "      <th>SLD_neg1</th>\n",
       "      <th>SLD_mode</th>\n",
       "      <th>SLD_pos1</th>\n",
       "      <th>SLD_pos2</th>\n",
       "      <th>SLD_pos3</th>\n",
       "      <th>SLD_pos4</th>\n",
       "      <th>SLD_pos5</th>\n",
       "      <th>SLD_pos6</th>\n",
       "      <th>SLD_pos7</th>\n",
       "      <th>SLD_pos8</th>\n",
       "      <th>SLD_pos9</th>\n",
       "      <th>SLD_pos10</th>\n",
       "      <th>SeqDupLev</th>\n",
       "      <th>Total_Dedup_Pct</th>\n",
       "      <th>Dedup_relPct_1</th>\n",
       "      <th>Dedup_relPct_2</th>\n",
       "      <th>Dedup_relPct_3</th>\n",
       "      <th>Dedup_relPct_4</th>\n",
       "      <th>Dedup_relPct_5</th>\n",
       "      <th>Dedup_relPct_6</th>\n",
       "      <th>Dedup_relPct_7</th>\n",
       "      <th>Dedup_relPct_8</th>\n",
       "      <th>Dedup_relPct_9</th>\n",
       "      <th>Dedup_relPct_ge10</th>\n",
       "      <th>Dedup_relPct_ge50</th>\n",
       "      <th>Dedup_relPct_ge100</th>\n",
       "      <th>Dedup_relPct_ge500</th>\n",
       "      <th>Dedup_relPct_ge1k</th>\n",
       "      <th>Dedup_relPct_ge5k</th>\n",
       "      <th>Dedup_relPct_ge10k</th>\n",
       "      <th>OverrepSeq</th>\n",
       "      <th>OverRepSeq_1</th>\n",
       "      <th>OverRepSeq_1_Pct</th>\n",
       "      <th>OverRepSeq_2</th>\n",
       "      <th>OverRepSeq_2_Pct</th>\n",
       "      <th>OverRepSeq_3</th>\n",
       "      <th>OverRepSeq_3_Pct</th>\n",
       "      <th>AdaptCont</th>\n",
       "      <th>KmerCont</th>\n",
       "      <th>Kmer_Seq_1</th>\n",
       "      <th>Kmer_Ratio_1</th>\n",
       "      <th>Kmer_Seq_2</th>\n",
       "      <th>Kmer_Ratio_2</th>\n",
       "      <th>Kmer_Seq_3</th>\n",
       "      <th>Kmer_Ratio_3</th>\n",
       "      <th>Kmer_Ratio_Min</th>\n",
       "      <th>Kmer_Num_Seq</th>\n",
       "      <th>study</th>\n",
       "      <th>barcode</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_name</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>sample_type_name</th>\n",
       "      <th>analyte_type</th>\n",
       "      <th>library_type</th>\n",
       "      <th>center</th>\n",
       "      <th>center_name</th>\n",
       "      <th>platform</th>\n",
       "      <th>platform_name</th>\n",
       "      <th>assembly</th>\n",
       "      <th>filename</th>\n",
       "      <th>files_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>analysis_id</th>\n",
       "      <th>aliquot_id</th>\n",
       "      <th>participant_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>published</th>\n",
       "      <th>uploaded</th>\n",
       "      <th>modified</th>\n",
       "      <th>state</th>\n",
       "      <th>sample_type_code</th>\n",
       "      <th>analyte_type_code</th>\n",
       "      <th>platform_full_name</th>\n",
       "      <th>file_type</th>\n",
       "      <th>reason</th>\n",
       "      <th>reagent_vendor</th>\n",
       "      <th>reagent_name</th>\n",
       "      <th>catalog_number</th>\n",
       "      <th>is_custom</th>\n",
       "      <th>target_file_url</th>\n",
       "      <th>probe_file_url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3461</th>\n",
       "      <td>261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13</td>\n",
       "      <td>61YETAAXX_1_ATCACG_1.fastq.gz</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>pass</td>\n",
       "      <td>41560889</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>47</td>\n",
       "      <td>pass</td>\n",
       "      <td>25</td>\n",
       "      <td>38</td>\n",
       "      <td>17.0</td>\n",
       "      <td>38</td>\n",
       "      <td>16.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>30</td>\n",
       "      <td>34</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>pass</td>\n",
       "      <td>27.231418</td>\n",
       "      <td>28.232857</td>\n",
       "      <td>17.734709</td>\n",
       "      <td>26.801016</td>\n",
       "      <td>23.878398</td>\n",
       "      <td>27.687630</td>\n",
       "      <td>25.841315</td>\n",
       "      <td>22.592656</td>\n",
       "      <td>24.753503</td>\n",
       "      <td>27.022491</td>\n",
       "      <td>25.430414</td>\n",
       "      <td>22.793592</td>\n",
       "      <td>23.867636</td>\n",
       "      <td>26.576534</td>\n",
       "      <td>26.730451</td>\n",
       "      <td>22.825379</td>\n",
       "      <td>23.769961</td>\n",
       "      <td>26.504633</td>\n",
       "      <td>26.626973</td>\n",
       "      <td>23.098432</td>\n",
       "      <td>23.905519</td>\n",
       "      <td>26.758423</td>\n",
       "      <td>26.348721</td>\n",
       "      <td>22.987337</td>\n",
       "      <td>23.790913</td>\n",
       "      <td>26.465633</td>\n",
       "      <td>26.532799</td>\n",
       "      <td>23.210656</td>\n",
       "      <td>23.805517</td>\n",
       "      <td>26.526281</td>\n",
       "      <td>26.465826</td>\n",
       "      <td>23.202376</td>\n",
       "      <td>23.770514</td>\n",
       "      <td>26.621140</td>\n",
       "      <td>26.395636</td>\n",
       "      <td>23.212710</td>\n",
       "      <td>23.661634</td>\n",
       "      <td>26.474574</td>\n",
       "      <td>26.580701</td>\n",
       "      <td>23.283091</td>\n",
       "      <td>23.731163</td>\n",
       "      <td>26.610283</td>\n",
       "      <td>26.462177</td>\n",
       "      <td>23.196377</td>\n",
       "      <td>pass</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>36</td>\n",
       "      <td>40</td>\n",
       "      <td>43</td>\n",
       "      <td>47</td>\n",
       "      <td>50</td>\n",
       "      <td>54</td>\n",
       "      <td>57</td>\n",
       "      <td>63</td>\n",
       "      <td>100</td>\n",
       "      <td>pass</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>pass</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>80.327895</td>\n",
       "      <td>73.951186</td>\n",
       "      <td>8.739390</td>\n",
       "      <td>2.841169</td>\n",
       "      <td>1.346374</td>\n",
       "      <td>0.923370</td>\n",
       "      <td>0.634790</td>\n",
       "      <td>0.504889</td>\n",
       "      <td>0.401431</td>\n",
       "      <td>0.351309</td>\n",
       "      <td>4.226179</td>\n",
       "      <td>1.895547</td>\n",
       "      <td>3.199739</td>\n",
       "      <td>0.360195</td>\n",
       "      <td>0.298664</td>\n",
       "      <td>0.104290</td>\n",
       "      <td>0.221475</td>\n",
       "      <td>warn</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>0.126886</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>warn</td>\n",
       "      <td>CTAT.TC</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>A.TTCGA</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>ACCCC.C</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>44.0</td>\n",
       "      <td>21</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13</td>\n",
       "      <td>LAML</td>\n",
       "      <td>ACUTE MYELOID LEUKEMIA</td>\n",
       "      <td>TB</td>\n",
       "      <td>Primary Blood Derived Cancer - Peripheral Blood</td>\n",
       "      <td>Total RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13_rnaseq_fastq.tar</td>\n",
       "      <td>6.494423e+09</td>\n",
       "      <td>6924e520152b16158c3ea24ef20550af</td>\n",
       "      <td>261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990</td>\n",
       "      <td>77952b61-6b89-4285-9443-a41246a7d8f2</td>\n",
       "      <td>ab60901b-d2ad-401a-9742-52df35ce1de0</td>\n",
       "      <td>40168b34-bbfe-4ee2-b7b5-2e33fbf64a18</td>\n",
       "      <td>AB</td>\n",
       "      <td>SRS131131</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>Live</td>\n",
       "      <td>03</td>\n",
       "      <td>T</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>fasta</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3462</th>\n",
       "      <td>261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13</td>\n",
       "      <td>61YETAAXX_1_ATCACG_2.fastq.gz</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>pass</td>\n",
       "      <td>83121778</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>50</td>\n",
       "      <td>45</td>\n",
       "      <td>47</td>\n",
       "      <td>pass</td>\n",
       "      <td>30</td>\n",
       "      <td>38</td>\n",
       "      <td>30.0</td>\n",
       "      <td>38</td>\n",
       "      <td>15.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>35</td>\n",
       "      <td>36</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>fail</td>\n",
       "      <td>12.495474</td>\n",
       "      <td>63.343577</td>\n",
       "      <td>9.489971</td>\n",
       "      <td>14.670979</td>\n",
       "      <td>60.915261</td>\n",
       "      <td>14.225736</td>\n",
       "      <td>13.044784</td>\n",
       "      <td>11.814219</td>\n",
       "      <td>24.198353</td>\n",
       "      <td>27.006609</td>\n",
       "      <td>25.681176</td>\n",
       "      <td>23.113862</td>\n",
       "      <td>23.447029</td>\n",
       "      <td>26.562247</td>\n",
       "      <td>26.857087</td>\n",
       "      <td>23.133636</td>\n",
       "      <td>23.456119</td>\n",
       "      <td>26.401801</td>\n",
       "      <td>26.787168</td>\n",
       "      <td>23.354911</td>\n",
       "      <td>23.428193</td>\n",
       "      <td>26.603210</td>\n",
       "      <td>26.603931</td>\n",
       "      <td>23.364666</td>\n",
       "      <td>23.439432</td>\n",
       "      <td>26.460188</td>\n",
       "      <td>26.654933</td>\n",
       "      <td>23.445447</td>\n",
       "      <td>23.526332</td>\n",
       "      <td>26.566157</td>\n",
       "      <td>26.542675</td>\n",
       "      <td>23.364837</td>\n",
       "      <td>23.523397</td>\n",
       "      <td>26.445991</td>\n",
       "      <td>26.512676</td>\n",
       "      <td>23.517937</td>\n",
       "      <td>23.541807</td>\n",
       "      <td>26.330577</td>\n",
       "      <td>26.670436</td>\n",
       "      <td>23.457181</td>\n",
       "      <td>23.558229</td>\n",
       "      <td>26.442730</td>\n",
       "      <td>26.512136</td>\n",
       "      <td>23.486905</td>\n",
       "      <td>fail</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>44</td>\n",
       "      <td>46</td>\n",
       "      <td>48</td>\n",
       "      <td>50</td>\n",
       "      <td>51</td>\n",
       "      <td>53</td>\n",
       "      <td>55</td>\n",
       "      <td>57</td>\n",
       "      <td>100</td>\n",
       "      <td>pass</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>warn</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>fail</td>\n",
       "      <td>0.410745</td>\n",
       "      <td>0.340027</td>\n",
       "      <td>0.054663</td>\n",
       "      <td>0.023789</td>\n",
       "      <td>0.015458</td>\n",
       "      <td>0.011768</td>\n",
       "      <td>0.010350</td>\n",
       "      <td>0.010206</td>\n",
       "      <td>0.009000</td>\n",
       "      <td>0.008803</td>\n",
       "      <td>0.270195</td>\n",
       "      <td>0.342614</td>\n",
       "      <td>1.343957</td>\n",
       "      <td>0.395589</td>\n",
       "      <td>0.607665</td>\n",
       "      <td>0.342961</td>\n",
       "      <td>96.212956</td>\n",
       "      <td>fail</td>\n",
       "      <td>ATCACG</td>\n",
       "      <td>47.417390</td>\n",
       "      <td>.TCACG</td>\n",
       "      <td>0.441099</td>\n",
       "      <td>ATC.CG</td>\n",
       "      <td>0.421891</td>\n",
       "      <td>pass</td>\n",
       "      <td>warn</td>\n",
       "      <td>CG.GATT</td>\n",
       "      <td>44.000004</td>\n",
       "      <td>AC.GCGC</td>\n",
       "      <td>44.000004</td>\n",
       "      <td>ACACG.C</td>\n",
       "      <td>44.000004</td>\n",
       "      <td>44.0</td>\n",
       "      <td>21</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13</td>\n",
       "      <td>LAML</td>\n",
       "      <td>ACUTE MYELOID LEUKEMIA</td>\n",
       "      <td>TB</td>\n",
       "      <td>Primary Blood Derived Cancer - Peripheral Blood</td>\n",
       "      <td>Total RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13_rnaseq_fastq.tar</td>\n",
       "      <td>6.494423e+09</td>\n",
       "      <td>6924e520152b16158c3ea24ef20550af</td>\n",
       "      <td>261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990</td>\n",
       "      <td>77952b61-6b89-4285-9443-a41246a7d8f2</td>\n",
       "      <td>ab60901b-d2ad-401a-9742-52df35ce1de0</td>\n",
       "      <td>40168b34-bbfe-4ee2-b7b5-2e33fbf64a18</td>\n",
       "      <td>AB</td>\n",
       "      <td>SRS131131</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>Live</td>\n",
       "      <td>03</td>\n",
       "      <td>T</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>fasta</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3463</th>\n",
       "      <td>261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13</td>\n",
       "      <td>622DDAAXX_1_1.fastq.gz</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>pass</td>\n",
       "      <td>34582299</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>47</td>\n",
       "      <td>pass</td>\n",
       "      <td>30</td>\n",
       "      <td>38</td>\n",
       "      <td>27.0</td>\n",
       "      <td>38</td>\n",
       "      <td>26.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>33</td>\n",
       "      <td>35</td>\n",
       "      <td>36</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>pass</td>\n",
       "      <td>26.712025</td>\n",
       "      <td>27.831353</td>\n",
       "      <td>17.814340</td>\n",
       "      <td>27.642282</td>\n",
       "      <td>23.939254</td>\n",
       "      <td>27.308237</td>\n",
       "      <td>25.576351</td>\n",
       "      <td>23.176158</td>\n",
       "      <td>24.874872</td>\n",
       "      <td>26.735585</td>\n",
       "      <td>25.151433</td>\n",
       "      <td>23.238110</td>\n",
       "      <td>23.873062</td>\n",
       "      <td>26.338144</td>\n",
       "      <td>26.456724</td>\n",
       "      <td>23.332070</td>\n",
       "      <td>23.843169</td>\n",
       "      <td>26.340744</td>\n",
       "      <td>26.295807</td>\n",
       "      <td>23.520281</td>\n",
       "      <td>23.804896</td>\n",
       "      <td>26.553735</td>\n",
       "      <td>26.136664</td>\n",
       "      <td>23.504705</td>\n",
       "      <td>23.887982</td>\n",
       "      <td>26.220423</td>\n",
       "      <td>26.206203</td>\n",
       "      <td>23.685392</td>\n",
       "      <td>23.898009</td>\n",
       "      <td>26.313621</td>\n",
       "      <td>26.092323</td>\n",
       "      <td>23.696048</td>\n",
       "      <td>23.807512</td>\n",
       "      <td>26.510711</td>\n",
       "      <td>26.000879</td>\n",
       "      <td>23.680898</td>\n",
       "      <td>23.717625</td>\n",
       "      <td>26.285280</td>\n",
       "      <td>26.239225</td>\n",
       "      <td>23.757871</td>\n",
       "      <td>23.753413</td>\n",
       "      <td>26.507991</td>\n",
       "      <td>26.058258</td>\n",
       "      <td>23.680338</td>\n",
       "      <td>pass</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>37</td>\n",
       "      <td>40</td>\n",
       "      <td>44</td>\n",
       "      <td>47</td>\n",
       "      <td>51</td>\n",
       "      <td>54</td>\n",
       "      <td>58</td>\n",
       "      <td>63</td>\n",
       "      <td>100</td>\n",
       "      <td>pass</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>pass</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>79.626184</td>\n",
       "      <td>73.229846</td>\n",
       "      <td>8.682177</td>\n",
       "      <td>2.826697</td>\n",
       "      <td>1.434117</td>\n",
       "      <td>0.917460</td>\n",
       "      <td>0.668410</td>\n",
       "      <td>0.559806</td>\n",
       "      <td>0.387599</td>\n",
       "      <td>0.392132</td>\n",
       "      <td>4.518259</td>\n",
       "      <td>2.061797</td>\n",
       "      <td>3.138964</td>\n",
       "      <td>0.313371</td>\n",
       "      <td>0.287809</td>\n",
       "      <td>0.191097</td>\n",
       "      <td>0.390460</td>\n",
       "      <td>warn</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>0.232136</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>TGGC.TG</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>ACT.CCT</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>ATCCTC.</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>44.0</td>\n",
       "      <td>21</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13</td>\n",
       "      <td>LAML</td>\n",
       "      <td>ACUTE MYELOID LEUKEMIA</td>\n",
       "      <td>TB</td>\n",
       "      <td>Primary Blood Derived Cancer - Peripheral Blood</td>\n",
       "      <td>Total RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13_rnaseq_fastq.tar</td>\n",
       "      <td>6.494423e+09</td>\n",
       "      <td>6924e520152b16158c3ea24ef20550af</td>\n",
       "      <td>261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990</td>\n",
       "      <td>77952b61-6b89-4285-9443-a41246a7d8f2</td>\n",
       "      <td>ab60901b-d2ad-401a-9742-52df35ce1de0</td>\n",
       "      <td>40168b34-bbfe-4ee2-b7b5-2e33fbf64a18</td>\n",
       "      <td>AB</td>\n",
       "      <td>SRS131131</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>Live</td>\n",
       "      <td>03</td>\n",
       "      <td>T</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>fasta</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3464</th>\n",
       "      <td>261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13</td>\n",
       "      <td>622DDAAXX_1_2.fastq.gz</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>https://storage.cloud.google.com/isb-cgc-open/...</td>\n",
       "      <td>pass</td>\n",
       "      <td>34582299</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>47</td>\n",
       "      <td>pass</td>\n",
       "      <td>28</td>\n",
       "      <td>38</td>\n",
       "      <td>25.0</td>\n",
       "      <td>38</td>\n",
       "      <td>25.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>32</td>\n",
       "      <td>34</td>\n",
       "      <td>35</td>\n",
       "      <td>36</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>pass</td>\n",
       "      <td>25.179036</td>\n",
       "      <td>27.663989</td>\n",
       "      <td>18.347905</td>\n",
       "      <td>28.809070</td>\n",
       "      <td>23.644135</td>\n",
       "      <td>27.163041</td>\n",
       "      <td>25.553711</td>\n",
       "      <td>23.639114</td>\n",
       "      <td>24.500060</td>\n",
       "      <td>26.758607</td>\n",
       "      <td>25.328191</td>\n",
       "      <td>23.413142</td>\n",
       "      <td>23.653301</td>\n",
       "      <td>26.365013</td>\n",
       "      <td>26.522033</td>\n",
       "      <td>23.459653</td>\n",
       "      <td>23.694560</td>\n",
       "      <td>26.204949</td>\n",
       "      <td>26.415956</td>\n",
       "      <td>23.684535</td>\n",
       "      <td>23.644564</td>\n",
       "      <td>26.406414</td>\n",
       "      <td>26.290791</td>\n",
       "      <td>23.658231</td>\n",
       "      <td>23.659179</td>\n",
       "      <td>26.237727</td>\n",
       "      <td>26.272652</td>\n",
       "      <td>23.830443</td>\n",
       "      <td>23.718234</td>\n",
       "      <td>26.384177</td>\n",
       "      <td>26.176677</td>\n",
       "      <td>23.720911</td>\n",
       "      <td>23.707734</td>\n",
       "      <td>26.226889</td>\n",
       "      <td>26.128568</td>\n",
       "      <td>23.936809</td>\n",
       "      <td>23.673562</td>\n",
       "      <td>26.092290</td>\n",
       "      <td>26.303848</td>\n",
       "      <td>23.930300</td>\n",
       "      <td>23.759317</td>\n",
       "      <td>26.301108</td>\n",
       "      <td>26.002894</td>\n",
       "      <td>23.936680</td>\n",
       "      <td>pass</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>36</td>\n",
       "      <td>40</td>\n",
       "      <td>44</td>\n",
       "      <td>47</td>\n",
       "      <td>51</td>\n",
       "      <td>54</td>\n",
       "      <td>58</td>\n",
       "      <td>63</td>\n",
       "      <td>100</td>\n",
       "      <td>pass</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>pass</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>83.822852</td>\n",
       "      <td>78.601951</td>\n",
       "      <td>7.146168</td>\n",
       "      <td>2.232441</td>\n",
       "      <td>1.158417</td>\n",
       "      <td>0.741047</td>\n",
       "      <td>0.543701</td>\n",
       "      <td>0.421290</td>\n",
       "      <td>0.350262</td>\n",
       "      <td>0.318886</td>\n",
       "      <td>3.700933</td>\n",
       "      <td>1.755167</td>\n",
       "      <td>2.250443</td>\n",
       "      <td>0.214197</td>\n",
       "      <td>0.215222</td>\n",
       "      <td>0.074879</td>\n",
       "      <td>0.274995</td>\n",
       "      <td>warn</td>\n",
       "      <td>.................................................</td>\n",
       "      <td>0.275410</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pass</td>\n",
       "      <td>warn</td>\n",
       "      <td>TCTGC.T</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>TCTGC.C</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>ATA.CAG</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>44.0</td>\n",
       "      <td>21</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13</td>\n",
       "      <td>LAML</td>\n",
       "      <td>ACUTE MYELOID LEUKEMIA</td>\n",
       "      <td>TB</td>\n",
       "      <td>Primary Blood Derived Cancer - Peripheral Blood</td>\n",
       "      <td>Total RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>unaligned</td>\n",
       "      <td>TCGA-AB-2909-03A-01T-0744-13_rnaseq_fastq.tar</td>\n",
       "      <td>6.494423e+09</td>\n",
       "      <td>6924e520152b16158c3ea24ef20550af</td>\n",
       "      <td>261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990</td>\n",
       "      <td>77952b61-6b89-4285-9443-a41246a7d8f2</td>\n",
       "      <td>ab60901b-d2ad-401a-9742-52df35ce1de0</td>\n",
       "      <td>40168b34-bbfe-4ee2-b7b5-2e33fbf64a18</td>\n",
       "      <td>AB</td>\n",
       "      <td>SRS131131</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>2014-05-13</td>\n",
       "      <td>Live</td>\n",
       "      <td>03</td>\n",
       "      <td>T</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>fasta</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          CGHub_analysisID                AliquotBarcode  \\\n",
       "3461  261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990  TCGA-AB-2909-03A-01T-0744-13   \n",
       "3462  261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990  TCGA-AB-2909-03A-01T-0744-13   \n",
       "3463  261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990  TCGA-AB-2909-03A-01T-0744-13   \n",
       "3464  261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990  TCGA-AB-2909-03A-01T-0744-13   \n",
       "\n",
       "                      fastqFilename  \\\n",
       "3461  61YETAAXX_1_ATCACG_1.fastq.gz   \n",
       "3462  61YETAAXX_1_ATCACG_2.fastq.gz   \n",
       "3463         622DDAAXX_1_1.fastq.gz   \n",
       "3464         622DDAAXX_1_2.fastq.gz   \n",
       "\n",
       "                                        FastQC_html_url  \\\n",
       "3461  https://storage.cloud.google.com/isb-cgc-open/...   \n",
       "3462  https://storage.cloud.google.com/isb-cgc-open/...   \n",
       "3463  https://storage.cloud.google.com/isb-cgc-open/...   \n",
       "3464  https://storage.cloud.google.com/isb-cgc-open/...   \n",
       "\n",
       "                                         FastQC_zip_url BasicStats  \\\n",
       "3461  https://storage.cloud.google.com/isb-cgc-open/...       pass   \n",
       "3462  https://storage.cloud.google.com/isb-cgc-open/...       pass   \n",
       "3463  https://storage.cloud.google.com/isb-cgc-open/...       pass   \n",
       "3464  https://storage.cloud.google.com/isb-cgc-open/...       pass   \n",
       "\n",
       "      numTotalSeq  numPoorSeq  minLength  maxLength  delLength  pctGC  \\\n",
       "3461     41560889           0         50         50          1     47   \n",
       "3462     83121778           0          6         50         45     47   \n",
       "3463     34582299           0         50         50          1     47   \n",
       "3464     34582299           0         50         50          1     47   \n",
       "\n",
       "     PerBaseSeqQual  PBSQ10_0  PBSQ50_0  PBSQ10_1  PBSQ50_1  PBSQ10_2  \\\n",
       "3461           pass        25        38      17.0        38      16.0   \n",
       "3462           pass        30        38      30.0        38      15.0   \n",
       "3463           pass        30        38      27.0        38      26.0   \n",
       "3464           pass        28        38      25.0        38      25.0   \n",
       "\n",
       "      PBSQ50_2  PBSQ10_3  PBSQ50_3  PBSQ10_4  PBSQ50_4  PBSQ10_5  PBSQ50_5  \\\n",
       "3461      38.0       9.0      38.0       7.0      38.0       2.0      37.0   \n",
       "3462      38.0      12.0      37.0       9.0      37.0       2.0      37.0   \n",
       "3463      38.0      23.0      38.0      24.0      38.0      23.0      38.0   \n",
       "3464      38.0      24.0      38.0      21.0      38.0      22.0      38.0   \n",
       "\n",
       "      PBSQ10_6  PBSQ50_6  PBSQ10_7  PBSQ50_7  PBSQ10_8  PBSQ50_8  PBSQ10_9  \\\n",
       "3461       2.0      37.0       2.0      37.0       2.0      37.0       2.0   \n",
       "3462       2.0      37.0       2.0      37.0       2.0      37.0       2.0   \n",
       "3463      22.0      38.0      13.0      38.0      11.0      37.0       2.0   \n",
       "3464      18.0      37.0      13.0      37.0       2.0      37.0       2.0   \n",
       "\n",
       "      PBSQ50_9  PBSQ10_10  PBSQ50_10 PerTileSeqQual PerSeqQualScore  PSQC_0  \\\n",
       "3461      36.0          2         36            NaN            pass       2   \n",
       "3462      36.0          2         36            NaN            pass       2   \n",
       "3463      37.0          2         36            NaN            pass       2   \n",
       "3464      36.0          2         36            NaN            pass       2   \n",
       "\n",
       "      PSQC_1  PSQC_2  PSQC_3  PSQC_4  PSQC_5  PSQC_6  PSQC_7  PSQC_8  PSQC_9  \\\n",
       "3461      11      30      34      36      36      37      37      37      37   \n",
       "3462      27      32      35      36      37      37      37      38      38   \n",
       "3463      22      33      35      36      37      37      37      37      37   \n",
       "3464      21      32      34      35      36      37      37      37      37   \n",
       "\n",
       "      PSQC_10 PerBaseSeqCont   PBSC_G_0   PBSC_A_0   PBSC_T_0   PBSC_C_0  \\\n",
       "3461       38           pass  27.231418  28.232857  17.734709  26.801016   \n",
       "3462       38           fail  12.495474  63.343577   9.489971  14.670979   \n",
       "3463       38           pass  26.712025  27.831353  17.814340  27.642282   \n",
       "3464       38           pass  25.179036  27.663989  18.347905  28.809070   \n",
       "\n",
       "       PBSC_G_1   PBSC_A_1   PBSC_T_1   PBSC_C_1   PBSC_G_2   PBSC_A_2  \\\n",
       "3461  23.878398  27.687630  25.841315  22.592656  24.753503  27.022491   \n",
       "3462  60.915261  14.225736  13.044784  11.814219  24.198353  27.006609   \n",
       "3463  23.939254  27.308237  25.576351  23.176158  24.874872  26.735585   \n",
       "3464  23.644135  27.163041  25.553711  23.639114  24.500060  26.758607   \n",
       "\n",
       "       PBSC_T_2   PBSC_C_2   PBSC_G_3   PBSC_A_3   PBSC_T_3   PBSC_C_3  \\\n",
       "3461  25.430414  22.793592  23.867636  26.576534  26.730451  22.825379   \n",
       "3462  25.681176  23.113862  23.447029  26.562247  26.857087  23.133636   \n",
       "3463  25.151433  23.238110  23.873062  26.338144  26.456724  23.332070   \n",
       "3464  25.328191  23.413142  23.653301  26.365013  26.522033  23.459653   \n",
       "\n",
       "       PBSC_G_4   PBSC_A_4   PBSC_T_4   PBSC_C_4   PBSC_G_5   PBSC_A_5  \\\n",
       "3461  23.769961  26.504633  26.626973  23.098432  23.905519  26.758423   \n",
       "3462  23.456119  26.401801  26.787168  23.354911  23.428193  26.603210   \n",
       "3463  23.843169  26.340744  26.295807  23.520281  23.804896  26.553735   \n",
       "3464  23.694560  26.204949  26.415956  23.684535  23.644564  26.406414   \n",
       "\n",
       "       PBSC_T_5   PBSC_C_5   PBSC_G_6   PBSC_A_6   PBSC_T_6   PBSC_C_6  \\\n",
       "3461  26.348721  22.987337  23.790913  26.465633  26.532799  23.210656   \n",
       "3462  26.603931  23.364666  23.439432  26.460188  26.654933  23.445447   \n",
       "3463  26.136664  23.504705  23.887982  26.220423  26.206203  23.685392   \n",
       "3464  26.290791  23.658231  23.659179  26.237727  26.272652  23.830443   \n",
       "\n",
       "       PBSC_G_7   PBSC_A_7   PBSC_T_7   PBSC_C_7   PBSC_G_8   PBSC_A_8  \\\n",
       "3461  23.805517  26.526281  26.465826  23.202376  23.770514  26.621140   \n",
       "3462  23.526332  26.566157  26.542675  23.364837  23.523397  26.445991   \n",
       "3463  23.898009  26.313621  26.092323  23.696048  23.807512  26.510711   \n",
       "3464  23.718234  26.384177  26.176677  23.720911  23.707734  26.226889   \n",
       "\n",
       "       PBSC_T_8   PBSC_C_8   PBSC_G_9   PBSC_A_9   PBSC_T_9   PBSC_C_9  \\\n",
       "3461  26.395636  23.212710  23.661634  26.474574  26.580701  23.283091   \n",
       "3462  26.512676  23.517937  23.541807  26.330577  26.670436  23.457181   \n",
       "3463  26.000879  23.680898  23.717625  26.285280  26.239225  23.757871   \n",
       "3464  26.128568  23.936809  23.673562  26.092290  26.303848  23.930300   \n",
       "\n",
       "      PBSC_G_10  PBSC_A_10  PBSC_T_10  PBSC_C_10 PerSeqGCcont  PSGCC_0  \\\n",
       "3461  23.731163  26.610283  26.462177  23.196377         pass        0   \n",
       "3462  23.558229  26.442730  26.512136  23.486905         fail        0   \n",
       "3463  23.753413  26.507991  26.058258  23.680338         pass        0   \n",
       "3464  23.759317  26.301108  26.002894  23.936680         pass        0   \n",
       "\n",
       "      PSGCC_1  PSGCC_2  PSGCC_3  PSGCC_4  PSGCC_5  PSGCC_6  PSGCC_7  PSGCC_8  \\\n",
       "3461       31       36       40       43       47       50       54       57   \n",
       "3462       43       44       46       48       50       51       53       55   \n",
       "3463       31       37       40       44       47       51       54       58   \n",
       "3464       31       36       40       44       47       51       54       58   \n",
       "\n",
       "      PSGCC_9  PSGCC_10 PerBaseNcont  PBNC_1  PBNC_2  PBNC_3  PBNC_4  PBNC_5  \\\n",
       "3461       63       100         pass     0.0     0.0     0.0     0.0     0.0   \n",
       "3462       57       100         pass     0.0     0.0     0.0     0.0     0.0   \n",
       "3463       63       100         pass     0.0     0.0     0.0     0.0     0.0   \n",
       "3464       63       100         pass     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "      PBNC_6  PBNC_7  PBNC_8  PBNC_9  PBNC_10 SeqLenDist  SLD_modeVal  \\\n",
       "3461     0.0     0.0     0.0     0.0      0.0       pass           50   \n",
       "3462     0.0     0.0     0.0     0.0      0.0       warn            6   \n",
       "3463     0.0     0.0     0.0     0.0      0.0       pass           50   \n",
       "3464     0.0     0.0     0.0     0.0      0.0       pass           50   \n",
       "\n",
       "      SLD_neg10  SLD_neg9  SLD_neg8  SLD_neg7  SLD_neg6  SLD_neg5  SLD_neg4  \\\n",
       "3461        NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "3462        NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "3463        NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "3464        NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "\n",
       "      SLD_neg3  SLD_neg2  SLD_neg1  SLD_mode  SLD_pos1  SLD_pos2  SLD_pos3  \\\n",
       "3461       NaN       NaN       NaN       1.0       NaN       NaN       NaN   \n",
       "3462       NaN       NaN       NaN       0.5       0.0       0.0       0.0   \n",
       "3463       NaN       NaN       NaN       1.0       NaN       NaN       NaN   \n",
       "3464       NaN       NaN       NaN       1.0       NaN       NaN       NaN   \n",
       "\n",
       "      SLD_pos4  SLD_pos5  SLD_pos6  SLD_pos7  SLD_pos8  SLD_pos9  SLD_pos10  \\\n",
       "3461       NaN       NaN       NaN       NaN       NaN       NaN        NaN   \n",
       "3462       0.0       0.0       0.0       0.0       0.0       0.0        0.0   \n",
       "3463       NaN       NaN       NaN       NaN       NaN       NaN        NaN   \n",
       "3464       NaN       NaN       NaN       NaN       NaN       NaN        NaN   \n",
       "\n",
       "     SeqDupLev  Total_Dedup_Pct  Dedup_relPct_1  Dedup_relPct_2  \\\n",
       "3461      pass        80.327895       73.951186        8.739390   \n",
       "3462      fail         0.410745        0.340027        0.054663   \n",
       "3463      pass        79.626184       73.229846        8.682177   \n",
       "3464      pass        83.822852       78.601951        7.146168   \n",
       "\n",
       "      Dedup_relPct_3  Dedup_relPct_4  Dedup_relPct_5  Dedup_relPct_6  \\\n",
       "3461        2.841169        1.346374        0.923370        0.634790   \n",
       "3462        0.023789        0.015458        0.011768        0.010350   \n",
       "3463        2.826697        1.434117        0.917460        0.668410   \n",
       "3464        2.232441        1.158417        0.741047        0.543701   \n",
       "\n",
       "      Dedup_relPct_7  Dedup_relPct_8  Dedup_relPct_9  Dedup_relPct_ge10  \\\n",
       "3461        0.504889        0.401431        0.351309           4.226179   \n",
       "3462        0.010206        0.009000        0.008803           0.270195   \n",
       "3463        0.559806        0.387599        0.392132           4.518259   \n",
       "3464        0.421290        0.350262        0.318886           3.700933   \n",
       "\n",
       "      Dedup_relPct_ge50  Dedup_relPct_ge100  Dedup_relPct_ge500  \\\n",
       "3461           1.895547            3.199739            0.360195   \n",
       "3462           0.342614            1.343957            0.395589   \n",
       "3463           2.061797            3.138964            0.313371   \n",
       "3464           1.755167            2.250443            0.214197   \n",
       "\n",
       "      Dedup_relPct_ge1k  Dedup_relPct_ge5k  Dedup_relPct_ge10k OverrepSeq  \\\n",
       "3461           0.298664           0.104290            0.221475       warn   \n",
       "3462           0.607665           0.342961           96.212956       fail   \n",
       "3463           0.287809           0.191097            0.390460       warn   \n",
       "3464           0.215222           0.074879            0.274995       warn   \n",
       "\n",
       "                                           OverRepSeq_1  OverRepSeq_1_Pct  \\\n",
       "3461  .................................................          0.126886   \n",
       "3462                                             ATCACG         47.417390   \n",
       "3463  .................................................          0.232136   \n",
       "3464  .................................................          0.275410   \n",
       "\n",
       "     OverRepSeq_2  OverRepSeq_2_Pct OverRepSeq_3  OverRepSeq_3_Pct AdaptCont  \\\n",
       "3461          NaN               NaN          NaN               NaN      pass   \n",
       "3462       .TCACG          0.441099       ATC.CG          0.421891      pass   \n",
       "3463          NaN               NaN          NaN               NaN      pass   \n",
       "3464          NaN               NaN          NaN               NaN      pass   \n",
       "\n",
       "     KmerCont Kmer_Seq_1  Kmer_Ratio_1 Kmer_Seq_2  Kmer_Ratio_2 Kmer_Seq_3  \\\n",
       "3461     warn    CTAT.TC     44.000000    A.TTCGA     44.000000    ACCCC.C   \n",
       "3462     warn    CG.GATT     44.000004    AC.GCGC     44.000004    ACACG.C   \n",
       "3463     fail    TGGC.TG     44.000000    ACT.CCT     44.000000    ATCCTC.   \n",
       "3464     warn    TCTGC.T     44.000000    TCTGC.C     44.000000    ATA.CAG   \n",
       "\n",
       "      Kmer_Ratio_3  Kmer_Ratio_Min  Kmer_Num_Seq study  \\\n",
       "3461     44.000000            44.0            21  TCGA   \n",
       "3462     44.000004            44.0            21  TCGA   \n",
       "3463     44.000000            44.0            21  TCGA   \n",
       "3464     44.000000            44.0            21  TCGA   \n",
       "\n",
       "                           barcode disease            disease_name  \\\n",
       "3461  TCGA-AB-2909-03A-01T-0744-13    LAML  ACUTE MYELOID LEUKEMIA   \n",
       "3462  TCGA-AB-2909-03A-01T-0744-13    LAML  ACUTE MYELOID LEUKEMIA   \n",
       "3463  TCGA-AB-2909-03A-01T-0744-13    LAML  ACUTE MYELOID LEUKEMIA   \n",
       "3464  TCGA-AB-2909-03A-01T-0744-13    LAML  ACUTE MYELOID LEUKEMIA   \n",
       "\n",
       "     sample_type                                 sample_type_name  \\\n",
       "3461          TB  Primary Blood Derived Cancer - Peripheral Blood   \n",
       "3462          TB  Primary Blood Derived Cancer - Peripheral Blood   \n",
       "3463          TB  Primary Blood Derived Cancer - Peripheral Blood   \n",
       "3464          TB  Primary Blood Derived Cancer - Peripheral Blood   \n",
       "\n",
       "     analyte_type library_type   center  center_name  platform platform_name  \\\n",
       "3461    Total RNA      RNA-Seq  BCCAGSC          NaN  ILLUMINA      Illumina   \n",
       "3462    Total RNA      RNA-Seq  BCCAGSC          NaN  ILLUMINA      Illumina   \n",
       "3463    Total RNA      RNA-Seq  BCCAGSC          NaN  ILLUMINA      Illumina   \n",
       "3464    Total RNA      RNA-Seq  BCCAGSC          NaN  ILLUMINA      Illumina   \n",
       "\n",
       "       assembly                                       filename    files_size  \\\n",
       "3461  unaligned  TCGA-AB-2909-03A-01T-0744-13_rnaseq_fastq.tar  6.494423e+09   \n",
       "3462  unaligned  TCGA-AB-2909-03A-01T-0744-13_rnaseq_fastq.tar  6.494423e+09   \n",
       "3463  unaligned  TCGA-AB-2909-03A-01T-0744-13_rnaseq_fastq.tar  6.494423e+09   \n",
       "3464  unaligned  TCGA-AB-2909-03A-01T-0744-13_rnaseq_fastq.tar  6.494423e+09   \n",
       "\n",
       "                              checksum                           analysis_id  \\\n",
       "3461  6924e520152b16158c3ea24ef20550af  261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990   \n",
       "3462  6924e520152b16158c3ea24ef20550af  261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990   \n",
       "3463  6924e520152b16158c3ea24ef20550af  261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990   \n",
       "3464  6924e520152b16158c3ea24ef20550af  261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990   \n",
       "\n",
       "                                aliquot_id  \\\n",
       "3461  77952b61-6b89-4285-9443-a41246a7d8f2   \n",
       "3462  77952b61-6b89-4285-9443-a41246a7d8f2   \n",
       "3463  77952b61-6b89-4285-9443-a41246a7d8f2   \n",
       "3464  77952b61-6b89-4285-9443-a41246a7d8f2   \n",
       "\n",
       "                            participant_id  \\\n",
       "3461  ab60901b-d2ad-401a-9742-52df35ce1de0   \n",
       "3462  ab60901b-d2ad-401a-9742-52df35ce1de0   \n",
       "3463  ab60901b-d2ad-401a-9742-52df35ce1de0   \n",
       "3464  ab60901b-d2ad-401a-9742-52df35ce1de0   \n",
       "\n",
       "                                 sample_id tss_id sample_accession  \\\n",
       "3461  40168b34-bbfe-4ee2-b7b5-2e33fbf64a18     AB        SRS131131   \n",
       "3462  40168b34-bbfe-4ee2-b7b5-2e33fbf64a18     AB        SRS131131   \n",
       "3463  40168b34-bbfe-4ee2-b7b5-2e33fbf64a18     AB        SRS131131   \n",
       "3464  40168b34-bbfe-4ee2-b7b5-2e33fbf64a18     AB        SRS131131   \n",
       "\n",
       "       published    uploaded    modified state sample_type_code  \\\n",
       "3461  2014-05-13  2014-05-13  2014-05-13  Live               03   \n",
       "3462  2014-05-13  2014-05-13  2014-05-13  Live               03   \n",
       "3463  2014-05-13  2014-05-13  2014-05-13  Live               03   \n",
       "3464  2014-05-13  2014-05-13  2014-05-13  Live               03   \n",
       "\n",
       "     analyte_type_code            platform_full_name file_type reason  \\\n",
       "3461                 T  Illumina Genome Analyzer IIx     fasta    NaN   \n",
       "3462                 T  Illumina Genome Analyzer IIx     fasta    NaN   \n",
       "3463                 T  Illumina Genome Analyzer IIx     fasta    NaN   \n",
       "3464                 T  Illumina Genome Analyzer IIx     fasta    NaN   \n",
       "\n",
       "     reagent_vendor reagent_name catalog_number is_custom target_file_url  \\\n",
       "3461            NaN          NaN            NaN       NaN             NaN   \n",
       "3462            NaN          NaN            NaN       NaN             NaN   \n",
       "3463            NaN          NaN            NaN       NaN             NaN   \n",
       "3464            NaN          NaN            NaN       NaN             NaN   \n",
       "\n",
       "     probe_file_url  \n",
       "3461            NaN  \n",
       "3462            NaN  \n",
       "3463            NaN  \n",
       "3464            NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf[ndf.CGHub_analysisID == '261c8bb3-0c4e-4cfd-bcaa-c0d156cf2990']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Files number analysis for each analysis_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2    10505\n",
       "1      648\n",
       "4      186\n",
       "6        1\n",
       "Name: study, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf.groupby('CGHub_analysisID').count()['study'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So interestingly, some analysis have more than 2 files (e.g. 4 & 6)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "count_df = ndf.groupby('CGHub_analysisID').count()['study']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CGHub_analysisID\n",
       "00065a62-5e18-4223-a884-12fca053a109    2\n",
       "0009f27d-a053-49b0-a030-ba9c1469224f    2\n",
       "Name: study, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CGHub_analysisID\n",
       "eb28bd24-74a2-4c68-83d9-c0982c7688b4    6\n",
       "Name: study, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_df[count_df == 6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CGHub_analysisID</th>\n",
       "      <th>AliquotBarcode</th>\n",
       "      <th>fastqFilename</th>\n",
       "      <th>center</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_name</th>\n",
       "      <th>platform_full_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20520</th>\n",
       "      <td>eb28bd24-74a2-4c68-83d9-c0982c7688b4</td>\n",
       "      <td>TCGA-BR-8592-01A-11R-2402-13</td>\n",
       "      <td>C1V55ACXX_7_AACTTG_1.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20521</th>\n",
       "      <td>eb28bd24-74a2-4c68-83d9-c0982c7688b4</td>\n",
       "      <td>TCGA-BR-8592-01A-11R-2402-13</td>\n",
       "      <td>C1V55ACXX_7_AACTTG_2.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20522</th>\n",
       "      <td>eb28bd24-74a2-4c68-83d9-c0982c7688b4</td>\n",
       "      <td>TCGA-BR-8592-01A-11R-2402-13</td>\n",
       "      <td>D1TBWACXX_8_AACTTG_1.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20523</th>\n",
       "      <td>eb28bd24-74a2-4c68-83d9-c0982c7688b4</td>\n",
       "      <td>TCGA-BR-8592-01A-11R-2402-13</td>\n",
       "      <td>D1TBWACXX_8_AACTTG_2.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20524</th>\n",
       "      <td>eb28bd24-74a2-4c68-83d9-c0982c7688b4</td>\n",
       "      <td>TCGA-BR-8592-01A-11R-2402-13</td>\n",
       "      <td>D20LMACXX_2_AACTTG_1.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20525</th>\n",
       "      <td>eb28bd24-74a2-4c68-83d9-c0982c7688b4</td>\n",
       "      <td>TCGA-BR-8592-01A-11R-2402-13</td>\n",
       "      <td>D20LMACXX_2_AACTTG_2.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           CGHub_analysisID                AliquotBarcode  \\\n",
       "20520  eb28bd24-74a2-4c68-83d9-c0982c7688b4  TCGA-BR-8592-01A-11R-2402-13   \n",
       "20521  eb28bd24-74a2-4c68-83d9-c0982c7688b4  TCGA-BR-8592-01A-11R-2402-13   \n",
       "20522  eb28bd24-74a2-4c68-83d9-c0982c7688b4  TCGA-BR-8592-01A-11R-2402-13   \n",
       "20523  eb28bd24-74a2-4c68-83d9-c0982c7688b4  TCGA-BR-8592-01A-11R-2402-13   \n",
       "20524  eb28bd24-74a2-4c68-83d9-c0982c7688b4  TCGA-BR-8592-01A-11R-2402-13   \n",
       "20525  eb28bd24-74a2-4c68-83d9-c0982c7688b4  TCGA-BR-8592-01A-11R-2402-13   \n",
       "\n",
       "                       fastqFilename   center disease            disease_name  \\\n",
       "20520  C1V55ACXX_7_AACTTG_1.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "20521  C1V55ACXX_7_AACTTG_2.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "20522  D1TBWACXX_8_AACTTG_1.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "20523  D1TBWACXX_8_AACTTG_2.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "20524  D20LMACXX_2_AACTTG_1.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "20525  D20LMACXX_2_AACTTG_2.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "\n",
       "        platform_full_name  \n",
       "20520  Illumina HiSeq 2000  \n",
       "20521  Illumina HiSeq 2000  \n",
       "20522  Illumina HiSeq 2000  \n",
       "20523  Illumina HiSeq 2000  \n",
       "20524  Illumina HiSeq 2000  \n",
       "20525  Illumina HiSeq 2000  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = ['CGHub_analysisID', 'AliquotBarcode', 'fastqFilename',\n",
    "        'center', 'disease', 'disease_name', 'platform_full_name']\n",
    "ndf[ndf.analysis_id == 'eb28bd24-74a2-4c68-83d9-c0982c7688b4'][cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# analysis ids with 4 fastq files\n",
    "n4_analysis_ids = count_df[count_df == 4].index.values\n",
    "n4_ndf = ndf[ndf.CGHub_analysisID.isin(n4_analysis_ids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BCCAGSC    744\n",
       "Name: center, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n4_ndf[cols].center.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So these analysis ids with over 2 fastqfiles are all from BCCAGSC. See an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['014ee344-2844-4bc4-842e-e6d6e1e8fd9b',\n",
       "       '019ce4dd-8274-4b2f-b55c-d69dc1a1604a',\n",
       "       '020e7211-cc75-4d68-86fa-65b697ed180a'], dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n4_ndf.CGHub_analysisID.unique()[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CGHub_analysisID</th>\n",
       "      <th>AliquotBarcode</th>\n",
       "      <th>fastqFilename</th>\n",
       "      <th>center</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_name</th>\n",
       "      <th>platform_full_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>014ee344-2844-4bc4-842e-e6d6e1e8fd9b</td>\n",
       "      <td>TCGA-AB-2882-03A-01T-0740-13</td>\n",
       "      <td>7043VAAXX_3_1.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>LAML</td>\n",
       "      <td>ACUTE MYELOID LEUKEMIA</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>014ee344-2844-4bc4-842e-e6d6e1e8fd9b</td>\n",
       "      <td>TCGA-AB-2882-03A-01T-0740-13</td>\n",
       "      <td>7043VAAXX_3_2.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>LAML</td>\n",
       "      <td>ACUTE MYELOID LEUKEMIA</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>014ee344-2844-4bc4-842e-e6d6e1e8fd9b</td>\n",
       "      <td>TCGA-AB-2882-03A-01T-0740-13</td>\n",
       "      <td>7046WAAXX_6_1.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>LAML</td>\n",
       "      <td>ACUTE MYELOID LEUKEMIA</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>014ee344-2844-4bc4-842e-e6d6e1e8fd9b</td>\n",
       "      <td>TCGA-AB-2882-03A-01T-0740-13</td>\n",
       "      <td>7046WAAXX_6_2.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>LAML</td>\n",
       "      <td>ACUTE MYELOID LEUKEMIA</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         CGHub_analysisID                AliquotBarcode  \\\n",
       "120  014ee344-2844-4bc4-842e-e6d6e1e8fd9b  TCGA-AB-2882-03A-01T-0740-13   \n",
       "121  014ee344-2844-4bc4-842e-e6d6e1e8fd9b  TCGA-AB-2882-03A-01T-0740-13   \n",
       "122  014ee344-2844-4bc4-842e-e6d6e1e8fd9b  TCGA-AB-2882-03A-01T-0740-13   \n",
       "123  014ee344-2844-4bc4-842e-e6d6e1e8fd9b  TCGA-AB-2882-03A-01T-0740-13   \n",
       "\n",
       "              fastqFilename   center disease            disease_name  \\\n",
       "120  7043VAAXX_3_1.fastq.gz  BCCAGSC    LAML  ACUTE MYELOID LEUKEMIA   \n",
       "121  7043VAAXX_3_2.fastq.gz  BCCAGSC    LAML  ACUTE MYELOID LEUKEMIA   \n",
       "122  7046WAAXX_6_1.fastq.gz  BCCAGSC    LAML  ACUTE MYELOID LEUKEMIA   \n",
       "123  7046WAAXX_6_2.fastq.gz  BCCAGSC    LAML  ACUTE MYELOID LEUKEMIA   \n",
       "\n",
       "               platform_full_name  \n",
       "120  Illumina Genome Analyzer IIx  \n",
       "121  Illumina Genome Analyzer IIx  \n",
       "122  Illumina Genome Analyzer IIx  \n",
       "123  Illumina Genome Analyzer IIx  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n4_ndf[n4_ndf.CGHub_analysisID == '014ee344-2844-4bc4-842e-e6d6e1e8fd9b'][cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "See only two disease are involved, 152 analysis IDs for LAML and 34 ones for STAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "LAML    152.0\n",
       "STAD     34.0\n",
       "Name: disease, dtype: float64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n4_ndf.disease.value_counts() / 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Illumina Genome Analyzer IIx    736\n",
       "Illumina HiSeq 2000               8\n",
       "Name: platform_full_name, dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n4_ndf.platform_full_name.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CGHub_analysisID</th>\n",
       "      <th>AliquotBarcode</th>\n",
       "      <th>fastqFilename</th>\n",
       "      <th>center</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_name</th>\n",
       "      <th>platform_full_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13007</th>\n",
       "      <td>94bf72c7-711b-4213-8a19-87c299fd1fd9</td>\n",
       "      <td>TCGA-BR-8676-01A-11R-2402-13</td>\n",
       "      <td>C1V55ACXX_5_AATAGG_1.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13008</th>\n",
       "      <td>94bf72c7-711b-4213-8a19-87c299fd1fd9</td>\n",
       "      <td>TCGA-BR-8676-01A-11R-2402-13</td>\n",
       "      <td>C1V55ACXX_5_AATAGG_2.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13009</th>\n",
       "      <td>94bf72c7-711b-4213-8a19-87c299fd1fd9</td>\n",
       "      <td>TCGA-BR-8676-01A-11R-2402-13</td>\n",
       "      <td>D1TBWACXX_6_AATAGG_1.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13010</th>\n",
       "      <td>94bf72c7-711b-4213-8a19-87c299fd1fd9</td>\n",
       "      <td>TCGA-BR-8676-01A-11R-2402-13</td>\n",
       "      <td>D1TBWACXX_6_AATAGG_2.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19858</th>\n",
       "      <td>e3b1348f-8ad6-48c4-8f3e-1e4a414032e2</td>\n",
       "      <td>TCGA-BR-8686-01A-11R-2402-13</td>\n",
       "      <td>C1V55ACXX_6_AAGACT_1.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19859</th>\n",
       "      <td>e3b1348f-8ad6-48c4-8f3e-1e4a414032e2</td>\n",
       "      <td>TCGA-BR-8686-01A-11R-2402-13</td>\n",
       "      <td>C1V55ACXX_6_AAGACT_2.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19860</th>\n",
       "      <td>e3b1348f-8ad6-48c4-8f3e-1e4a414032e2</td>\n",
       "      <td>TCGA-BR-8686-01A-11R-2402-13</td>\n",
       "      <td>D1TBWACXX_7_AAGACT_1.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19861</th>\n",
       "      <td>e3b1348f-8ad6-48c4-8f3e-1e4a414032e2</td>\n",
       "      <td>TCGA-BR-8686-01A-11R-2402-13</td>\n",
       "      <td>D1TBWACXX_7_AAGACT_2.fastq.gz</td>\n",
       "      <td>BCCAGSC</td>\n",
       "      <td>STAD</td>\n",
       "      <td>STOMACH ADENOCARCINOMA</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           CGHub_analysisID                AliquotBarcode  \\\n",
       "13007  94bf72c7-711b-4213-8a19-87c299fd1fd9  TCGA-BR-8676-01A-11R-2402-13   \n",
       "13008  94bf72c7-711b-4213-8a19-87c299fd1fd9  TCGA-BR-8676-01A-11R-2402-13   \n",
       "13009  94bf72c7-711b-4213-8a19-87c299fd1fd9  TCGA-BR-8676-01A-11R-2402-13   \n",
       "13010  94bf72c7-711b-4213-8a19-87c299fd1fd9  TCGA-BR-8676-01A-11R-2402-13   \n",
       "19858  e3b1348f-8ad6-48c4-8f3e-1e4a414032e2  TCGA-BR-8686-01A-11R-2402-13   \n",
       "19859  e3b1348f-8ad6-48c4-8f3e-1e4a414032e2  TCGA-BR-8686-01A-11R-2402-13   \n",
       "19860  e3b1348f-8ad6-48c4-8f3e-1e4a414032e2  TCGA-BR-8686-01A-11R-2402-13   \n",
       "19861  e3b1348f-8ad6-48c4-8f3e-1e4a414032e2  TCGA-BR-8686-01A-11R-2402-13   \n",
       "\n",
       "                       fastqFilename   center disease            disease_name  \\\n",
       "13007  C1V55ACXX_5_AATAGG_1.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "13008  C1V55ACXX_5_AATAGG_2.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "13009  D1TBWACXX_6_AATAGG_1.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "13010  D1TBWACXX_6_AATAGG_2.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "19858  C1V55ACXX_6_AAGACT_1.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "19859  C1V55ACXX_6_AAGACT_2.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "19860  D1TBWACXX_7_AAGACT_1.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "19861  D1TBWACXX_7_AAGACT_2.fastq.gz  BCCAGSC    STAD  STOMACH ADENOCARCINOMA   \n",
       "\n",
       "        platform_full_name  \n",
       "13007  Illumina HiSeq 2000  \n",
       "13008  Illumina HiSeq 2000  \n",
       "13009  Illumina HiSeq 2000  \n",
       "13010  Illumina HiSeq 2000  \n",
       "19858  Illumina HiSeq 2000  \n",
       "19859  Illumina HiSeq 2000  \n",
       "19860  Illumina HiSeq 2000  \n",
       "19861  Illumina HiSeq 2000  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n4_ndf[n4_ndf.platform_full_name == 'Illumina HiSeq 2000'][cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
